<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Short Variant Calling</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:30px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Short Variant Calling</h1>
<h3 class="subtitle">NGS workflow</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Rapidly dropping sequencing costs and the ability to obtain information about the entire genetic code has made whole genome sequening (WGS) a powerful research tool for detecting genetic variation. This workshop will take you thourgh the process of calling short variants (SNPs and indels) in whole geonome sequence data from three samples.</p>
<ul>
<li>The main part of the workshop will guide you through a basic short variant calling workflow for just one sample. The aim is to get familiar with the bam and vcf file formats, and how to interpret the results of variant calling in Integrative Genomics Viewer (IGV).</li>
<li>Next, if you have time, we will expand the workflow and perform joint variant calling on three samples. Here we will also give you an idea of how you can combine individual linux commands into a workflow that can be started as an SBATCH script.</li>
<li>The last part of the exercise will take you thourgh some additional steps that are recommended by GATK best practise short variant detection.</li>
</ul>
</div>
<div id="samples" class="section level1">
<h1><span class="header-section-number">2</span> Samples</h1>
<p>The 1000 Genomes Project was the first project to sequence the entire genomes of a large number of people, to provide a comprehensive resource on human genetic variation. In the first phase of the project each sample was seqcuended to 4x genome coverage. At this depth, sequencing can not discover all variants in each sample, but can allow the detection of most variants with frequencies of 1% in the 1000 Genomes cohort. Data from the 1000 Genomes Project available through freely accessible public databases, and in this workshop we will use 3 samples from the low converage phase of the 1000 Genomes project.</p>
<table>
<thead>
<tr class="header">
<th>Sample</th>
<th>Population</th>
<th>Coverage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>HG00097</td>
<td>European</td>
<td>30X</td>
</tr>
<tr class="even">
<td>HG00100</td>
<td>European</td>
<td>30X</td>
</tr>
<tr class="odd">
<td>HG00100</td>
<td>European</td>
<td>30X</td>
</tr>
</tbody>
</table>
</div>
<div id="the-lct-locus" class="section level1">
<h1><span class="header-section-number">3</span> The LCT locus</h1>
<p>We will work with data for a small regions on chromosome 2 around the LCT gene. Genetic variants in this region control how lactose is tolerated in adults, and the variant rs4988235 (located at at position chr2:136608596 in HG19) has been shown to lead to lactose persistence in adults. To learn more about the genetic bases for lactose tolerance please read the first three pages of this publication by Mattar et al. The variant (rs4988235) is here referred to as LCT-13910C&gt;T. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3401057/pdf/ceg-5-113.pdf">Lactose intolerance: diagnosis, genetic, and clinical factors</a>.</p>
</div>
<div id="data" class="section level1">
<h1><span class="header-section-number">4</span> Data</h1>
<p>All input data for this exercise is located in the folder:</p>
<pre class="bash"><code>/sw/courses/ngsintro/reseq/data
# These files are read-only so that the files are not accidentally deleted</code></pre>
<div id="the-reference-genome" class="section level2">
<h2><span class="header-section-number">4.1</span> The reference genome</h2>
<p>Before we can align our sample we need a reference genome. For our exercises, we’ll use only human chromosome 2, however in a real situation you should use the full genome. Create a symbolic link to the reference in your workspace:</p>
<pre class="bash"><code>ln -s /sw/courses/ngsintro/reseq/data/ref/human_g1k_v37_chr2.fasta</code></pre>
<p>Do the same with the fastq files</p>
<pre class="bash"><code>ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_1.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_2.fq</code></pre>
</div>
<div id="running-commands" class="section level2">
<h2><span class="header-section-number">4.2</span> Running commands</h2>
<p>Throughout the exercises, we will illustrate commands in the format:</p>
<pre class="bash"><code>command &lt;parameter1&gt; &lt;parameter2&gt; ...</code></pre>
<p>This signifies that you should replace <code>&lt;parameter&gt;</code> with the correct parameter type, for example your input file name, output file name, directory name, etc. If you don’t know which parameter you should supply, please ask.</p>
<div class="boxy boxy-lightbulb">
<p><strong>General guide</strong></p>
<p>
<ul>
<li>Use tab completion when possible.</li>
<li>Running a command without parameters will, often, return a default help message on how to run the command.</li>
<li>Copying and pasting commands from the exercise to terminal can result in formatting errors. Especially backslash (<code>\</code>), which is used to show that the command continues on the next line, can be a bit tricky.</li>
<li>A line starting with <code>#</code> is a comment</li>
<li>Once a command has resulted in successful completion, save it! You will redo the procedure again with another sample and this will save time.</li>
<li>If you change the node you are working on you will need to reload the tool modules. (See <strong>Accessing programs</strong> below)</li>
<li>Check that the output file exists and is a reasonable size (use <code>ls -l</code>) after a command is completed as a quick way to see that nothing is wrong. A common mistake people make is to attempt to load input files that do not exist or create output files where they cannot write.</li>
<li>Giving good names to your outfiles that describes what has been done will help you.</li>
<li>Again, replace <code>&lt;parameter&gt;</code> with the correct parameter in the commands</li>
<li>Google errors, someone in the world has run into EXACTLY the same problem you had and asked about it on a forum somewhere.</li>
</ul>
</p>
</div>
</div>
<div id="using-uppmax" class="section level2">
<h2><span class="header-section-number">4.3</span> Using Uppmax</h2>
<p>Connect to UPPMAX.</p>
<pre class="bash"><code>$ ssh -Y username@rackham.uppmax.uu.se</code></pre>
<p>Book a node. Make sure you only do this once depeding on the day.</p>
<p>Use this reservation on day 1 of variant-calling.</p>
<pre class="bash"><code>salloc -A g2019031 -t 04:00:00 -p core -n 1 --no-shell --reservation=g2019031_29 &amp;</code></pre>
<p>Use this reservation on day 2 of variant-calling.</p>
<pre class="bash"><code>salloc -A g2019031 -t 04:00:00 -p core -n 1 --no-shell --reservation=g2019031_30 &amp;</code></pre>
<p>Once your job allocation has been granted (should not take long) you can connect to the node using <code>ssh</code>. To find out the name of your node, use:</p>
<pre class="bash"><code>squeue -u &lt;username&gt;
# Node name is found under nodelist header. You should only see one node here.</code></pre>
<p>and to connect to the node:</p>
<pre class="bash"><code>ssh -Y &lt;nodename&gt;</code></pre>
</div>
<div id="accessing-programs" class="section level2">
<h2><span class="header-section-number">4.4</span> Accessing programs</h2>
<p>We’re going to need several programs that are installed in the module system. To access the bioinformatics modules you first need to load the <code>bioinfo-tools</code> module:</p>
<pre class="bash"><code>module load bioinfo-tools</code></pre>
<p>This makes it possible to load the individual programs we need:</p>
<pre class="bash"><code>module load bwa/0.7.17
module load samtools/1.8
module load GATK/3.7
module load picard/2.10.3
# It is not necessary to add version number, but it is recommended if you want to rerun the exact same commands later.
# Picard and GATK are java programs, which means that we need the path to the program file, therefore UPPMAX sets a variable when you load these modules ($GATK_HOME or $PICARD_HOME).</code></pre>
</div>
<div id="creating-a-workspace" class="section level2">
<h2><span class="header-section-number">4.5</span> Creating a workspace</h2>
<p>During this lab you should work in your folder under the course’s nobackup folder, just like you have done during the previous labs. Start by creating a workspace for this exercise in your folder, and then move into it.</p>
<pre class="bash"><code>mkdir /proj/g2019031/nobackup/&lt;username&gt;/ngsworkflow
cd /proj/g2019031/nobackup/&lt;username&gt;/ngsworkflow</code></pre>
</div>
<div id="data-for-this-lab" class="section level2">
<h2><span class="header-section-number">4.6</span> Data for this lab</h2>
<p>All input data for this exercise is located in the folder:</p>
<pre class="bash"><code>/sw/courses/ngsintro/vc/data/
# These files are read-only so that the files are not accidentally deleted</code></pre>
<p>Bam and VCF files can be reached at this new location on Uppmax:</p>
<pre class="bash"><code>$ cd /sw/courses/ngsintro/g2019031/vc</code></pre>
</div>
<div id="the-reference-genome-1" class="section level2">
<h2><span class="header-section-number">4.7</span> The reference genome</h2>
<p>Before we can align our sample we need a reference genome. For our exercises, we’ll use only human chromosome 2, however in a real situation you should use the full genome. Create a symbolic link to the reference in your workspace:</p>
<pre class="bash"><code>ln -s /sw/courses/ngsintro/reseq/data/ref/human_g1k_v37_chr2.fasta</code></pre>
<p>Do the same with the fastq files</p>
<pre class="bash"><code>ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_1.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_2.fq</code></pre>
<div id="bwa-index" class="section level3">
<h3><span class="header-section-number">4.7.1</span> BWA index</h3>
<p>Some associated files are expected by the aligner. You will only need to create these files once.</p>
<pre class="bash"><code>bwa index -a bwtsw human_g1k_v37_chr2.fasta
# BWA is a program that takes different commands as the first argument, here index, used with the bwtsw algorithm</code></pre>
<p>Check to see that several new files have been created using <code>ls -l</code>.</p>
</div>
<div id="samtools-and-picard-indices" class="section level3">
<h3><span class="header-section-number">4.7.2</span> Samtools and Picard indices</h3>
<p>We will also build two other sequence dictionaries for the reference, which just lists the names and lengths of all the chromosomes. Other programs will need these later.</p>
<pre class="bash"><code>samtools faidx human_g1k_v37_chr2.fasta
java -Xmx7g -jar $PICARD_HOME/picard.jar CreateSequenceDictionary R=human_g1k_v37_chr2.fasta O=human_g1k_v37_chr2.dict</code></pre>
</div>
</div>
<div id="data-pre-processing" class="section level2">
<h2><span class="header-section-number">4.8</span> Data pre-processing</h2>
<div id="aligning-reads" class="section level3">
<h3><span class="header-section-number">4.8.1</span> Aligning reads</h3>
<p><img src="data/ngs-workflow/wf_align.png" /></p>
<p>We are skipping the quality control and trimming of reads for this exercise, but two excellent tools are <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> &amp; <a href="http://multiqc.info/">MultiQC</a>. You will get a chance to try these tools later.</p>
<p>Should we instead run the first sampe (R1 and R2) in fastqc?</p>
<pre class="bash"><code>module load FastQC/0.11.8
mkdir fastqc
fastqc -q HG00097_1.fq HG00097_2.fq -o fastqc</code></pre>
<p>We are going to start by aligning the reads for the HG00097 sample, using <code>bwa mem</code>. We give the paths to our reference genome and the forward (<code>_1.fq</code>) and reverse (<code>_2.fq</code>) read fastq files.</p>
<p>At the same time we add something called read groups to the reads in the fastq files. Read groups allow us to trace various technical features, such as which flowcell that was used to generate the reads. Read gruop info needs to be in the file when we perform variant calling with HaplotypeCaller.
For a detailed description of read groups, please read this article at <a href="https://gatkforums.broadinstitute.org/gatk/discussion/6472/read-groups">GATK-forum</a>. We don’t have all of the information needed to create correct read groups since we work with public data today, but we assume that each fastq file was generated from one library preparation (called “1” below) derived from one biological sample (called “HG00097” below) that was run on one lane of a flowcell in the Illumina machine. We will add <em>toy</em> read groups corresponding to this. UPDATE!!!</p>
<p>We parse the output from BWA to samtools sort, which sorts the sam file according to chromosome position and then converts the sam file to the binary bam format. This saves space since no intermediary sam file is created!</p>
<pre class="bash"><code>bwa mem -R &quot;@RG\\tID:1\\tPU:1\\tSM:HG00097\\tLB:1\\tPL:illumina&quot; -t 1 human_g1k_v37_chr2.fasta HG00097_1.fq HG00097_2.fq | samtools sort &gt; HG00097_RG.bam
# -t 1 is the number of threads to use (the number of cores you booked). If you would have analysed the entire genome more cores and threads would have been necessary.
# You have to use a file redirect &quot;&gt;&quot; for the output, otherwise it will be written to stdout (your screen).</code></pre>
<p>Now look at the output file in IGV! With and without readgroups.</p>
<p>We now need to index this BAM, so that programs can randomly access the sorted data without reading the whole file. This creates an index file similarly named to the input BAM file, except with a .bai extension.</p>
<pre class="bash"><code>samtools index HG00097_RG.bam</code></pre>
<p>#########KOM hit!</p>
<p><i class="fas fa-clipboard-list"></i> Compare the size of the SAM and BAM files using <code>ls -l</code></p>
</div>
<div id="bam-processing" class="section level3">
<h3><span class="header-section-number">4.8.2</span> BAM Processing</h3>
<p>The reads have now been aligned to the reference, but these files can still be improved a bit before we use them for variant calling. Genome Analysis Tool Kit (GATK) is a popular set of tools used for variant calling. GATK comes with <em>best practises</em>, which gives a suggestion on how to get from raw BAM to filtered variants.</p>
<p>In brief, their current best practises for <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-">Germline short variant discovery (SNPs + Indels)</a> consists of:
* Marking duplicates
* Base quality score recalibration
* Sample wise variant calling
* Joint genotyping of all samples
* Variant filtration</p>
<p>However, this may need adjustment depending for example on your experimental setup. <strong>It is therefore important to check the best practises in each case</strong>.</p>
</div>
<div id="marking-duplicates" class="section level3">
<h3><span class="header-section-number">4.8.3</span> Marking duplicates</h3>
<p><img src="data/ngs-workflow/wf_markdup.png" /></p>
<p>We use Picard to mark duplicate reads:</p>
<pre class="bash"><code>java -Xmx7g -jar $PICARD_HOME/picard.jar MarkDuplicates \
  INPUT=&lt;input_bam&gt; \
  OUTPUT=&lt;marked_bam&gt; \
  METRICS_FILE=&lt;metrics_file&gt; \
  READ_NAME_REGEX=null

# marked_bam: you need to specify an output name, which will be used in the following steps.
# It is suggested to use file names that describe what has been done, e.g. &lt;sample&gt;.dedup.bam
# The metrics file is a text file that will be created and contains some stats about duplicates.</code></pre>
<p><i class="fas fa-clipboard-list"></i> How many duplicates were detected? Check the metrics file.</p>
<p>Again, we need to index the new BAM file before proceeding.</p>
<pre class="bash"><code>samtools index &lt;marked_bam&gt;</code></pre>
<p>Create a bam file with only the duplicate reads:</p>
<pre class="bash"><code>samtools view -f 1024 HG00097.dedup.RG.bam | cut -f1 | sort | uniq &gt; HG00097.dupreads.txt
java -jar $PICARD_HOME/picard.jar SortSam INPUT=HG00097.dedup.bam OUTPUT=HG00097.sorted_by_name.bam SORT_ORDER=queryname
java -jar $PICARD_HOME/picard.jar FilterSamReads INPUT=HG00097.sorted_by_name.bam OUTPUT=HG00097.only_dups.bam FILTER=includeReadList READ_LIST_FILE=HG00097.dupreads.txt SORT_ORDER=coordinate CREATE_INDEX=true </code></pre>
</div>
</div>
<div id="variant-calling" class="section level2">
<h2><span class="header-section-number">4.9</span> Variant Calling</h2>
<p><img src="data/ngs-workflow/wf_varcall.png" /></p>
<p>Now we’ll run the GATK <code>HaplotypeCaller</code> on our BAM file and output a gVCF file.</p>
<pre class="bash"><code>java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T HaplotypeCaller \
  -R human_g1k_v37_chr2.fasta \
  --emitRefConfidence GVCF \
  -I &lt;input_bam&gt; \
  -o &lt;output&gt;.g.vcf

# The output gVCF file needs to have the ending .g.vcf
# You will have one g.vcf file per sample, so make sure the sample id is in the file name</code></pre>
</div>
<div id="further-samples" class="section level2">
<h2><span class="header-section-number">4.10</span> Further samples</h2>
<p>Rerun the workflow for additional samples. If you want to repeat the procedure you can now use your saved commands and rerun the mapping and variant calling steps for at least one more sample (HG00100 or HG00101) from the course directory before continuing with the next step. Alternatively, you can copy the .g.vcf files of the two additional samples.</p>
<pre class="bash"><code>cp /sw/courses/ngsintro/vc/data/vcf/&lt;sample&gt;.g.vcf .</code></pre>
</div>
<div id="joint-genotyping" class="section level2">
<h2><span class="header-section-number">4.11</span> Joint genotyping</h2>
<p>Now you will call genotypes from all the gVCF files produced in the previous step with <code>GenotypeGVCFs</code>. This will create one file with unfiltered variants for all samples (HG00097, HG00100 and HG00101).</p>
<pre class="bash"><code>java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T GenotypeGVCFs \
  -R human_g1k_v37_chr2.fasta \
  --variant &lt;sample1&gt;.g.vcf \
  --variant &lt;sample2&gt;.g.vcf \
  --variant &lt;sample3&gt;.g.vcf \
  -o &lt;raw_variants&gt;.vcf</code></pre>
</div>
<div id="filtering-variants" class="section level2">
<h2><span class="header-section-number">4.12</span> Filtering Variants</h2>
<p><img src="data/ngs-workflow/wf_filtration.png" /></p>
<p>The next thing we will do is filter the variants. For a large dataset with many variants it is recommended to use GATK’s VQSR technique for training filter thresholds on known variants. Since we are working with a small data set we’re going to use hard filters and <a href="https://software.broadinstitute.org/gatk/documentation/article?id=2806">parameters suggested by GATK.</a></p>
<p>The parameters are slightly different for SNPs and INDELs, which we have called together and are both in our VCF. You therefore need to perform the filtering in two separate steps: first select all SNPs and filter them, then select all indels and filter them.</p>
<div id="snps" class="section level4">
<h4><span class="header-section-number">4.12.0.1</span> SNPs</h4>
<pre class="bash"><code>java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T SelectVariants \
  -R human_g1k_v37_chr2.fasta \
  -V &lt;raw_variants&gt;.vcf \
  -selectType SNP \
  -o &lt;raw_snps&gt;.vcf

java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T VariantFiltration \
  -R human_g1k_v37_chr2.fasta \
  -V &lt;raw_snps&gt;.vcf \
  -o &lt;filtered_snps&gt;.vcf \
  --filterExpression &quot;QD &lt; 2.0&quot; --filterName QDfilter \
  --filterExpression &quot;MQ &lt; 40.0&quot; --filterName MQfilter \
  --filterExpression &quot;FS &gt; 60.0&quot; --filterName FSfilter

# Each filterName option has to immediately follow the filterExpression it matches.
# Do not worry about the warnings here, but if you have time, try to find out why you get these warnings!</code></pre>
</div>
<div id="indels" class="section level4">
<h4><span class="header-section-number">4.12.0.2</span> Indels</h4>
<pre class="bash"><code>java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T SelectVariants \
  -R human_g1k_v37_chr2.fasta \
  -V &lt;raw_variants&gt;.vcf \
  -selectType INDEL \
  -o &lt;raw_indels&gt;.vcf

java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T VariantFiltration \
  -R human_g1k_v37_chr2.fasta \
  -V &lt;raw_indels&gt;.vcf \
  -o &lt;filtered_indels&gt;.vcf \
  --filterExpression &quot;QD &lt; 2.0&quot; \
  --filterName QDfilter \
  --filterExpression &quot;FS &gt; 200.0&quot; \
  --filterName FSfilter</code></pre>
</div>
<div id="merge-snps-and-indels" class="section level3">
<h3><span class="header-section-number">4.12.1</span> Merge SNPs and Indels</h3>
<p>Once you have the filtered calls for both SNPs and INDELs, you can combine these using <code>CombineVariants</code>:</p>
<pre class="bash"><code>java -Xmx7g -jar $GATK_HOME/GenomeAnalysisTK.jar -T CombineVariants \
  -R human_g1k_v37_chr2.fasta \
  --variant:snp &lt;filtered_snps&gt;.vcf \
  --variant:indel &lt;filtered_indels&gt;.vcf \
  -o &lt;my_variants_filtered&gt;.vcf \
  -genotypeMergeOptions PRIORITIZE -priority snp,indel
</code></pre>
<p>Open your filtered VCF with <code>less</code> and page through it. It still has all the variant lines, but the FILTER column that was blank before is now filled in, with PASS or a list of the filters it failed. Note also that the filters that were run are described in the header section.</p>
</div>
</div>
</div>
<div id="analyse-your-results" class="section level1">
<h1><span class="header-section-number">5</span> Analyse your results</h1>
<p>There are several tools you can use to analyse your results, once you have your multi-sample vcf file. You can get various statistics, add annotations, view the alignments etc.</p>
<div id="vcftools" class="section level2">
<h2><span class="header-section-number">5.1</span> vcftools</h2>
<p><a href="https://vcftools.github.io/man_latest.html">Vcftools</a> can be used to work with vcf files, either for manipulation or getting statistics. We can use vcftools to remove all variants that did not pass our filters:</p>
<pre class="bash"><code>module load vcftools/0.1.15

vcftools --vcf &lt;my_variants_filtered&gt;.vcf \
  --out &lt;output_prefix&gt; \
  --remove-filtered-all \
  --recode --recode-INFO-all
# --recode is used to create a new vcf file as output, otherwise only statistics are calculated.</code></pre>
<p><i class="fas fa-comments"></i> You can now try to answer these questions:</p>
<ul>
<li>How many variants passed the filters? (Check the log file)<br />
</li>
<li>Try to find out how many variants that are indels (Hint: <code>man vcftools</code> will list available options.)</li>
</ul>
</div>
<div id="variant-annotation" class="section level2">
<h2><span class="header-section-number">5.2</span> Variant annotation</h2>
<p><img src="data/ngs-workflow/wf_annotation.png" /></p>
<p>There are several tools for variant annotation. Some of the most frequently used are VEP (Variant Effect Predictor), SnpEff and Annovar. We will try <a href="http://annovar.openbioinformatics.org/en/latest/">Annovar</a>. Note that these tools are predictors and may give slightly different results. The output is also dependent on your choice of gene set.</p>
<p>Load the module.</p>
<pre class="bash"><code>module load annovar/2017.07.16</code></pre>
<p>Annovar uses downloaded annotation tracks to annotate the input file. These can be downloaded using Annovar, however many annotation tracks are already downloaded in the uppmax installation of annovar, and will be used in this lab.</p>
<pre class="bash"><code>#$ANNOVAR_HOME is set when annovar module is loaded. To see which datasets are already downloaded
# you can use
ls $ANNOVAR_HOME/humandb/</code></pre>
<p>All datasets available through Annovar can be found <a href="http://annovar.openbioinformatics.org/en/latest/user-guide/download/">here</a>.</p>
<p>Annovar can use a vcf file as input and also output the results as a vcf file. However, to make the results easier to read we will first create an annovar formatted file and produce tab delimited results.</p>
<pre class="bash"><code>convert2annovar.pl -format vcf4 -allsample -withfreq &lt;vcffile&gt; &gt; &lt;annovarfile&gt;</code></pre>
<p>Now you are ready to annotate your variants. Here, we will use RefSeq genes, dbSNP-ids and allele frequencies from 1000 genomes.</p>
<pre class="bash"><code># Use table_annovar.pl -h to find all options

table_annovar.pl &lt;annovarfile&gt; $ANNOVAR_HOME/humandb -buildver hg19 -outfile &lt;myanno&gt; -remove -protocol refGene,snp138,1000g2015aug_all,1000g2015aug_eur,1000g2015aug_afr -operation g,f,f,f,f

# buildver: Should be the same build as the reference used for alignment.
# protocol: Datasets used for annotation
# operation: Are they gene (g), filter (f) or region (r) based annotations?</code></pre>
<p>Your results can be found in <code>&lt;myanno&gt;.hg19_multianno.txt.</code></p>
<p><i class="fas fa-comments"></i> You can now try to answer these questions:</p>
<ul>
<li>What is the position of the variant that leads to retained lactase activity in adults? It is refered to as rs4988235 in dbSNP.</li>
<li>Is the variant located within a gene, if so, which gene? (What effect could this have on the LCT gene?)</li>
<li>What are the allele frequencies of the variant in different populations?</li>
<li>What is the allele frequency in the Swedish population? See the <a href="https://swegen-exac.nbis.se">SweGen Variant Frequency Browser</a></li>
<li>Are there any coding (exonic) variants in the LCT gene?</li>
<li>Add one or more datasets and rerun the annotation!</li>
</ul>
</div>
<div id="using-igv" class="section level2">
<h2><span class="header-section-number">5.3</span> Using IGV</h2>
<p>Next, we want to look at the data. For that, we will use IGV (Integrative Genomics Viewer). You can launch it from UPPMAX using your graphical forwarding, <code>-X</code> or <code>-Y</code>, as you may have already tried, see <a href="http://www.uppmax.uu.se/support-sv/user-guides/integrative-genomics-viewer--igv--guide/">here</a>. However, here we will show how to run it on your local machine.</p>
<p>Go to the IGV <a href="https://software.broadinstitute.org/software/igv/download">download page</a>, and follow the instructions. If you have not downloaded IGV before, it will prompt you to fill in some information and agree to license. Launch the viewer through webstart. The 1.2 Gb version should be sufficient for our data.</p>
<p><strong>Download data</strong></p>
<p>We need to download some data to our local machines so IGV can find it. Open a new terminal, but do not log in to UPPMAX. You may want to create a new folder for the exercise (cd to the new folder). Now we’re going to use the command <code>scp</code> (secure copy) to copy .bam (and .bai) and vcf files. <strong>&lt;username&gt; is your UPPMAX username.</strong> It will prompt you for your UPPMAX password, then it should download your files.
You will need the bam files of all 3 samples: HG00097, HG00100 and HG00101.</p>
<pre class="bash"><code># copy bam files
scp &lt;username&gt;@rackham.uppmax.uu.se:/sw/courses/ngsintro/vc/HG00* ./
# copy vcf file
scp &lt;username&gt;@rackham.uppmax.uu.se:/sw/courses/ngsintro/vc/filtered_variants.vcf ./</code></pre>
<p>If you are missing some of the BAM files or the VCF you can do this instead:</p>
<pre class="bash"><code># Copy the files from the course directory:
scp &lt;username&gt;@rackham.uppmax.uu.se:/sw/courses/ngsintro/vc/data/bam/&lt;sample&gt;.dedup.recal.ba* ./
scp &lt;username&gt;@rackham.uppmax.uu.se:/sw/courses/ngsintro/vc/data/vcf/my_variants_filtered.vcf ./</code></pre>
<p><strong>Choose reference</strong></p>
<p>First we want to make sure we have the right reference, i.e. the same reference used for mapping the reads.</p>
<ul>
<li>In IGV, go to the popup menu in the upper left and set it to <strong>Human hg19</strong>.</li>
</ul>
<p><strong>Prepare BAMs</strong></p>
<ul>
<li>Go under the Tools menu and select <strong>Run igvtools…</strong>. Choose the command <strong>Count</strong> and then use the Browse button next to the Input File line to select the BAMs (not the bai) that you just downloaded. It will autofill the output file.</li>
<li>Now hit the Run button. This generates a .tdf file for each BAM. This allows us to see the coverage value for our BAM file even at zoomed out views.</li>
</ul>
<p><strong>Load BAM &amp; VCF</strong></p>
<p>Close the <strong>igvtools</strong> window and go back to the File menu, select <strong>Load from File…</strong> and select your BAMs (not the .bai or the .tdf), which should appear in the tracks window. Next, load the vcf file. You will have to zoom in before you can see any reads. You can either select a region by click and drag, or by typing a region (chr:from-to) or gene name in the text box at the top.</p>
<p><i class="fas fa-comments"></i> You can now try to answer these questions:</p>
<ul>
<li>What is the read length?</li>
<li>Can you find the variant shown to be associated with lactose tolerance?</li>
<li>Do you have coverage in this region in all samples?</li>
<li>Which samples do you think should avoid drinking milk? Any heterozygotes?</li>
<li>Does this agree with the genotype in the vcf file?</li>
</ul>
<p><strong>Customizing the view</strong></p>
<ul>
<li><p>Change the track features. If you control-click (or right click) on the track name at the left, you will get a popup menu with several options. For example, click the gene track and play with the view.</p></li>
<li><p>If you go to a read alignment track, you can control some useful features of the display. One is how you color the reads and another is the grouping. Are there any colored reads? Why?</p></li>
<li><p>Go under the View menu and select Preferences. Click on the Alignments tab. There are a number of things we can configure. Feel free to play with them.</p></li>
</ul>
</div>
</div>
<div id="extra-labs" class="section level1">
<h1><span class="header-section-number">6</span> Extra labs</h1>
<p>Please continue here if you have time left!</p>
<div id="ucsc-genome-browser" class="section level2">
<h2><span class="header-section-number">6.1</span> UCSC Genome Browser</h2>
<p>The UCSC Genome Browser is a collection of tools for viewing and manipulating genomic data from public databases. In this extra lab we will upload our filtered variants to the UCSC genome browser, and look at evolutionary conservation of the genomic region surrounding the lactase persistent allele. We also encourage you to explore other types of data available in the UCSC Genome Browser. Plenty of documentation is available about UCSC Genome Browser, so please look at these sites to learn more:</p>
<ul>
<li><a href="http://www.genome.ucsc.edu/training/index.html" class="uri">http://www.genome.ucsc.edu/training/index.html</a><br />
</li>
<li><a href="http://www.genome.ucsc.edu/training/ucscGeneFishing.pdf" class="uri">http://www.genome.ucsc.edu/training/ucscGeneFishing.pdf</a></li>
</ul>
<ol style="list-style-type: decimal">
<li><p>Go to <a href="https://genome.ucsc.edu" class="uri">https://genome.ucsc.edu</a>.</p></li>
<li><p>Click on <strong>Genomes</strong> in the upper menu and select <strong>Human GRCh37/hg19</strong>. This will take you to the version of the human reference assembly that we have been working with.</p></li>
<li><p>Click on <strong>My Data</strong> in the upper menu, select <strong>Custom Tracks</strong>, and click on <strong>add custom tracks</strong> on the next page.</p></li>
<li><p>Click on the <strong>Choose File</strong> button to upload data, browse to the vcf file with filtered variants on your local computer, and press <strong>Submit</strong>. Then click <strong>go</strong> to view the uploaded variants in the Genome Browser.</p></li>
<li><p>Zoom in to the relevant part of the genome. We only have data for a small region on chromosome 2 surrounding the <em>LCT</em> gene. An easy way to find the relevant region is to type <strong>LCT</strong> in the search window on the top of the page, press <strong>go</strong>, and then select <strong>LCT (uc002tuu.1) at chr2:136545415-136594750 - Homo sapiens lactase (LCT), mRNA</strong> under <strong>UCSC Genes</strong>.Then use the zoom out buttons to view both the <em>LCT</em> gene and the <em>MCM6</em> gene.</p></li>
<li><p>Zoom in to the variant that leads to retained lactase activity in adults (rs4988235). Look at the <strong>Multiz Alignment of 100 Vertebrates</strong> track that shows evolutionary conservation of this genomic region across 100 vertebrates. (It is possible to directly enter the rs-number in the search window.)</p></li>
</ol>
<div class="boxy boxy-comments">
<p>You can try to answer these questions:</p>
<ul>
<li>Which species appear to have the DNA sequence surrounding the rs4988235 variant in their genomes?<br />
</li>
<li>What allele does these species have at the variant position?</li>
</ul>
</div>
<ol start="7" style="list-style-type: decimal">
<li>Play around with the various annotations tracks that are available in the UCSC genome browser and explore what is known about this particular region of the human genome. All tracks can be displayed in different modes, from <strong>full</strong> to <strong>hide</strong> and this regulates at what detail the track will be displayed.</li>
</ol>
</div>
<div id="bedtools" class="section level2">
<h2><span class="header-section-number">6.2</span> BEDTools</h2>
<p><a href="http://bedtools.readthedocs.org/en/latest/">BEDTools</a> is a <a href="http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html">collection of tools</a> used for many different tasks in genomics. It supports different input formats, including .bam, .bed, .gff or vcf. The <a href="http://genome.ucsc.edu/FAQ/FAQformat">bed format</a> is central; each line describes a feature, where the three first columns are required (chromosome, start coord, end coord) and the rest describe the feature (e.g. a gene). We will try the <strong>intersect</strong> tool:</p>
<pre class="bash"><code>module load BEDTools/2.27.1

#Copy this bed file from the course directory and look at it using e.g. less
cp /sw/courses/ngsintro/vc/data/genes.bed .

bedtools intersect -wo -a &lt;vcffile&gt; -b &lt;bedfile&gt; &gt; &lt;subset.vcf&gt;</code></pre>
<p><i class="fas fa-comments"></i> You can now try to answer these questions:</p>
<ul>
<li>How many variants are found within the given gene regions? (Hint: use <code>wc -l</code> to count the lines in your output file.)</li>
<li>How would you find all variants that are <em>NOT</em> in genes? (Look at the available options using bedtools intersect <code>-h</code>)</li>
</ul>
</div>
<div id="samtools-bit-flag" class="section level2">
<h2><span class="header-section-number">6.3</span> Samtools &amp; bit flag</h2>
<p>Viewing duplicates and understanding the bit flag. We can look at the duplicates that we marked with Picard above, using a filter on the bit flag. The mark for duplicates is the bit for 1024, we can use samtools view to look at them. First go to <a href="https://broadinstitute.github.io/picard/explain-flags.html">this online utility</a> that is helpful for decoding sam flags. Notice that it can also be used in reverse to find the appropriate flag. Now look at your data.</p>
<pre class="bash"><code># We can count the marked reads, with samtools and the -c option.
samtools view -f 1024 -c &lt;marked_bam&gt;

# Or if we want to look at these reads:
samtools view -f 1024 &lt;marked_bam&gt; | less</code></pre>
<p><i class="fas fa-comments"></i> You can now try to answer these questions:</p>
<ul>
<li>Do you get the same number of duplicates as in the metrics file?</li>
<li>Why do we use samtools to look at the BAM file? Could we have looked at it with just less?</li>
</ul>
</div>
<div id="vep" class="section level2">
<h2><span class="header-section-number">6.4</span> VEP</h2>
<p>Variant Effect Predictor (<a href="https://www.ensembl.org/info/docs/tools/vep/index.html">VEP</a>) is another popular tool for variant annotation.</p>
<pre class="bash"><code>module load vep/92
#To see which cache is default use
#echo $VEP_CACHE

vep --cache --dir_cache $VEP_CACHE --assembly GRCh37 --offline --force_overwrite --symbol --af_1kg -i &lt;vcffile&gt; -o &lt;outfile&gt;
# To list all options see: http://www.ensembl.org/info/docs/tools/vep/script/vep_options.html</code></pre>
<p>Do you get the same annotations as with annovar?</p>
</div>
</div>
<div id="additional-information" class="section level1">
<h1><span class="header-section-number">7</span> Additional information</h1>
<ul>
<li><p>Read e.g. the first three pages of <a href="http://dx.doi.org/10.2147/CEG.S32368">Matter <em>el.al.</em>, 2012</a> to find out more details about the lactase persistence allele. The variant (rs4988235) is here referred to as <em>LCT</em>-13910C&gt;T.</p></li>
<li><p>Here is a technical documentation of <a href="https://www.illumina.com/documents/products/technotes/technote_understanding_quality_scores.pdf">Illumina Quality Scores</a></p></li>
<li><p>Tools used or referenced</p>
<ul>
<li><a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
<li><a href="http://multiqc.info/">MultiQC</a></li>
<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html">Picard</a></li>
<li><a href="https://software.broadinstitute.org/gatk/">GATK</a></li>
<li><a href="http://www.htslib.org/">samtools</a></li>
</ul></li>
</ul>
<hr />
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
