---
title: "&nbsp;"
output:
  bookdown::html_document2:
    toc: false
    number_sections: false
---

```{r,child="assets/header-lab.Rmd"}
```

```{r,eval=TRUE,include=FALSE}
library(yaml)
upid <- yaml::read_yaml("_site.yml")$uppmax_project
nscid <- yaml::read_yaml("_site.yml")$nsc_project

```

# Precourse {.tabset .tabset-fade}

This workshop is aimed towards biologists, researchers, computer scientists or data analysts with limited experience in analysing NGS data.

## Preparation

These are steps to be completed before the workshop.

### 1. Create UPPMAX and NSC accounts

Remote computing clusters (UPPMAX and NSC) will be use for data analyses. A SUPR/SNIC account is needed to use these resources.

If you do not already have one, create an account at [https://supr.snic.se/](https://supr.snic.se/).

<img src="data/common/supr-login.png" class="fancyimage" style="width:70%;">

Log in to [SUPR/SNIC](https://supr.snic.se/) and request membership to the project IDs **`r paste0(upid)`** and **`r paste0(nscid)`**.

<img src="data/common/supr-request.png" class="fancyimage" style="width:70%;">

Once you are accepted to a project, you should see that project listed under your active projects.

<img src="data/common/supr-projects.png" class="fancyimage" style="width:70%;">

Finally you need to request a login account to NSC. Login to SUPR and go to the Accounts page. Under the **Possible Resource Account Requests** heading click on **Request Account on Tetralith @ NSC** button and confirm it on the next page.

<img src="data/common/supr-tetralith.png" class="fancyimage" style="width:70%;">

Checking your request and approving your account requires some manual work, so you might have to wait for some time (up to a working day) before the next step. When the account is ready to be created, you will receive an email to your registered email address (shown in your SUPR contact information) with information on how to proceed. You will get a URL that you use to choose the password (within seven days). When that has been done, the account ready for use within 15 minutes and you can then login using your chosen password.

<div class="boxy boxy-warning boxy-exclamation">
**Note:** You will get one username & password for the account on UPPMAX, and one username and password for the account on NSC. Please keep track of both, we will tell you when to use which account during the workshop. 
</div>

### 2. Install tools for SSH & SFTP access

#### 2.1 SSH

[![](data/common/mobaxterm.png)](http://mobaxterm.mobatek.net)

You need a program to connect to a remote cluster (UPPMAX and NSC). Linux and Mac users already have terminal on their systems. If you are on a Windows system, we recommend [MobaXterm](http://mobaxterm.mobatek.net). It is recommended that you INSTALL the program and not use the portable version. MobaXterm also has an integrated SFTP file browser.

#### 2.2 XQuartz 

[![](data/common/xquartz.png)](https://www.xquartz.org/)

Mac users will need to download and install [XQuartz](https://www.xquartz.org/) for X11 forwarding. *ie*; to forward remotely opened windows to local machine.


#### 2.3 SFTP Browser

[![](data/common/filezilla.png)](https://filezilla-project.org/)

When you need to transfer data between UPPMAX and your computer, you can use the tools SCP or SFTP through the terminal. Windows users can use the SFTP browser available with MobaXterm. If you prefer a GUI to upload and download files from UPPMAX, we recommend installing [FileZilla](https://filezilla-project.org/).


### 3. Test the connection to UPPMAX

<div class="boxy boxy-primary boxy-lightbulb">
For this step, you will have to use the terminal a bit. You can get started by following **Tutorial One** at this link *[Unix tutorial for beginners](http://www.ee.surrey.ac.uk/Teaching/Unix/)*. You must use [https://scilifelab.github.io/courses/ngsintro/common/emu/](https://scilifelab.github.io/courses/ngsintro/common/emu/) [(or this mirror)](https://bellard.org/jslinux/vm.html?url=https://bellard.org/jslinux/buildroot-x86.cfg) to try the commands in the tutorial, so that you don't mess up any real world system. If you any questions regarding this tutorial contact: *martin.dahlo@scilifelab.uu.se*.
</div>

**Connect to UPPMAX**
Open the terminal (Window users can use MobaXterm) and type `ssh -Y user@rackham.uppmax.uu.se`, then enter your password. The password will not be visible as you type.

**Create a user folder**
Then make sure that you can read and write in the workshop folder at UPPMAX. Go to ``r paste0("/proj/",upid,"/nobackup/")`` and create a directory with your username. For example `mkdir jody`. You will work inside this directory for the workshop. For example ``r paste0("/proj/",upid,"/nobackup/jody")``. If you cannot write to the folder, the most likely reason is that you have not requested access to the workshop project via SUPR. This is described in step 1 above.

<div class="boxy boxy-warning boxy-exclamation">
**Note:** It may take an hour or so from request approval, before you can actually write to the folder. We will check before the workshop that all students have logged in and done this, so do not forget!
</div>

### 4. Install other tools

#### 4.1 Zoom

[![](data/common/zoom.png)](https://zoom.us/)  

Please install the latest version of [Zoom](https://zoom.us/).

#### 4.2 IGV

Please install IGV on your local system before the start of the workshop. 

<div class="boxy boxy-primary boxy-lightbulb">
[![](data/common/IGVlogo.png)](https://www.broadinstitute.org/igv/)  

**Download IGV** (Integrated Genome Browser) from the [Broad Institute](https://www.broadinstitute.org/igv/) on your own computer and have the mouse genome (mm10) as well as the human genome (hg19) available.
</div>

#### 4.3 R

We will use R on Uppmax and therefore, you do not need to install it on your own system. Optionally, if you would like to try it out on your own system, instructions are given below.

<div class="boxy boxy-clipboard-list boxy-orange">
**Optional**

Install R and RStudio on your system.

[![](data/common/r.png)](https://cloud.r-project.org/)

<i class="fas fa-chevron-circle-right"></i> **Install R statistical software** from [r-project.org](https://cloud.r-project.org/). If you have an old version of R and you do not use it, uninstall it and then install a newer version. Make sure you have one of the recent versions of R (Preferably one version older than the latest). For Windows users, it is recommended that you DO NOT install to `C:\Program Files\R\`. Instead, install to `C:\R\`.

[![](data/common/rstudio.png)](https://www.rstudio.com/products/rstudio/download/)

<i class="fas fa-chevron-circle-right"></i> **Install [RStudio](https://www.rstudio.com/products/rstudio/download/)**. RStudio provides you with tools like code editor with highlighting, project management, version control, package building, debugger, profiler and more.

<i class="fas fa-chevron-circle-right"></i> Extra R packages used in the workshop exercises (if any) are listed below. It is recommended that you install this in advance. Simply copy and paste the code into R.

For Linux and Mac users, you may have to install some extra OS specific libraries before installing the R packages. Pick the packages as needed based on your system below.

**libcurl**

* deb: `libcurl4-openssl-dev` (Debian, Ubuntu, etc)
* rpm: `libcurl-devel` (Fedora, CentOS, RHEL)
* csw: `libcurl_dev` (Solaris)

**libssl**

* deb: `libssl-dev` (Debian, Ubuntu, etc)
* rpm: `openssl-devel` (Fedora, CentOS, RHEL)
* csw: `libssl_dev` (Solaris)
* brew: `openssl@1.1` (Mac OSX)

**libxml2**

* deb: `libxml2-dev` (Debian, Ubuntu, etc)
* rpm: `libxml2-devel` (Fedora, CentOS, RHEL)
* csw: `libxml2_dev` (Solaris)

`libudunits2-dev`  
`libopenblas-base`

Install the R packages. Simply copy and paste the code into R.

```{r,eval=TRUE,block.title=NULL,echo=FALSE,comment="",class.output="r"}
# this code block reads package names from '_site.yml' and prints them as installation instruction.

pkg <- yaml::read_yaml("_site.yml")

if(!is.null(pkg$packages$packages_cran_student)) {
 cat("# install from cran\n")
 cat(paste0("install.packages(c('",paste(pkg$packages$packages_cran_student,sep="",collapse="','"),"'))"))
 cat("\n")
}

if(!is.null(pkg$packages$packages_bioc_student)) {
 cat("# install from bioconductor\n")
 cat(paste0("BiocManager::install(c('",paste(pkg$packages$packages_bioc_student,sep="",collapse="','"),"'))"))
}

if(!is.null(pkg$packages$packages_github_student)) {
 cat("# install from github\n")
 cat(paste0("remotes::install_github(c('",paste(pkg$packages$packages_github_student,sep="",collapse="','"),"'))"))
}
```

</div>

***

## Syllabus

The syllabus for this workshop are as follows.

* Working on the unix/linux command line
  * Command line navigation and related commands: cd, mkdir, rm, rmdir
  * Commonly used linux tools: cp, mv, tar, less, more, head, tail, nano, grep, top, man
  * Wildcards
  * Ownership and permissions
  * Symbolic links
  * Piping commands
* Working on remote computing cluster
  * Logging on to UPPMAX
  * Booking resources
  * Job templates, submission and queues
  * Modules
* Commonly used bioinformatic tools and pipelines
* Working with integrated genome viewer
* Variant-calling workflow
  * Mapping reads to the reference genome
  * Variant detection
  * VCF file format
* RNA-Seq workflow
  * RNA-Seq experimental design and considerations
  * QC, mapping and gene expression counts
  * Differential gene expression analyses
* Current advances in NGS technologies

***

## Learning outcomes

After this workshop you should be able to:

* Describe the basic principles of next generation sequencing.
* Use the Linux command line interface to manage simple file processing operations, and organize directory structures.
* Connect to and work on a remote high performance compute cluster.
* Apply programs in Linux for analysis of NGS data.
* Summarise the applications of current NGS technologies, including the weakness and strengths of the approaches and when it is appropriate to use which one of them.
* Explain common NGS file formats.
* Interpret quality control of NGS reads.
* Explain the steps involved in variant calling using whole genome sequencing data.
* Independently perform a basic variant calling workflow on example data.
* Explain the steps involved in differential gene expression using RNA seq data.
* Independently perform differential gene expression analysis on example data.



***

#
