---
title: "Quality control with FastQC"
---

```{r,child="assets/header-lab.Rmd"}
```

```{r,eval=TRUE,include=FALSE}
library(yaml)
upid <- yaml::read_yaml("_site.yml")$uppmax_project
upres2 <- yaml::read_yaml("_site.yml")$uppmax_res_2
```

```{css,include=FALSE}
.workLocally{
background-color: red;
}
```

```{r,include=FALSE}
## VARIABLES
#colours
fastqdir <- "/sw/courses/ngsintro/reseq/data/fastq"
col_uppmax <- "#f4f8e8"
col_local <- "#e5f4f8"
```
# Data
We will run FastQC on three low-coverage whole genome sequencing (WGS) samples from the public 1000 Genomes project. To speed up the analysis we will only use data from a small genomic region. These are the exact same samples as will be used in the variant-calling workflow lab on Wednesday.  

Sample        | Description
------------- | ---------- 
HG00097       | Low coverage WGS
HG00100       | Low coverage WGS
HG00101       | Low coverage WGS 

# Run FastQC 

## Connect to Uppmax
Please connect to the Rackham cluster on UPPMAX using `ssh`:
```bash
$ ssh -Y username@rackham.uppmax.uu.se
```  

## Create a workspace 
You should work in your folder under the course's nobackup folder, just like you have done during the previous labs. Start by creating a workspace for this exercise in your folder, and then move into it. 
```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('mkdir /proj/',upid,'/nobackup/<username>/qc\n'))
cat(paste0('cd /proj/',upid,'/nobackup/<username>/qc'))
```

## Symbolic links to data   
The raw data files are located in 
```{r,echo=FALSE,comment="",class.output="bash"}		
 cat(paste0(fastqdir))
```

Instead of copying the files to your workspace you should create symbolic links (soft-links) to them. Soft-linking files and folders allows you to work with them as if they were in your current directory, but without multiplying them.  
Create symbolic links to the fastq files in your workspace:
```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('ln -s ',fastqdir,'/HG00097_1.fq\n')) 
cat(paste0('ln -s ',fastqdir,'/HG00097_2.fq\n'))  
cat(paste0('ln -s ',fastqdir,'/HG00100_1.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00100_2.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00101_1.fq\n')) 
cat(paste0('ln -s ',fastqdir,'/HG00101_2.fq\n'))
```

## Book a node 
When you login to Rackham you enter the *login node*, which is not intended for compute intensive tasks. To be able to run analyses directly in the terminal you should therefore book a compute node (or in this case just one core of a node). Make sure you only do this once each day because we have reserved one core per student for the course.    
Use this reservation on day 1 of variant-calling:
```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('salloc -A ',upid,' -t 04:00:00 -p core -n 1 --no-shell --reservation=',upres2,' &'))
```
Once your job allocation has been granted (should not take long) you can connect to the node using `ssh`. To find out the name of your node, use:
```bash
squeue -u <username>
```
The node name is found under nodelist header, you should only see one. Connect to that node:  
```bash
ssh -Y <nodename>
```
  
## Accessing FastQC
FastQC is installed in the module system on UPPMAX. Modules must be loaded *every time* you login to Rackham, or when you connect to a new compute node.  
First load the `bioinfo-tools` module:
```bash
module load bioinfo-tools
```
This makes it possible to load FastQC:
```bash
module load FastQC/0.11.8
```
 
## Run FastQC  
Run `FastQC` on the HG00097 fastq files:
```bash
fastqc -q *.fastqc
```
The output is a .html document that shows quality scores along the reads, and other information.  
  
# Check the results
Now its time to look at the results. 

## Option 1 - Download output to your laptop
You can download the files you just generated to you laptop and open them in your webbrowser. To do that you can use this command:
```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('scp <username>@rackham.uppmax.uu.se:/proj/',upid,'/nobackup/<username>/qc/*.html .'))
```
Please replace `<username>` with your UPPMAX user name.  

## Option 2 - Look at precomputed output
Alternatively, you can look at the files that we have uploaded to the course website: 
[HG00097_1.fq](data/qc/HG00097_1_fastqc.html)
[HG00097_2.fq](data/qc/HG00097_2_fastqc.html)
[HG00100_1.fq](data/qc/HG00100_1_fastqc.html)
[HG00100_2.fq](data/qc/HG00100_2_fastqc.html)
[HG00101_1.fq](data/qc/HG00101_1_fastqc.html)
[HG00101_2.fq](data/qc/HG00101_2_fastqc.html)
  
### Questions   
1. Look at the section "Sequence Length Distribution". How long are the reads?
2. Look at the "per base sequence quality". How many bases in the reads have a median phred-score above 28?
3. Are the quality scores higher for the first strand reads or the second strand reads?

# Documentation
  * [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
