<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Quality control with FastQC</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">





</div>


<!-- rmd lab header -->
<p><br></p>
<div id="fastqc" class="section level1">
<h1><span class="header-section-number">1</span> FastQC</h1>
<p>FastQC performes a series of quality control analyses, called modules. The output is a HTML report with one section for each module, and a summary evaluation of the results in the top. Entirely normal results are marked with a green tick, slightly abnormal results raise warnings (orange exclamation mark), and very unusual results raise failures (red cross).</p>
<p>It is important to stress that although the analyses appear to give a pass/fail result, these evaluations must be taken in the context of what you expect from your library. A ‘normal’ sample as far as FastQC is concerned is random and diverse. Some experiments may be expected to produce libraries which are biased in particular ways. You should treat the summary evaluations therefore as pointers to where you should concentrate your attention and understand why your library may not look random and diverse.</p>
</div>
<div id="data" class="section level1">
<h1><span class="header-section-number">2</span> Data</h1>
<p>We will run FastQC on three low-coverage whole genome sequencing (WGS) samples from the public 1000 Genomes project. To speed up the analysis we will only use data from a small genomic region. These are the exact same samples as will be used in the variant-calling workflow lab on Wednesday.</p>
<table>
<thead>
<tr class="header">
<th>Sample</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>HG00097</td>
<td>Low coverage WGS</td>
</tr>
<tr class="even">
<td>HG00100</td>
<td>Low coverage WGS</td>
</tr>
<tr class="odd">
<td>HG00101</td>
<td>Low coverage WGS</td>
</tr>
</tbody>
</table>
</div>
<div id="run-fastqc" class="section level1">
<h1><span class="header-section-number">3</span> Run FastQC</h1>
<div id="connect-to-uppmax" class="section level2">
<h2><span class="header-section-number">3.1</span> Connect to Uppmax</h2>
<p>Please connect to the Rackham cluster on UPPMAX using ssh:</p>
<pre class="bash"><code>$ ssh -Y username@rackham.uppmax.uu.se</code></pre>
</div>
<div id="create-a-workspace" class="section level2">
<h2><span class="header-section-number">3.2</span> Create a workspace</h2>
<p>You should work in your folder under the course’s nobackup folder, just like you have done during the previous labs. Start by creating a workspace for this exercise in your folder, and then move into it.</p>
<pre class="bash"><code>mkdir /proj/g2021013/nobackup/&lt;username&gt;/qc
cd /proj/g2021013/nobackup/&lt;username&gt;/qc</code></pre>
</div>
<div id="symbolic-links-to-data" class="section level2">
<h2><span class="header-section-number">3.3</span> Symbolic links to data</h2>
<p>The raw data files are located in</p>
<pre class="bash"><code>/sw/courses/ngsintro/reseq/data/fastq</code></pre>
<p>Instead of copying the files to your workspace you should create symbolic links (soft-links) to them. Soft-linking files and folders allows you to work with them as if they were in your current directory, but without multiplying them. Create symbolic links to the fastq files in your workspace:</p>
<pre class="bash"><code>ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_1.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00097_2.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00100_1.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00100_2.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00101_1.fq
ln -s /sw/courses/ngsintro/reseq/data/fastq/HG00101_2.fq</code></pre>
</div>
<div id="book-a-node" class="section level2">
<h2><span class="header-section-number">3.4</span> Book a node</h2>
<p>To be able to run analyses in the terminal you should book a compute node (or in this case just one core of a node). Make sure you only do this once each day because we have reserved one core per student for the course. If you haven’t already reserved a core today please use this command:</p>
<pre class="bash"><code>salloc -A g2021013 -t 04:00:00 -p core -n 1 --no-shell --reservation=g2021013_10 &amp;</code></pre>
<p>Once your job allocation has been granted (should not take long) you can connect to the node using ssh. To find out the name of your node, use:</p>
<pre class="bash"><code>squeue -u &lt;username&gt;</code></pre>
<p>The node name is found under nodelist header, you should only see one. Connect to that node:</p>
<pre class="bash"><code>ssh -Y &lt;nodename&gt;</code></pre>
</div>
<div id="accessing-fastqc" class="section level2">
<h2><span class="header-section-number">3.5</span> Accessing FastQC</h2>
<p>FastQC is installed in the module system on UPPMAX. Modules must be loaded <em>every time</em> you login to Rackham, or when you connect to a new compute node.<br />
First load the bioinfo-tools module:</p>
<pre class="bash"><code>module load bioinfo-tools</code></pre>
<p>This makes it possible to load FastQC:</p>
<pre class="bash"><code>module load FastQC/0.11.8</code></pre>
</div>
<div id="run-fastqc-1" class="section level2">
<h2><span class="header-section-number">3.6</span> Run FastQC</h2>
<p>Run FastQC on all fastq files:</p>
<pre class="bash"><code>fastqc -q *.fq</code></pre>
<p>The output is .html documents that shows quality scores along the reads, and other information. Please check what new files were generated with the command:</p>
<pre class="bash"><code>ls -lrt.  </code></pre>
</div>
</div>
<div id="check-the-results" class="section level1">
<h1><span class="header-section-number">4</span> Check the results</h1>
<p>The output from FastQC is a HTML report that should be opened in a web browser. When you have X-forwarding enabled you can open it on Rackham with this command:</p>
<pre class="bash"><code>firefox --no-remote filename.html</code></pre>
<p>It can be slow to look at graphical files via SSH. We have therefore made the output files that you just created available through the links below, so that you can look at them via your local web-browser:</p>
<table>
<thead>
<tr class="header">
<th>Sample</th>
<th>Read 1</th>
<th>Read 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>HG00097</td>
<td><a href="data/qc/HG00097_1_fastqc.html" target="_blank">HG00097_1.fq</a></td>
<td><a href="data/qc/HG00097_2_fastqc.html" target="_blank">HG00097_2.fq</a></td>
</tr>
<tr class="even">
<td>HG00100</td>
<td><a href="data/qc/HG00100_1_fastqc.html" target="_blank">HG00100_1.fq</a></td>
<td><a href="data/qc/HG00100_2_fastqc.html" target="_blank">HG00100_2.fq</a></td>
</tr>
<tr class="odd">
<td>HG00101</td>
<td><a href="data/qc/HG00101_1_fastqc.html" target="_blank">HG00101_1.fq</a></td>
<td><a href="data/qc/HG00101_2_fastqc.html" target="_blank">HG00101_2.fq</a></td>
</tr>
</tbody>
</table>
<div id="per-base-sequence-quality" class="section level2">
<h2><span class="header-section-number">4.1</span> Per Base Sequence Quality</h2>
<p>This module shows the distribution of the quality scores at each position in the reads. The quality scores are represented by a Box and Whisker plot with the following elements:</p>
<ul>
<li>The central red line is the median value.</li>
<li>The yellow box represents the inter-quartile range (25-75%).</li>
<li>The upper and lower whiskers represent the 10% and 90% points</li>
<li>The blue line represents the mean quality</li>
</ul>
<p>The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red).</p>
<div id="questions" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Questions</h3>
<ol style="list-style-type: decimal">
<li>Which positions in the reads have a median phred-score above 28 (very good quality calls) in each sample?<br />
</li>
<li>Do any of the samples have warnings or failures in the Per Base Sequence Quality module?</li>
<li>Why? Please look in the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/2%20Per%20Base%20Sequence%20Quality.html" target="_blank">documentation of this module</a>.</li>
</ol>
</div>
</div>
<div id="sequence-length-distribution" class="section level2">
<h2><span class="header-section-number">4.2</span> Sequence Length Distribution</h2>
<p>This module shows the length distribution of the reads in the file.</p>
<div id="questions-1" class="section level3">
<h3><span class="header-section-number">4.2.1</span> Questions</h3>
<ol start="4" style="list-style-type: decimal">
<li>How long are the reads?</li>
<li>Do any of the samples have warnings or failures in the Sequence Length Distribution module?</li>
<li>Why? Please look in the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html" target="_blank">documentation of this module</a></li>
</ol>
</div>
</div>
</div>
<div id="answers" class="section level1">
<h1><span class="header-section-number">5</span> Answers</h1>
<p>When you have finished the exercise, please have a look at this document with <a href="lab_qc_answers.pdf" target="_blank">answers to questions</a>, and compare them with your answers.</p>
</div>
<div id="documentation" class="section level1">
<h1><span class="header-section-number">6</span> Documentation</h1>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc" target="_blank">FastQC</a></li>
<li>If you want to learn more details about FastQC please have a look at this video by the <a href="http://www.youtube.com/watch?v=bz93ReOv87Y" target="_blank">Babraham Bioinformatics Institute</a>.</li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
