<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Filetypes</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.5.1/jquery-3.5.1.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">





</div>


<!-- rmd lab header -->
<p><br></p>
<div class="boxy boxy-yellow boxy-exclamation">
<p>In code blocks, the dollar sign (<code>$</code>) is not to be printed. The dollar sign is usually an indicator that the text following it should be typed in a terminal window.</p>
</div>
<div id="connect-to-uppmax" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Connect to UPPMAX</h1>
<p>The first step of this lab is to open a ssh connection to UPPMAX. Please go to the Contents page and click “Connecting to UPPMAX” under the “Additional content” topic. Once connected to UPPMAX, return here and continue reading the instructions below.</p>
</div>
<div id="logon-to-a-node" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Logon to a node</h1>
<p>Usually you would do most of the work in this lab directly on one of the login nodes at UPPMAX, but we have arranged for you to have one core each for better performance. This was covered briefly in the lecture notes.</p>
<pre class="bash"><code>$ salloc -A snic2022-22-1124 -t 07:00:00 -p core -n 1 --no-shell &amp;</code></pre>
<p>check which node you got (replace <strong>username</strong> with your UPPMAX username)</p>
<pre class="bash"><code>$ squeue -u username</code></pre>
<p>should look something like this</p>
<pre class="bash"><code>dahlo@rackham2 work $ squeue -u dahlo
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           3132376      core       sh    dahlo  R       0:04      1 r292
dahlo@rackham2 work $</code></pre>
<p>where <strong>r292</strong> is the name of the node I got (yours will probably be different).
Note the numbers in the Time column. They show for how long the job has been running. When it reaches the time limit you requested (7 hours in this case) the session will shut down, and you will lose all unsaved data. Connect to this node from within UPPMAX.</p>
<pre class="bash"><code>$ ssh -Y r292</code></pre>
<p><i class="fas fa-lightbulb"></i> There is a UPPMAX specific tool called <code>jobinfo</code> that supplies the same kind of information as <code>squeue</code> that you can use as well (<code>$ jobinfo -u username</code>).</p>
</div>
<div id="copy-lab-files" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Copy lab files</h1>
<p>Now you will need some files. To avoid all the course participants editing the same file all at once, undoing each other’s edits, each participant will get their own copy of the needed files. The files are located in the folder <code>/sw/courses/ngsintro/linux/filetypes</code></p>
<p>Next, copy the lab files from this folder. <code>-r</code> means recursively, which means all the files including sub-folders of the source folder. Without it, only files directly in the source folder would be copied, NOT sub-folders and files in sub-folders.</p>
<p><i class="fas fa-lightbulb"></i> Remember to use tab-complete to avoid typos and too much writing.</p>
<pre class="bash"><code>$ cp -r &lt;source&gt; &lt;destination&gt;
$ cp -r /sw/courses/ngsintro/linux/filetypes /proj/snic2022-22-1124/nobackup/username/</code></pre>
<p>Have a look in <strong>/proj/snic2022-22-1124/nobackup/username/</strong>.</p>
<pre class="bash"><code>$ cp -r &lt;source&gt; &lt;destination&gt;
cd /proj/snic2022-22-1124/nobackup/username/filetypes
$ tree</code></pre>
<p>This will print a file tree, which gives you a nice overview of the folders where you are standing in. As you can see, you have a couple of files and a couple of empty folders. In the <strong>0_ref</strong> folder you have a reference genome in fasta format and annotations for the genome in GTF format. In <strong>0_seq</strong> you have a fastq file containing the reads we will align.</p>
</div>
<div id="run-pipeline" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Run pipeline</h1>
<p>The best way to see all the different file formats is to run a small pipeline and see which files we encounter along the way. The pipeline is roughly the same steps you’ll do in the variant-calling part of the course, so for now we’ll stick with the dummy pipeline which some of you might have encoutered in the <a href="lab_uppmax_pipeline.html" target="_blank">extra material</a> for the UPPMAX exercise.</p>
<p>The programs in the dummy pipeline does not actually do any analysis but they work the same way as the real deal, although slightly simplified, to get you familiar with how to work with analysis programs. The data is from a sequencing of the adenovirus genome, which is tiny compared to the human genome (36kb vs 3gb).</p>
<p>The starting point of the pipeline are fresh reads from the sequencing machine in fastq format, and a reference genome in fasta format. The goal of the exercise is to look at our aligned reads in a genome viewer together with the annotations of the adenovirus genome.</p>
<p>First, let’s go through the steps of the pipeline:</p>
<ul>
<li><strong>Build an index for the reference genome.</strong> This will speed up the alignment process. Not possible to do the analysis without it.</li>
<li><strong>Align the reads.</strong></li>
<li><strong>Convert the SAM file to a BAM file.</strong> We want to use the space efficiently.</li>
<li><strong>Sort the BAM file.</strong> We have to sort it to be able to index it.</li>
<li><strong>Index the BAM file.</strong> We have to index it to make it fast to access the data in the file.</li>
<li><strong>View the aligned data together with the annotations.</strong></li>
</ul>
<p>The first thing you usually do is to load the modules for the programs you want to run. During this exercise we’ll only run my dummy scripts that don’t actually do any analysis, so they don’t have a module of their own. What we can do instead is to manually do what module loading usually does: to modify the <strong><code>$PATH</code> variable</strong>.</p>
<p>The <code>$PATH</code> variable specifies directories where the computer should look for programs. For instance, when you type <code>nano</code>, how does the computer know which program to start? You gave it the name <code>nano</code>, but that could refer to any file named nano in the computer, yet it starts the correct one every time. The answer is that it looks in the directories stored in the <code>$PATH</code> variable.</p>
<p>To see which directories that are available by default, type</p>
<pre class="bash"><code>$ echo $PATH</code></pre>
<p>It should give you something like this, a list of directories, separated by colon signs:</p>
<pre class="bash"><code>dahlo@rackham2 work $ echo $PATH
/home/dahlo/perl//bin/:/home/dahlo/.pyenv/shims:/home/dahlo/.pyenv/bin:
/usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:
/sbin:/opt/thinlinc/bin:/sw/uppmax/bin:/home/dahlo/usr/bin</code></pre>
<p>Try loading a module, and then look at the <code>$PATH</code> variable again. You’ll see that there are a few extra directories there now, after the module has been loaded.</p>
<pre class="bash"><code>dahlo@rackham2 work $ module load bioinfo-tools samtools/1.10
dahlo@rackham2 work $ echo $PATH
/sw/apps/bioinfo/samtools/1.10/rackham/bin:/home/dahlo/perl/bin:/home/dahlo/.pyenv/shims:
/home/dahlo/.pyenv/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:
/usr/sbin:/sbin:/opt/thinlinc/bin:/sw/uppmax/bin:/home/dahlo/usr/bin</code></pre>
<p>To pretend that we are loading a module, we will just add a the directory containing my dummy scripts to the <code>$PATH</code> variable, and it will be like we loaded the module for them.</p>
<pre class="bash"><code>$ export PATH=$PATH:/sw/courses/ngsintro/linux/uppmax_pipeline_exercise/dummy_scripts</code></pre>
<p>This will set the <code>$PATH</code> variable to whatever it is at the moment, and add a directory at the end of it. Note the lack of a dollar sign in front of the variable name directly after <code>export</code>. You don’t use dollar sign when <strong>assigning</strong> values to variables, and you always use dollar signs when <strong>getting</strong> values from variables.</p>
<p><i class="fas fa-exclamation-circle"></i> <strong>IMPORTANT:</strong> The export command affects only the terminal you type it in. If you have 2 terminals open, only the terminal you typed it in will have a modified path. If you close that terminal and open a new one, it will not have the modified path.</p>
<div id="build-index" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Build index</h2>
<ol style="list-style-type: decimal">
<li><strong>Build an index for the reference genome.</strong></li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>
<p>All aligners will have to index the reference genome you are aligning your data against. This is only done once per reference genome, and then you reuse that index whenever you need it. All aligners have their own kind of index unfortunately, so you will have to build one index for each aligner you want to use. In this lab, we will use the dummy aligher called <code>align_reads</code>, and we will build a index using it’s indexing progam, called <code>reference_indexer</code>.</p>
<p>First, have a look in the <strong>0_ref</strong> folder</p>
<pre class="bash"><code>$ ll 0_ref</code></pre>
<p>You should see 2 files: the fasta file, the gtf file. Have a look at each of them with <code>less</code>, just to see how they look inside. To do the actual indexing of the genome:</p>
<p><i class="fas fa-clipboard-list"></i> Run reference_indexer.</p>
<pre class="bash"><code>Syntax: reference_indexer -r &lt;name of the fasta file you want to index&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713540674" aria-expanded="false" aria-controls="acc2023012713540674">
</button>
</p>
<div id="acc2023012713540674" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ reference_indexer -r 0_ref/ad2.fa</code></pre>
</div>
</div>
<p>Since the genome is so small this should only take a second or so. The human genome will probably take a couple of hours. Look in the <strong>0_ref folder</strong> again and see if anything has changed.</p>
<pre class="bash"><code>$ ll 0_ref</code></pre>
<p>The new file you see is the index file created by <strong>reference_indexer</strong>. This index is in the same format as you would get from the real program <strong>samtools</strong>. Try viewing the index file with <code>less</code> and see how it looks. The samtools type of index contains one row per fasta record in the reference file. In this case, there is only one record for the adenovirus genome, and it’s called <code>ad2</code> in the fasta file. The human reference genome typically have one record per chromosome, so a index of the human genome would then have 24 rows.</p>
<p>The numbers after the record name specifies how many bases the record has, how far into the file (in bytes) the record starts, the number of bases on each line in the record, and how many bytes each line takes up in the file.
Using this information the program can quickly jump to the start location of each record, without having to read the file from the first row every time.</p>
<p>Other aligners might use more complex indexing of the file to speed up the alignment process even further, e.g. creating an index over where it can find all possible “words” that you can form with 5 or so bases, making it easier to find possible matching sites for reads. If the read starts with <code>ATGTT</code> you can quickly look in the index and see all places in the geonome that contains this word and start looking if the rest of the read matches the area around the word.</p>
<p>This greatly decreases the number of places you have to look when looking for a match. These types of indexes usually take a long time to create (5+ hours maybe), but since you only have to do it once per reference genome it’s easily worth it, seeing how the alignment process probably would take 100s of hours without the index, instead of 6-12 hours.</p>
<p>We are now ready to align our reads.</p>
</div>
<div id="align-reads" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Align reads</h2>
<ol style="list-style-type: decimal">
<li>Build an index for the reference genome.</li>
<li><strong>Align the reads.</strong></li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>
<p><i class="fas fa-clipboard-list"></i> Align reads using <strong>align_reads</strong>, naming the output file <strong>ad2.sam</strong>, placed in the <strong>1_alignment</strong> folder.</p>
<pre class="bash"><code>Syntax: align_reads -r &lt;reference genome&gt; -i &lt;fastq file with reads&gt; -o &lt;name of the output file&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713541759" aria-expanded="false" aria-controls="acc2023012713541759">
</button>
</p>
<div id="acc2023012713541759" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ align_reads -r 0_ref/ad2.fa -i 0_seq/ad2.fq -o 1_alignment/ad2.sam</code></pre>
</div>
</div>
<p>This will create a SAM file in <strong>1_alignment</strong> called <strong>ad2.sam</strong>. Have a look at it with less. If you think the file looks messy, add a <code>-S</code> after less to make it stop wrapping long lines, <code>less -S 1_alignment/ad2.sam</code> and scroll sideways using the arrow keys. As you can see there is one row per aligned read in this file. Each row contains information about the read, like the name of the read, where in the reference genome it aligned, and also a copy of the reads sequence and quality score, among other things.</p>
</div>
<div id="sam-to-bam" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> SAM to BAM</h2>
<ol style="list-style-type: decimal">
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li><strong>Convert the SAM file to a BAM file.</strong></li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>
<p><i class="fas fa-clipboard-list"></i> The next step is to convert the SAM file to a BAM file. This is more or less just compressing the file, like creating a zip file. To do that we will use the dummy program <strong>sambam_tools</strong>, telling it we want to convert a file to BAM (<strong>-f bam</strong>), which file we want to convert (<strong>-i</strong>), where it should save the resulting BAM file (<strong>-o</strong>). Save the BAM file in the <strong>2_bam</strong> folder and name it <strong>ad2.bam</strong>.</p>
<pre class="bash"><code>Syntax: sambam_tool -f bam -i &lt;sam file&gt; -o &lt;bam&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713541107" aria-expanded="false" aria-controls="acc2023012713541107">
</button>
</p>
<div id="acc2023012713541107" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ sambam_tool -f bam -i 1_alignment/ad2.sam -o 2_bam/ad2.bam</code></pre>
</div>
</div>
<p>Have a look in the <strong>2_bam</strong> folder.</p>
<pre class="bash"><code>$ ll 2_bam</code></pre>
<p>The created BAM file is an exact copy of the SAM file, but stored in a much more efficient format. Aligners usually have an option to output BAM format directly, saving you the trouble to convert it yourself, but not all tools can do this (they really should though). Have a look at the difference in file size, though in this example it’s quite an extreme difference (2.9 MB vs 0.3 MB). The quality score of all reads is the same (BBBBBBBBB..), and files with less differences are easier to compress. Usually the BAM file is about 25% of the size of the SAM file.</p>
<p>Since the BAM format is a binary format we can’t look at it with <code>less</code>. We would have to use a tool, like <strong>samtools</strong> which you will probably see later in the week, to first convert the file back to a SAM file before we can read it. In that case we can just look at the SAM file before converting it since they will be the same.</p>
</div>
<div id="sort-index-bam" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> Sort &amp; index BAM</h2>
<ol style="list-style-type: decimal">
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li><strong>Sort the BAM file.</strong></li>
<li><strong>Index the BAM file.</strong></li>
<li>View the aligned data together with the annotations.</li>
</ol>
<p>A BAM file is taking up much less space than the SAM file, but we can still improve performance. An indexed BAM file is infinitely faster for programs to work with, but before we can index it, we have to sort it since it’s not possible to index an unsorted file in any meaningful way.</p>
<p><i class="fas fa-clipboard-list"></i> To sort the BAM file we’ll use the <strong>sambam_tool</strong> again, but specifying a different function, <strong>-f sort</strong> instead. Tell it to store the sorted BAM file in the <strong>3_sorted</strong> folder and name the file <strong>ad2.sorted.bam</strong></p>
<pre class="bash"><code>Syntax: sambam_tool -f sort -i &lt;unsorted bam file&gt; -o &lt;sorted bam file&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713546575" aria-expanded="false" aria-controls="acc2023012713546575">
</button>
</p>
<div id="acc2023012713546575" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ sambam_tool -f sort -i 2_bam/ad2.bam -o 3_sorted/ad2.sorted.bam</code></pre>
</div>
</div>
<p>This will sort the <strong>ad2.bam</strong> file and create a new BAM file which is sorted, called <strong>ad2.sorted.bam</strong>.</p>
<p><i class="fas fa-clipboard-list"></i> Now when we have a sorted BAM file, we can index it. Use the command</p>
<pre class="bash"><code>Syntax: sambam_tool -f index -i &lt;sorted bam file&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713546386" aria-expanded="false" aria-controls="acc2023012713546386">
</button>
</p>
<div id="acc2023012713546386" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ sambam_tool -f index -i 3_sorted/ad2.sorted.bam</code></pre>
</div>
</div>
<p>This will create an index named <strong>ad2.sorted.bam.bai</strong> in the same folder as the <strong>ad2.sorted.bam</strong> file is located. It’s nicer to have the <strong>.bam</strong> and <strong>.bai</strong> named to the same “prefix”, so rename the <strong>.bai</strong> file to not have the <strong>.bam</strong> in its name.</p>
<pre class="bash"><code>$ mv 3_sorted/ad2.sorted.bam.bai 3_sorted/ad2.sorted.bai</code></pre>
</div>
<div id="view-in-a-genome-viewer" class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> View in a genome viewer</h2>
<ol style="list-style-type: decimal">
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li><strong>View the aligned data together with the annotations.</strong></li>
</ol>
<p>Now that we have to data aligned and prepared for easy access, we will view it in a genome viewer together with the annotations for the genome. Have a look at the annotations file with <code>less</code>.</p>
<pre class="bash"><code>$ less -S 0_ref/ad2.gtf</code></pre>
<p>The <strong>-S</strong> will tell less to not wrap the lines, and instead show one line per line. If the line is longer than the window, you can user the left and right arrow to scroll to the left and right. Many tabular files are much more readable when using the <code>-S</code> option. Try viewing the file without it and see the difference.</p>
<p>To view the file, we will use the program <strong>IGV</strong> (Integrated Genome Viewer). Before we can do this, we have to load the module for IGV.</p>
<p><i class="fas fa-lightbulb"></i> If you are using a Mac you might have to install the program <a href="https://www.xquartz.org/" target="_blank">XQuartz</a>, if you have not already installed that program. By using <code>-Y</code> in your ssh command you enable graphical transfer over ssh, but you will also have to have a program able to receive the graphics in order to display it.</p>
<pre class="bash"><code>$ module load bioinfo-tools IGV/2.4.2</code></pre>
<p>Start it by typing the following command (now we’ll find out if you used <code>-Y</code> in all your ssh connections!):</p>
<pre class="bash"><code>$ igv.sh</code></pre>
<div class="boxy boxy-lightbulb boxy-orange">
<p><strong>Optional</strong></p>
<p>If you notice that IGV over Xforwarding is excruciatingly slow, you can try to use the web based ThinLinc client instead. Unfortunately this requires you to have set up a two factor authentification (2FA) with UPPMAX, so it’s something you can try on your own. <a href="https://www.uppmax.uu.se/support/user-guides/setting-up-two-factor-authentication/" target="_blank">Instructions for setting up the 2FA at UPPMAX</a>. When you are all set up, go to the address <a href="https://rackham-gui.uppmax.uu.se" target="_blank">https://rackham-gui.uppmax.uu.se</a> and login with your normal UPPMAX username and password together with your 2FA (described at the login screen). This will get you a remote desktop on one of the login nodes, and you can open a terminal and run IGV there instead. Once IGV is started, either using Xforwarding or the remote desktop in your web browser, we are ready to go.</p>
</div>
<p>There are 3 files we have to load in IGV.</p>
<p>The first is the reference genome. Press the menu button located at <strong>“Genomes - Load Genome from File…”</strong> and find your reference genome in <strong>0_ref/ad2.fa</strong>. If you are having trouble finding your files, note that IGV always starts in your home directory. Use the dropdown menu at the top to navigate to <strong>/proj/snic2022-22-1124/nobackup/…</strong>.</p>
<p>The second file you have to load is the reads. Press the menu button <strong>“File - Load from File…”</strong> and select your <strong>3_sorted/ad2.sorted.bam</strong>.</p>
<p>The last file you have to load is the annotation data. Press <strong>“File - Load from File…”</strong> again and select you annotation file in <strong>0_ref/ad2.gtf</strong>.</p>
<p>This will show you the reference genome, how all the reads are aligned to it, and all the annotation data. Try zooming in on an area and have a look at the reads and annotations. The figures you see in the picture are all derived from the data in the files you have given it.</p>
<p>At the top of the window you have the overview of the current chromosome you are looking at, which tells you the scale you are zoomed at for the moment. When you zoom in you will see a red rectangle apper which shows you which portion of the chromosome you are looking at. Just below the scale you’ll see the coverage graph, which tells you how many reads cover each position along the reference genome. The colored bands you see here and there are SNPs, i.e. positions where the reads of your sample does not match the reference genome.</p>
<p>All the reads, the larger area in the middle of the window, are drawn from the data in the BAM file using the chromosome name, the starting position and the ending position of each read. When you zoom in more you will be able to see individual reads and how they are aligned. The annotation in GTF format are all plotted using the data in the GTF file, visible just under all the reads, are shown as blue rectangles.</p>
<p>The reference genome, a fasta file containing the DNA sequence of the reference genome, is visible at the bottom of the window if you zoom to the smallest level so you can see the bases of the genome.</p>
</div>
</div>
<div id="create-a-cram-file" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Create a CRAM file</h1>
<p>The CRAM format is even more efficient than the BAM format. To create a CRAM file we’ll have to use <strong>samtools</strong>, so we will load the module for it.</p>
<pre class="bash"><code>$ module load bioinfo-tools samtools/1.10</code></pre>
<p><i class="fas fa-clipboard-list"></i> Tell samtools that you want CRAM output (<strong><code>-C</code></strong>) and specify which reference genome it should use to do the CRAM conversion (<strong><code>-T</code></strong>)</p>
<pre class="bash"><code>Syntax: samtools view -C -T &lt;reference genome&gt; -o &lt;name of cram file&gt; &lt;bam file to convert&gt;</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2023012713547270" aria-expanded="false" aria-controls="acc2023012713547270">
</button>
</p>
<div id="acc2023012713547270" class="collapse">
<div class="card card-body">
<pre class="bash"><code>$ samtools view -C -T 0_ref/ad2.fa -o 4_cram/ad2.cram 3_sorted/ad2.sorted.bam</code></pre>
</div>
</div>
<p>Compare the sizes of the convered BAM file and the newly created CRAM file:</p>
<pre class="bash"><code>$ ll -h 3_sorted/ad2.sorted.bam 4_cram/ad2.cram</code></pre>
<p>This will list both the files, and print the file size in a human readable format (<strong><code>-h</code></strong>). The CRAM file is roughly 1/3 of the size of the BAM file. This is probably because all the reads in the simulated data has the same quality value (BBBBBBBBBB). Fewer types of quality values are easier to compress, hence this amazing compression ratio. Real data will have much more diverse quality scores, and the CRAM file would be pethaps 70-80% of the original BAM file.</p>
<div class="boxy boxy-clipboard-list boxy-orange">
<p><strong>Optional</strong></p>
<p>If you have been fast to finish this lab and you still have time left (or just can’t get enough of linux stuff), please have a look at the <a href="lab_linux_advanced.html" target="_blank">advanced linux tutorial</a> where you can learn the basics in bash programming using variables, loops and control statements. This is the lab that is scheduled for after lunch today.</p>
</div>
<hr />
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
