---
title: "Read quality"
author: "Malin Larsson"
description: "Quality control of short reads with FastQC"
format: html
---

```{r,eval=TRUE,include=FALSE}
library(yaml)
library(here)
id_project <- yaml::read_yaml(here("_quarto.yml"))$id_project
path_workspace <- yaml::read_yaml(here("_quarto.yml"))$path_workspace
```

```{css,include=FALSE}
.workLocally{
background-color: red;
}
```

```{r,include=FALSE}
## VARIABLES
#colours
fastqdir <- "/sw/courses/ngsintro/vc/data/fastq"
```

## FastQC

FastQC performs a series of quality control analyses, called modules. The output is an HTML report with one section for each module, and a summary evaluation of the results in the top. Entirely normal results are marked with a green tick, slightly abnormal results raise warnings (orange exclamation mark), and very unusual results raise failures (red cross).

It is important to stress that although the analyses appear to give a pass/fail result, these evaluations must be taken in the context of what you expect from your library. A 'normal' sample as far as FastQC is concerned is random and diverse. Some experiments may be expected to produce libraries which are biased in particular ways. You should treat the summary evaluations therefore as pointers to where you should concentrate your attention and understand why your library may not look random and diverse.

## Data

We will run FastQC on three low-coverage whole genome sequencing (WGS) samples from the public 1000 Genomes project. To speed up the analysis we will only use data from a small genomic region. These are the exact same samples as will be used in the variant-calling workflow lab on Wednesday.  

Sample        | Description
------------- | ----------
HG00097       | Low coverage WGS
HG00100       | Low coverage WGS
HG00101       | Low coverage WGS

## Run FastQC

### Connect to PDC

During this lab it is best to connect to Dardel via a remote desktop (ThinLinc). Please refer to [**Connecting to PDC**](../other/lab_connect_pdc.html), section 2 Remote desktop, for instructions.

### Log on to a node

Check which node you got when you booked resources this morning (replace `username` with your username)

```bash
squeue -u username
```

should look something like this

```bash
user@login1 ~ $ squeue -u user
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           5583899    shared interact    user  R       2:22      1 nid001009
user@login1 ~ $
```

where `nid001009` is the name of the node (yours will probably be different). Note the numbers in the Time column. They show for how long the job has been running. When it reaches the time limit you requested, the session will shut down, and you will lose all unsaved data.

If the list is empty, you can run the allocation command again, and it should be in the list:

```bash
salloc -A {{< meta id_project >}} --reservation={{< meta id_slurm_res_tue >}} -t 04:00:00 -p shared -c 1
```

Connect to this node from the login node.

```bash
ssh -Y nid001009
```

### Create a workspace

You should work in a subfolder in your home folder on Dardel, just like you have done during the previous labs. 
Start by going to your course folder using this command:

```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('cd ', path_workspace))
```

Create a folder for this exercise and move into it:

```bash
mkdir qc
cd qc
```

### Symbolic links to data

The raw data files are located in

```{r,echo=FALSE,comment="",class.output="bash"}		
 cat(paste0(fastqdir))
```

Instead of copying the files to your workspace, you should create symbolic links (soft-links) to them. Soft-linking files and folders allows you to work with them as if they were in your current directory, but without creating actual copies.  Create symbolic links to the FASTQ files in your workspace:

```{r,echo=FALSE,comment="",class.output="bash"}
cat(paste0('ln -s ',fastqdir,'/HG00097_1.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00097_2.fq\n'))  
cat(paste0('ln -s ',fastqdir,'/HG00100_1.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00100_2.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00101_1.fq\n'))
cat(paste0('ln -s ',fastqdir,'/HG00101_2.fq\n'))
```

### Inspecting the raw reads

The files come from paired-end sequencing, where each fragment is
sequenced from two ends.
The two reads are sometimes called R1 and R2,
and are typically stored in two different files with file names
that have a `1` and a `2` in their names.
Use the program `less` to look at the contents of the FASTQ file containing the
R2 reads of the `HG00101` sample:
```bash
less HG00101_2.fq
```
While `less` is running, use the arrow keys to scroll up and down.
Use the `Q` to quit the program when you are done.

### Accessing FastQC

FastQC is installed in the module system on PDC. Modules must be loaded *every time* you login to Dardel, or when you connect to a new compute node.  
First load the bioinfo-tools module:

```bash
module load bioinfo-tools
```

This makes it possible to load FastQC:

```bash
module load fastqc/0.12.1
```

### Run FastQC

Run FastQC on all FASTQ files:

```bash
fastqc -q *.fq
```

The output is one `.html` document for each input FASTQ file, which shows quality scores along the reads and other information. Please check what new files were generated with the command:

```bash
ls -lrt
```

## Check the results

The output from FastQC is an HTML report that should be opened in a web browser.

One option is to download the output to your local computer and open the file in your local web browser.  
To do this, first open a new terminal window on your local computer and navigate to a folder where you want to keep the files.   
Then download all generated `.html` files to that folder with this command (replace `username` with your username):
```bash
scp username@dardel.pdc.kth.se:{{< meta path_workspace >}}/qc/*.html .
```
The files can now be opened from your local web browser by navigating to the same folder.  

If you have connected to PDC via ThinLinc, you can open HTML files on Dardel with this command:

```bash
firefox --no-remote filename.html &
```

We have also made the output files that you just created available through the links below so that you can look at them via your local web-browser:

Sample        | Read 1 | Read 2
------------- | ---------- | --------
HG00097       | [HG00097_1.fq](assets/HG00097_1_fastqc.html) | [HG00097_2.fq](assets/HG00097_2_fastqc.html)
HG00100       | [HG00100_1.fq](assets/HG00100_1_fastqc.html) | [HG00100_2.fq](assets/HG00100_2_fastqc.html)
HG00101       | [HG00101_1.fq](assets/HG00101_1_fastqc.html) | [HG00101_2.fq](assets/HG00101_2_fastqc.html)

### Per Base Sequence Quality

This module shows the distribution of the quality scores at each position in the reads. The quality scores are represented by a Box and Whisker plot with the following elements:

- The central red line is the median value.
- The yellow box represents the inter-quartile range (25-75%).
- The upper and lower whiskers represent the 10% and 90% points
- The blue line represents the mean quality

The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red).

### Trim qualities

Use the quality-trimming functionality of Cutadapt to remove low-quality 3' read ends of the HG00101 sample.

First, load the appropriate module:

```bash
module load cutadapt
```

Then run this Cutadapt with the `-q` option:
```bash
cutadapt -q 20 -o HG00101_trimmed_1.fq -p HG00101_trimmed_2.fq HG00101_1.fq HG00101_2.fq
```

Generate new FastQC reports for the trimmed reads:
```bash
fastqc -q HG00101_trimmed_*.fq
```

Finally, inspect the `HG00101_trimmed_2.fq` file using `less`:
```bash
less HG00101_trimmed_2.fq
```

#### Questions

1. Which positions in the reads have a median phred-score above 28 (very good quality calls) in each sample?  
2. Do any of the samples have warnings or failures in the Per Base Sequence Quality module?
3. Why? Please look in the [documentation of this module](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/2%20Per%20Base%20Sequence%20Quality.html).  
4. How do quality scores change for HG00101 after quality trimming?

### Sequence Length Distribution

This module shows the length distribution of the reads in the file.  

#### Questions

5. How long are the reads?
6. Do any of the samples have warnings or failures in the Sequence Length Distribution module?
7. Why? Please look in the [documentation of this module](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html)
8. How does the length distribution change for HG00101 after quality trimming?
9. What is the first read whose quality was trimmed in the `HG00101_trimmed_2.fq` file?

## Answers  

When you have finished the exercise, please have a look at this document with [answers to questions](lab_qc_answers.pdf), and compare them with your answers.  

## Documentation

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc)
- If you want to learn more details about FastQC please have a look at this video by the [Babraham Bioinformatics Institute]( http://www.youtube.com/watch?v=bz93ReOv87Y).
