<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>RNA-Seq Analyses</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">RNA-Seq Analyses</h1>
<h3 class="subtitle">NGS workflow</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<!-- ------------ Only edit title, subtitle & author above this ------------ -->
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>RNA-seq has become a powerful approach to study the continually changing cellular transcriptome. Here, one of the most common questions is to identify genes that are differentially expressed between two conditions, e.g. controls and treatment. The <strong>main</strong> exercise in this tutorial will take you through a basic bioinformatic analysis pipeline to answer just that, it will show you how to find differentially expressed (DE) genes.</p>
<p><strong>Main exercise</strong></p>
<ul>
<li>01 Check the quality of the raw reads with <strong>FastQC</strong></li>
<li>02 Map the reads to the reference genome using <strong>HISAT2</strong></li>
<li>03 Assess the post-alignment quality using <strong>QualiMap</strong></li>
<li>04 Count the reads overlapping with genes using <strong>featureCounts</strong></li>
<li>05 Find DE genes using <strong>DESeq2</strong> in R</li>
</ul>
<p>RNA-seq experiment does not necessarily end with a list of DE genes. If you have time after completing the <strong>main</strong> exercise, try one (or more) of the <strong>bonus</strong> exercises. The <strong>bonus</strong> exercises can be run independently of each other, so choose the one that matches your interest. Bonus sections are listed below.</p>
<p><strong>Bonus exercises</strong></p>
<ul>
<li>01 Functional annotation of DE genes using <strong>GO/Reactome</strong> databases</li>
<li>02 RNA-Seq figures and plots using <strong>R</strong></li>
<li>03 Visualisation of RNA-seq BAM files using <strong>IGV</strong> genome browser</li>
</ul>
<div class="boxy boxy-primary boxy-exclamation">
<p><strong>General guide</strong></p>
<ul>
<li><p>In paths, remember to change <code>&lt;username&gt;</code> with your actual UPPMAX username.</p></li>
<li><p>You are welcome to try your own solutions to the problems, before checking the solution. Click the <code style="background-color:#4C979F;color:white;">+</code> button to see the suggested solution. There is more than one way to complete a task. Discuss with person next to you and ask us when in doubt.</p></li>
<li><p>Input code blocks are displayed like shown below. The code language is displayed above the block. Shell scripts (<strong>SH</strong>) are to be executed in the linux terminal such as bash. <strong>R</strong> scripts are to be run in R after running R` in the terminal or in R/RStudio locally. Output text blocks are marked as <strong>OUTPUT</strong>.</p></li>
</ul>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>command</code></pre>
</div>
</div>
</div>
<div id="data-description" class="section level1">
<h1><span class="header-section-number">2</span> Data description</h1>
<p>The data used in this exercise is from the paper: <strong>Poitelon, Yannick, <em>et al</em>. YAP and TAZ control peripheral myelination and the expression of laminin receptors in Schwann cells. <a href="https://www.nature.com/articles/nn.4316">Nature neuroscience 19.7 (2016): 879</a></strong>. In this study, YAP and TAZ genes were knocked-down in Schwann cells to study myelination, using the sciatic nerve in mice as a model.</p>
<p>Myelination is essential for nervous system function. Schwann cells interact with neurons and the basal lamina to myelinate axons using receptors, signals and transcription factors. Hippo pathway is a conserved pathway involved in cell contact inhibition, and it acts to promote cell proliferation and inhibits apoptosis. The pathway integrates mechanical signals (cell polarity, mechanotransduction, membrane tension) and gene expression response. In addition to its role in organ size control, the Hippo pathway has been implicated in tumorigenesis, for example its deregulation occurs in a broad range of human carcinomas. Transcription co-activators YAP and TAZ are two major downstream effectors of the Hippo pathway, and have redundant roles in transcriptional activation.</p>
<p>The material for RNA-seq was collected from 2 conditions (<strong>Wt</strong> and <strong>KO</strong>), each with 3 biological replicates.</p>
<table class="table table-striped table-hover table-responsive" style="width: auto !important; ">
<thead>
<tr>
<th style="text-align:right;">
Accession
</th>
<th style="text-align:right;">
Condition
</th>
<th style="text-align:right;">
Replicate
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
SRR3222409
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #fb8072 !important;">KO</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #80b1d3 !important;">1</span>
</td>
</tr>
<tr>
<td style="text-align:right;">
SRR3222410
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #fb8072 !important;">KO</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #fdb462 !important;">2</span>
</td>
</tr>
<tr>
<td style="text-align:right;">
SRR3222411
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #fb8072 !important;">KO</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #bebada !important;">3</span>
</td>
</tr>
<tr>
<td style="text-align:right;">
SRR3222412
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8dd3c7 !important;">Wt</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #80b1d3 !important;">1</span>
</td>
</tr>
<tr>
<td style="text-align:right;">
SRR3222413
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8dd3c7 !important;">Wt</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #fdb462 !important;">2</span>
</td>
</tr>
<tr>
<td style="text-align:right;">
SRR3222414
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #8dd3c7 !important;">Wt</span>
</td>
<td style="text-align:right;">
<span style=" font-weight: bold;    color: white !important;border-radius: 4px; padding-right: 4px; padding-left: 4px; background-color: #bebada !important;">3</span>
</td>
</tr>
</tbody>
</table>
<div class="boxy boxy-primary boxy-exclamation">
<p>For the purpose of this tutorial, to shorten the time needed to run various bioinformatics steps, we have picked reads for a single chromosome (Chr 19) and downsampled the reads. We randomly sampled, without replacement, 25% reads from each sample, using <code>fastq-sample</code> from the toolset <a href="https://homes.cs.washington.edu/~dcjones/fastq-tools/">fastq-tools</a>.</p>
</div>
</div>
<div id="main-exercise" class="section level1">
<h1><span class="header-section-number">3</span> Main exercise</h1>
<p>The main exercise covers Differential Gene Expression (DGE) workflow from raw reads to a list of differentially expressed genes.</p>
<div id="using-uppmax" class="section level2">
<h2><span class="header-section-number">3.1</span> Using Uppmax</h2>
<p>Connect to UPPMAX.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ssh -Y username@rackham.uppmax.uu.se</code></pre>
</div>
<p>Book a node.</p>
<p>For the RNA-Seq part of the course, we will work on the Rackham cluster. A standard compute node on cluster Rackham has 128 GB of RAM and 20 cores. We will use 1 core per person for this session. Therefore, each core gives you 6.4 GB of RAM. The code below is valid to run at the start of the day. If you are running it in the middle of a day, you need to decrease the time (<code>-t</code>). Do not run this twice and also make sure you are not running computations on a login node.</p>
<p>Book resources for RNA-Seq day 1.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>salloc -A g2020009 -t 03:00:00 -p core -n 1 --reservation=g2020009_12</code></pre>
</div>
<p>Book resources for RNA-Seq day 2.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>salloc -A g2020009 -t 04:00:00 -p core -n 1 --reservation=g2020009_13</code></pre>
</div>
<div id="set-up-directory" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Set-up directory</h3>
<p>Setting up the directory structure is an important step as it helps to keep our raw data, intermediate data and results in an organised manner. All work must be carried out at this location <code>/proj/g2020009/nobackup/&lt;username&gt;/</code> where <code>&lt;username&gt;</code> is your user name.</p>
<p>Create a directory named <code>rnaseq</code>. All RNA-Seq related activities must be carried out in this sub-directory named <code>rnaseq</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>mkdir rnaseq</code></pre>
</div>
<p><i class="fas fa-clipboard-list"></i> Set up the below directory structure in your project directory.</p>
<pre><code>&lt;username&gt;/
rnaseq/
  +-- 1_raw/
  +-- 2_fastqc/
  +-- 3_mapping/
  +-- 4_qualimap/
  +-- 5_dge/
  +-- 6_multiqc/
  +-- reference/
  |   +-- mouse_chr19_hisat2/
  +-- scripts/
  +-- funannot/
  +-- plots</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216454719" aria-expanded="false" aria-controls="acc2020102216454719">
</button>
</p>
<div id="acc2020102216454719" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cd rnaseq
mkdir 1_raw 2_fastqc 3_mapping 4_qualimap 5_dge 6_multiqc reference scripts funannot plots
cd reference
mkdir mouse_chr19_hisat2
cd ..</code></pre>
</div>
</div>
</div>
<p>The <code>1_raw</code> directory will hold the raw fastq files (soft-links). <code>2_fastqc</code> will hold FastQC outputs. <code>3_mapping</code> will hold the mapping output files. <code>4_qualimap</code> will hold the QualiMap output files. <code>5_dge</code> will hold the counts from featureCounts and all differential gene expression related files. <code>6_multiqc</code> will hold MultiQC outputs. <code>reference</code> directory will hold the reference genome, annotations and aligner indices. The <code>funannot</code> and <code>plots</code> directory are optional for bonus steps.</p>
<p><i class="fas fa-lightbulb"></i> It might be a good idea to open an additional terminal window. One to navigate through directories and another for scripting in the <code>scripts</code> directory.</p>
</div>
<div id="create-symbolic-links" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Create symbolic links</h3>
<p>We have the raw fastq files in this remote directory: <code>/sw/courses/ngsintro/rnaseq/main/1_raw/</code>. We are going to create symbolic links (soft-links) for these files from our <code>1_raw</code> directory to the remote directory. We do this because fastq files tend to be large files and simply copying them would use up a lot of storage space. Soft-linking files and folders allows us to work with those files as if they were actually there. Use <code>pwd</code> to check if you are standing in the correct directory. You should be standing here:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/1_raw</code></pre>
<p>Run below to create softlinks. Note that the command ends in a space followed by a period.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ln -s /sw/courses/ngsintro/rnaseq/main/1_raw/*.gz .</code></pre>
</div>
<p>Check if your files have linked correctly. You should be able to see as below.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>SRR3222409-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222409-19_1.fq.gz
SRR3222409-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222409-19_2.fq.gz
SRR3222410-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222410-19_1.fq.gz
SRR3222410-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222410-19_2.fq.gz
SRR3222411-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222411-19_1.fq.gz
SRR3222411-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222411-19_2.fq.gz
SRR3222412-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222412-19_1.fq.gz
SRR3222412-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222412-19_2.fq.gz
SRR3222413-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222413-19_1.fq.gz
SRR3222413-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222413-19_2.fq.gz
SRR3222414-19_1.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222414-19_1.fq.gz
SRR3222414-19_2.fq.gz -&gt; /sw/courses/ngsintro/rnaseq/main/1_raw/SRR3222414-19_2.fq.gz</code></pre>
</div>
</div>
</div>
<div id="fastqc" class="section level2">
<h2><span class="header-section-number">3.2</span> FastQC</h2>
<p><strong>Quality check using FastQC</strong></p>
<p>After receiving raw reads from a high throughput sequencing centre it is essential to check their quality. <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> provides a simple way to do some quality control check on raw sequence data. It provides a modular set of analyses which you can use to get a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.</p>
<p><i class="fas fa-clipboard-list"></i> Change into the <code>2_fastqc</code> directory. Use <code>pwd</code> to check if you are standing in the correct directory. You should be standing here:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/2_fastqc</code></pre>
<p>Load Uppmax modules <code>bioinfo-tools</code> and FastQC <code>FastQC/0.11.5</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load FastQC/0.11.8</code></pre>
</div>
<p>Once the module is loaded, FastQC program is available through the command <code>fastqc</code>. Use <code>fastqc --help</code> to see the various parameters available to the program. We will use <code>-o</code> to specify the output directory path and finally, the name of the input fastq file to analyse. The syntax will look like below.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>fastqc -o . ../1_raw/filename.fq.gz</code></pre>
</div>
<p>Based on the above command, we will write a bash loop to process all fastq files in the directory. Writing multi-line commands through the terminal can be a pain. Therefore, we will run larger scripts from a bash script file. Move to your <code>scripts</code> directory and create a new file named <code>fastqc.sh</code>.</p>
<p>You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/scripts</code></pre>
<p>The command below creates a new file in the current directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>touch fastqc.sh</code></pre>
</div>
<p>Use a text editor (<code>nano</code>,<code>Emacs</code>,<code>gedit</code> etc.) to edit <code>fastqc.sh</code>.</p>
<p><i class="fas fa-lightbulb"></i> <code>gedit</code> behaves like a regular text editor with a standard graphical interface.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>gedit fastqc.sh&amp;</code></pre>
</div>
<p><i class="fas fa-lightbulb"></i> Adding <code>&amp;</code> at the end sends that process to the background, so that the console is free to accept new commands.</p>
<p>Then add the lines below and save the file.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

module load bioinfo-tools
module load FastQC/0.11.8

for i in ../1_raw/*.gz
do
    echo &quot;Running $i ...&quot;
    fastqc -o . &quot;$i&quot;
done</code></pre>
</div>
<p>This script loops through all files ending in .gz. In each iteration of the loop, it executes fastqc on the file. The <code>-o .</code> flag to fastqc indicates that the output must be exported in this current directory.</p>
<p>While standing in the <code>2_fastqc</code> directory, run the file <code>fastqc.sh</code>. Use <code>pwd</code> to check if you are standing in the correct directory.</p>
<p>You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/2_fastqc</code></pre>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/fastqc.sh</code></pre>
</div>
<p>After the fastqc run, there should be a <code>.zip</code> file and a <code>.html</code> file for every fastq file. The <code>.html</code> file is the report that you need. Open the <code>.html</code> in the browser and view it. You do not need to necessarily look at all files now. We will do a comparison with all samples when using the MultiQC tool.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>firefox file.html &amp;</code></pre>
</div>
<div class="boxy boxy-clipboard-list boxy-orange">
<p><strong>Optional</strong></p>
<p>Download the <code>.html</code> file to your computer and view it.</p>
<p><i class="fas fa-lightbulb"></i> All users can use an SFTP browser like <a href="https://filezilla-project.org/">Filezilla</a> or <a href="https://cyberduck.io/">Cyberduck</a> for a GUI interface. Windows users can also use the MobaXterm SFTP file browser to drag and drop.</p>
<p>Linux and Mac users can use SFTP or SCP by running the below command in a <strong>LOCAL</strong> terminal and <strong>NOT</strong> on Uppmax. Open a terminal locally on your computer, move to a suitable download directory and run the command below.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>scp user@rackham.uppmax.uu.se:/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/2_fastqc/SRR3222409_1_fastqc.html ./</code></pre>
</div>
</div>
<p><i class="fas fa-clipboard-list"></i> Go back to the FastQC website and compare your report with the sample report for <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html">Good Illumina data</a> and <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html">Bad Illumina data</a>.</p>
<p><i class="fas fa-comments"></i> Discuss based on your reports, whether your data is of good enough quality and/or what steps are needed to fix it.</p>
</div>
<div id="hisat2" class="section level2">
<h2><span class="header-section-number">3.3</span> HISAT2</h2>
<p><strong>Mapping reads using HISAT2</strong></p>
<p>After verifying that the quality of the raw sequencing reads is acceptable, we will map the reads to the reference genome. There are many mappers/aligners available, so it may be good to choose one that is adequate for your type of data. Here, we will use a software called <a href="http://www.ccb.jhu.edu/software/hisat/index.shtml">HISAT2</a> (Hierarchical Indexing for Spliced Alignment of Transcripts) as it is a good general purpose splice-aware aligner. It is fast and does not require a lot of RAM. Before we begin mapping, we need to obtain genome reference sequence (<code>.fasta</code> file) and build an aligner index. Due to time constraints, we will build an index only on chromosome 19.</p>
<div id="get-reference" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Get reference</h3>
<p>It is best if the reference genome (<code>.fasta</code>) and annotation (<code>.gtf</code>) files come from the same source to avoid potential naming conventions problems. It is also good to check in the manual of the aligner you use for hints on what type of files are needed to do the mapping. We will not be using the annotation (<code>.gtf</code>) during mapping, but we will use it during quantification.</p>
<p><i class="fas fa-comments"></i> What is the idea behind building an aligner index? What files are needed to build one? Where do we take them from? Could one use an index that was generated for another project? Check out the <a href="http://www.ccb.jhu.edu/software/hisat/manual.shtml">HISAT2 manual</a> indexer section for answers. Browse through <a href="https://www.ensembl.org/index.html">Ensembl</a> and try to find the files needed. Note that we are working with Mouse (<em>Mus musculus</em>).</p>
<p><i class="fas fa-clipboard-list"></i> Move into the <code>reference</code> directory and download the Chr 19 genome (<code>.fasta</code>) file and the genome-wide annotation file (<code>.gtf</code>) from Ensembl.</p>
<p>You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/reference</code></pre>
<p>You are most likely to use the <a href="https://www.ensembl.org/index.html">latest version</a> of ensembl release genome and annotations when starting a new analysis. For this exercise, we will choose ensembl version 99.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>wget ftp://ftp.ensembl.org/pub/release-99/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.19.fa.gz
wget ftp://ftp.ensembl.org/pub/release-99/gtf/mus_musculus/Mus_musculus.GRCm38.99.gtf.gz</code></pre>
</div>
<p>Decompress the files for use.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>gunzip Mus_musculus.GRCm38.dna.chromosome.19.fa.gz
gunzip Mus_musculus.GRCm38.99.gtf.gz</code></pre>
</div>
<p>From the full gtf file, we will also extract chr 19 alone to create a new gtf file for use later.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cat Mus_musculus.GRCm38.99.gtf | grep -E &quot;^#|^19&quot; &gt; Mus_musculus.GRCm38.99-19.gtf</code></pre>
</div>
<p>Check what you have in your directory.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216454228" aria-expanded="false" aria-controls="acc2020102216454228">
</button>
</p>
<div id="acc2020102216454228" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
</div>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>drwxrwsr-x 2 user gXXXXXXX 4.0K Jan 22 21:59 mouse_chr19_hisat2
-rw-rw-r-- 1 user gXXXXXXX  26M Jan 22 22:46 Mus_musculus.GRCm38.99-19.gtf
-rw-rw-r-- 1 user gXXXXXXX 771M Jan 22 22:10 Mus_musculus.GRCm38.99.gtf
-rw-rw-r-- 1 user gXXXXXXX  60M Jan 22 22:10 Mus_musculus.GRCm38.dna.chromosome.19.fa</code></pre>
</div>
</div>
<div id="build-index" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Build index</h3>
<p>Move into the <code>reference</code> directory if not already there. Load module HISAT2. Remember to load <code>bioinfo-tools</code> if you haven’t done so already.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216459098" aria-expanded="false" aria-controls="acc2020102216459098">
</button>
</p>
<div id="acc2020102216459098" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load HISAT2/2.1.0</code></pre>
</div>
</div>
</div>
<p><i class="fas fa-lightbulb"></i> To search for the tool or other versions of a tool, use <code>module spider fastqc</code>.</p>
<p>Create a new bash script in your <code>scripts</code> directory named <code>hisat2_index.sh</code> and add the following lines:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

# load module
module load bioinfo-tools
module load HISAT2/2.1.0

hisat2-build \
  -p 1 \
  Mus_musculus.GRCm38.dna.chromosome.19.fa \
  mouse_chr19_hisat2/mouse_chr19_hisat2</code></pre>
</div>
<p>The above script builds a HISAT2 index using the command <code>hisat2-build</code>. It should use 1 core for computation. The paths to the <code>.fasta</code> genome file is specified. And the output path is specified. The output describes the output directory and the prefix for output files. The output directory must be present before running this script. Check <code>hisat2-build --help</code> for the arguments and descriptions.</p>
<p>Use <code>pwd</code> to check if you are standing in the correct directory. Then, run the script from the <code>reference</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/hisat2_index.sh</code></pre>
</div>
<p>Once the indexing is complete, move into the <code>mouse_chr19_hisat2</code> directory and make sure you have all the files.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216457249" aria-expanded="false" aria-controls="acc2020102216457249">
</button>
</p>
<div id="acc2020102216457249" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
</div>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 23M Sep 19 12:30 mouse_chr19_hisat2.1.ht2
-rw-rw-r-- 1 user gXXXXXXX 14M Sep 19 12:30 mouse_chr19_hisat2.2.ht2
-rw-rw-r-- 1 user gXXXXXXX  53 Sep 19 12:29 mouse_chr19_hisat2.3.ht2
-rw-rw-r-- 1 user gXXXXXXX 14M Sep 19 12:29 mouse_chr19_hisat2.4.ht2
-rw-rw-r-- 1 user gXXXXXXX 25M Sep 19 12:30 mouse_chr19_hisat2.5.ht2
-rw-rw-r-- 1 user gXXXXXXX 15M Sep 19 12:30 mouse_chr19_hisat2.6.ht2
-rw-rw-r-- 1 user gXXXXXXX  12 Sep 19 12:29 mouse_chr19_hisat2.7.ht2
-rw-rw-r-- 1 user gXXXXXXX   8 Sep 19 12:29 mouse_chr19_hisat2.8.ht2</code></pre>
</div>
<p>The index for the whole genome would be created in a similar manner. It just requires more time to run.</p>
</div>
<div id="map-reads" class="section level3">
<h3><span class="header-section-number">3.3.3</span> Map reads</h3>
<p>Now that we have the index ready, we are ready to map reads. Move to the directory <code>3_mapping</code>. Use <code>pwd</code> to check if you are standing in the correct directory.</p>
<p>You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/3_mapping</code></pre>
<p>We will create softlinks to the fastq files from here to make things easier.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cd 3_mapping
ln -s ../1_raw/* .</code></pre>
</div>
<p>These are the parameters that we want to specify for the mapping run:</p>
<ul>
<li>Specify the number of threads</li>
<li>Specify the full genome index path</li>
<li>Specify a summary file</li>
<li>Indicate read1 and read2 since we have paired-end reads</li>
<li>Direct the output (SAM) to a file</li>
</ul>
<p>HISAT2 aligner arguments can be obtained by running <code>hisat2 --help</code>. We should end with a script that looks like below to run one of the samples.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>hisat2 \
-p 1 \
-x ../reference/mouse_chr19_hisat2/mouse_chr19_hisat2 \
--summary-file &quot;SRR3222409-19.summary&quot; \
-1 SRR3222409-19_1.fq.gz \
-2 SRR3222409-19_1.fq.gz \ 
-S SRR3222409-19.sam</code></pre>
</div>
<p>Let’s break this down a bit. <code>-p</code> options denotes that it will use 1 thread. In a real world scenario, we might want to use 8 to 12 threads. <code>-x</code> denotes the full path (with prefix) to the aligner index we built in the previous step. <code>--summary-file</code> specifies a summary file to write alignments metrics such as mapping rate etc. <code>-1</code> and <code>-2</code> specifies the input fastq files. Both are used here because this is a paired-end sequencing. <code>-S</code> denotes the output SAM file.</p>
<p>But, we will not run it as above. We will make some more changes to it. We want to make the following changes:</p>
<ul>
<li>Rather than writing the output as a SAM file, we want to direct the output to another tool <code>samtools</code> to order alignments by coordinate and to export in BAM format</li>
<li>Generalise the above script to be used as a bash script to read any two input files and to automatically create the output filename.</li>
</ul>
<p><i class="fas fa-clipboard-list"></i> Now create a new bash script file named <code>hisat2_align.sh</code> in your <code>scripts</code> directory and add the script below to it.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216452002" aria-expanded="false" aria-controls="acc2020102216452002">
</button>
</p>
<div id="acc2020102216452002" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

module load bioinfo-tools
module load HISAT2/2.1.0
module load samtools/1.8

# get output filename prefix
prefix=$( basename $1 | sed -E &#39;s/_.+$//&#39; )

hisat2 \
-p 1 \
-x ../reference/mouse_chr19_hisat2/mouse_chr19_hisat2 \
--summary-file &quot;${prefix}.summary&quot; \
-1 $1 \
-2 $2 | samtools sort -O BAM &gt; &quot;${prefix}.bam&quot;</code></pre>
</div>
</div>
</div>
<p>In the above script, the two input fasta files are not hard coded, rather we use <code>$1</code> and <code>$2</code> because they will be passed in as arguments.</p>
<p>The output filename prefix is automatically created using this line <code>prefix=$( basename $1 | sed -E 's/_.+$//' )</code> from input filename of <code>$1</code>. For example, a file with path <code>/bla/bla/sample_1.fq.gz</code> will have the directory stripped off using the function <code>basename</code> to get <code>sample_1.fq.gz</code>. This is piped (<code>|</code>) to <code>sed</code> where all text starting from <code>_</code> to end of string (specified by this regular expression <code>_.+$</code> matching <code>_1.fq.gz</code>) is removed and the prefix will be just <code>sample</code>. This approach will work only if your filenames are labelled suitably.</p>
<p>Lastly, the SAM output is piped (<code>|</code>) to the tool samtools for sorting and written in BAM format.</p>
<p>Now we can run the bash script like below while standing in the <code>3_mapping</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/hisat2_align.sh sample_1.fq.gz sample_2.fq.gz</code></pre>
</div>
<p><i class="fas fa-clipboard-list"></i> Similarly run the other samples.</p>
<div class="boxy boxy-clipboard-list boxy-orange">
<p><strong>Optional</strong></p>
<p>Try to create a new bash loop script (<code>hisat2_align_batch.sh</code>) to iterate over all fastq files in the directory and run the mapping using the <code>hisat2_align.sh</code> script. Note that there is a bit of a tricky issue here. You need to use two fastq files (<code>_1</code> and <code>_2</code>) per run rather than one file. There are many ways to do this and here is one.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-optional collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216452132" aria-expanded="false" aria-controls="acc2020102216452132">
</button>
</p>
<div id="acc2020102216452132" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>## find only files for read 1 and extract the sample name
lines=$(find *_1.fq.gz | sed &quot;s/_1.fq.gz//&quot;)

for i in ${lines}
do
  ## use the sample name and add suffix (_1.fq.gz or _2.fq.gz)
  echo &quot;Mapping ${i}_1.fq.gz and ${i}_2.fq.gz ...&quot;
  bash ../scripts/hisat2_align.sh ${i}_1.fq.gz ${i}_2.fq.gz
done</code></pre>
</div>
</div>
</div>
<p>Run the <code>hisat2_align_batch.sh</code> script in the <code>3_mapping</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/hisat2_align_batch.sh</code></pre>
</div>
</div>
<p>At the end of the mapping jobs, you should have the following list of output files for every sample:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 16M Sep 19 12:30 SRR3222409-19.bam
-rw-rw-r-- 1 user gXXXXXXX 604 Sep 19 12:30 SRR3222409-19.summary</code></pre>
</div>
<p>The <code>.bam</code> file contains the alignment of all reads to the reference genome in binary format. BAM files are not human readable directly. To view a BAM file in text format, you can use <code>samtools view</code> functionality.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load samtools/1.8
samtools view SRR3222409-19Aligned.sortedByCoord.out.bam | head</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>SRR3222409.13658290 163 19  3084385 60  95M3S   =   3084404 120 CTTTAAGATAAGTGCCGGTTGCAGCCAGCTGTGAGAGCTGCACTCCCTTCTCTGCTCTAAAGTTCCCTCTTCTCAGAAGGTGGCACCACCCTGAGCTG  DB@D@GCHHHEFHIIG&lt;CHHHIHHIIIIHHHIIIIGHIIIIIFHIGHIHIHIIHIIHIHIIHHHHIIIIIIHFFHHIIIGCCCHHHH1GHHIIHHIII  AS:i:-3 ZS:i:-18    XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:95 YS:i:-6 YT:Z:CP NH:i:1
SRR3222409.13658290 83  19  3084404 60  101M    =   3084385 -120    TGCAGCCAGCTGTGAGAGCTGCACTCCCTTCTCTGCTCTAAAGTTCCCTCTTCTCAGAAGGTGGCACCACCCTGAGCTGCTGGCAGTGAGTCTGTTCCAAG   IIIIHECHHH?IHHHIIIHIHIIIHEHHHCHHHIHIIIHHIHIIIHHHHHHIHEHIIHIIHHIIHHIHHIGHIGIIIIIIIHHIIIHHIHEHCHHG@&lt;&lt;BD   AS:i:-6 ZS:i:-16    XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:76T24  YS:i:-3 YT:Z:CP NH:i:1
SRR3222409.13741570 163 19  3085066 60  15M2I84M    =   3085166 201 ATAGTACCTGGCAACAAAAAAAAAAAAGCTTTTGGCTAAAGACCAATGTGTTTAAGAGATAAAAAAAGGGGTGCTAATACAGAAGCTGAGGCCTTAGAAGA   0B@DB@HCCH1&lt;&lt;CGECCCGCHHIDD?01&lt;&lt;G1&lt;/&lt;1&lt;FH1F11&lt;1111&lt;&lt;&lt;&lt;11&lt;CGC1&lt;G1&lt;F//DHHI0/01&lt;&lt;1FG11111111&lt;111&lt;1D1&lt;1D1&lt;   AS:i:-20    XN:i:0  XM:i:3  XO:i:1  XG:i:2  NM:i:5  MD:Z:40T19C27T10    YS:i:0  YT:Z:CP NH:i:1</code></pre>
</div>
<p><i class="fas fa-comments"></i> Can you identify what some of these columns are? SAM format description is available <a href="https://samtools.github.io/hts-specs/">here</a>.</p>
<p><i class="fas fa-clipboard-list"></i> The summary file gives a summary of the mapping run. This file is used by MultiQC later to collect mapping statistics. Inspect one of the mapping log files to identify the number of uniquely mapped reads and multi-mapped reads.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216458968" aria-expanded="false" aria-controls="acc2020102216458968">
</button>
</p>
<div id="acc2020102216458968" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cat SRR3222409-19.summary</code></pre>
</div>
</div>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>156992 reads; of these:
  156992 (100.00%) were paired; of these:
    7984 (5.09%) aligned concordantly 0 times
    147178 (93.75%) aligned concordantly exactly 1 time
    1830 (1.17%) aligned concordantly &gt;1 times
    ----
    7984 pairs aligned concordantly 0 times; of these:
      690 (8.64%) aligned discordantly 1 time
    ----
    7294 pairs aligned 0 times concordantly or discordantly; of these:
      14588 mates make up the pairs; of these:
        5987 (41.04%) aligned 0 times
        4085 (28.00%) aligned exactly 1 time
        4516 (30.96%) aligned &gt;1 times
98.09% overall alignment rate</code></pre>
</div>
<p>Next, we need to index these BAM files. Indexing creates <code>.bam.bai</code> files which are required by many downstream programs to quickly and efficiently locate reads anywhere in the BAM file.</p>
<p><i class="fas fa-clipboard-list"></i> Index all BAM files. Write a for-loop to index all BAM files using the command <code>samtools index file.bam</code>.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216452561" aria-expanded="false" aria-controls="acc2020102216452561">
</button>
</p>
<div id="acc2020102216452561" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load samtools/1.8

for i in *.bam
  do
    echo &quot;Indexing $i ...&quot;
    samtools index $i
  done</code></pre>
</div>
</div>
</div>
<p>Now, we should have <code>.bai</code> index files for all BAM files.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 16M Sep 19 12:30 SRR3222409-19.bam
-rw-rw-r-- 1 user gXXXXXXX 43K Sep 19 12:32 SRR3222409-19.bam.bai</code></pre>
</div>
<p><i class="fas fa-exclamation-circle"></i> If you are running short of time or unable to run the mapping, you can copy over results for all samples that have been prepared for you before class. They are available at this location: <code>/sw/courses/ngsintro/rnaseq/main/3_mapping/</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp -r /sw/courses/ngsintro/rnaseq/main/3_mapping/* /proj/g2020009/nobackup/[user]/rnaseq/3_mapping/</code></pre>
</div>
<p><br></p>
</div>
</div>
<div id="qualimap" class="section level2">
<h2><span class="header-section-number">3.4</span> QualiMap</h2>
<p><strong>Post-alignment QC using QualiMap</strong></p>
<p>Some important quality aspects, such as saturation of sequencing depth, read distribution between different genomic features or coverage uniformity along transcripts, can be measured only after mapping reads to the reference genome. One of the tools to perform this post-alignment quality control is QualiMap. QualiMap examines sequencing alignment data in SAM/BAM files according to the features of the mapped reads and provides an overall view of the data that helps to the detect biases in the sequencing and/or mapping of the data and eases decision-making for further analysis.</p>
<p><i class="fas fa-clipboard-list"></i> Read through <a href="http://qualimap.bioinfo.cipf.es/doc_html/intro.html">QualiMap</a> documentation and see if you can figure it out how to run it to assess post-alignment quality on the RNA-seq mapped samples. Here is the RNA-Seq specific tool <a href="http://qualimap.bioinfo.cipf.es/doc_html/analysis.html#rnaseqqc">explanation</a>. The tool is already installed on Uppmax as a module.</p>
<p><i class="fas fa-clipboard-list"></i> Load the QualiMap module version 2.2.1 and create a bash script named <code>qualimap.sh</code> in your <code>scripts</code> directory.</p>
<p>Add the following script to it. Note that we are using the smaller GTF file with chr19 only.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

# load modules
module load bioinfo-tools
module load QualiMap/2.2.1

# get output filename prefix
prefix=$( basename &quot;$1&quot; .bam)

unset DISPLAY

qualimap rnaseq -pe \
  -bam $1 \
  -gtf &quot;../reference/Mus_musculus.GRCm38.99-19.gtf&quot; \
  -outdir &quot;../4_qualimap/${prefix}/&quot; \
  -outfile &quot;$prefix&quot; \
  -outformat &quot;HTML&quot; \
  --java-mem-size=6G &gt;&amp; &quot;${prefix}-qualimap.log&quot;</code></pre>
</div>
<p>The line <code>prefix=$( basename "$1" .bam)</code> is used to remove directory path and <code>.bam</code> from the input filename and create a prefix which will be used to label output. The export <code>DISPLAY=:0</code> forces a ‘headless mode’ on the JAVA application, which would otherwise throw an error about X11 display. If that doesn’t work, one can also try <code>export DISPLAY=:0</code> or <code>export DISPLAY=""</code>. The last part <code>&gt;&amp; "${prefix}-qualimap.log"</code> saves the <strong>standard-out</strong> as a log file.</p>
<p><i class="fas fa-clipboard-list"></i> Create a new bash loop script named <code>qualimap_batch.sh</code> with a bash loop to run the qualimap script over all BAM files. The loop should look like below.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216458017" aria-expanded="false" aria-controls="acc2020102216458017">
</button>
</p>
<div id="acc2020102216458017" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>for i in ../3_mapping/*.bam
do
    echo &quot;Running QualiMap on $i ...&quot;
    bash ../scripts/qualimap.sh $i
done</code></pre>
</div>
</div>
</div>
<p>Run the loop script <code>qualimap_batch.sh</code> in the directory <code>4_qualimap</code>.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216458556" aria-expanded="false" aria-controls="acc2020102216458556">
</button>
</p>
<div id="acc2020102216458556" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/qualimap_batch.sh</code></pre>
</div>
</div>
</div>
<p>Qualimap should have created a directory for every BAM file.</p>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>drwxrwsr-x 5 user gXXXXXXX 4.0K Jan 22 22:53 SRR3222409-19
-rw-rw-r-- 1 user gXXXXXXX  669 Jan 22 22:53 SRR3222409-19-qualimap.log</code></pre>
</div>
<p>Inside every directory, you should see:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>drwxrwsr-x 2 user gXXXXXXX 4.0K Jan 22 22:53 css
drwxrwsr-x 2 user gXXXXXXX 4.0K Jan 22 22:53 images_qualimapReport
-rw-rw-r-- 1 user gXXXXXXX  12K Jan 22 22:53 qualimapReport.html
drwxrwsr-x 2 user gXXXXXXX 4.0K Jan 22 22:53 raw_data_qualimapReport
-rw-rw-r-- 1 user gXXXXXXX 1.2K Jan 22 22:53 rnaseq_qc_results.txt</code></pre>
</div>
<p><i class="fas fa-exclamation-circle"></i> You can download the HTML files locally to your computer if you wish. If you do so, note that you MUST also download the dependency files (ie; css folder and images_qualimapReport folder), otherwise the HTML file may not render correctly.</p>
<p><i class="fas fa-clipboard-list"></i> Inspect the HTML output file and try to make sense of it.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>firefox qualimapReport.html &amp;</code></pre>
</div>
<p><i class="fas fa-exclamation-circle"></i> If you are running out of time or were unable to run QualiMap, you can also copy pre-run QualiMap output from this location: <code>/sw/courses/ngsintro/rnaseq/main/4_qualimap/</code>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp -r /sw/courses/ngsintro/rnaseq/main/4_qualimap/* /proj/g2020009/nobackup/&lt;username&gt;/rnaseq/4_qualimap/</code></pre>
</div>
<p><i class="fas fa-comments"></i> Check the QualiMap report for one sample and discuss if the sample is of good quality. You only need to do this for one file now. We will do a comparison with all samples when using the MultiQC tool.</p>
<p><br></p>
</div>
<div id="featurecounts" class="section level2">
<h2><span class="header-section-number">3.5</span> featureCounts</h2>
<p><strong>Counting mapped reads using featureCounts</strong></p>
<p>After ensuring mapping quality, we can move on to enumerating reads mapping to genomic features of interest. Here we will use <strong>featureCounts</strong>, an ultrafast and accurate read summarization program, that can count mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations.</p>
<p><i class="fas fa-clipboard-list"></i> Read featureCounts <a href="http://bioinf.wehi.edu.au/subread-package/SubreadUsersGuide.pdf">documentation</a> and see if you can figure it out how to use paired-end reads using an unstranded library to count fragments overlapping with exonic regions and summarise over genes.</p>
<p><i class="fas fa-clipboard-list"></i> Load the subread module on Uppmax. Create a bash script named <code>featurecounts.sh</code> in the directory <code>scripts</code>.</p>
<p>We could run featureCounts on each BAM file, produce a text output for each sample and combine the output. But the easier way is to provide a list of all BAM files and featureCounts will combine counts for all samples into one text file.</p>
<p>Below is the script that we will use:</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216450327" aria-expanded="false" aria-controls="acc2020102216450327">
</button>
</p>
<div id="acc2020102216450327" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

# load modules
module load bioinfo-tools
module load subread/2.0.0

featureCounts \
  -a &quot;../reference/Mus_musculus.GRCm38.99.gtf&quot; \
  -o &quot;counts.txt&quot; \
  -F &quot;GTF&quot; \
  -t &quot;exon&quot; \
  -g &quot;gene_id&quot; \
  -p \
  -s 0 \
  -T 1 \
  ../3_mapping/*.bam</code></pre>
</div>
</div>
</div>
<p>In the above script, we indicate the path of the annotation file (<code>-a "../reference/Mus_musculus.GRCm38.99.gtf"</code>), specify the output file name (<code>-o "counts.txt"</code>), specify that that annotation file is in GTF format (<code>-F "GTF"</code>), specify that reads are to be counted over exonic features (<code>-t "exon"</code>) and summarised to the gene level (<code>-g "gene_id"</code>). We also specify that the reads are paired-end (<code>-p</code>), the library is unstranded (<code>-s 0</code>) and the number of threads to use (<code>-T 1</code>).</p>
<p>Run the featurecounts bash script in the directory <code>5_dge</code>. Use <code>pwd</code> to check if you are standing in the correct directory.</p>
<p>You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/5_dge</code></pre>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216450400" aria-expanded="false" aria-controls="acc2020102216450400">
</button>
</p>
<div id="acc2020102216450400" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>bash ../scripts/featurecounts.sh</code></pre>
</div>
</div>
</div>
<p>You should have two output files:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 2.8M Sep 15 11:05 counts.txt
-rw-rw-r-- 1 user gXXXXXXX  658 Sep 15 11:05 counts.txt.summary</code></pre>
</div>
<p><i class="fas fa-comments"></i> Inspect the files and try to make sense of them.</p>
<div class="boxy boxy-primary boxy-exclamation">
<p><strong>Important</strong></p>
<p>For downstream steps, we will NOT use this counts.txt file. Instead we will use <strong>counts_full.txt</strong> from the back-up folder. This contains counts across all chromosomes. This is located here: <code>/sw/courses/ngsintro/rnaseq/main/5_dge/</code>.</p>
<p>
<button class="btn btn-sm btn-primary btn-collapse btn-collapse-normal collapsed" type="button" data-toggle="collapse" data-target="#acc2020102216455370" aria-expanded="false" aria-controls="acc2020102216455370">
</button>
</p>
<div id="acc2020102216455370" class="collapse">
<div class="card card-body">
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/main/5_dge/counts_full.txt /proj/g2020009/nobackup/&lt;username&gt;/rnaseq/5_dge/</code></pre>
</div>
</div>
</div>
</div>
<p><br></p>
</div>
<div id="multiqc" class="section level2">
<h2><span class="header-section-number">3.6</span> MultiQC</h2>
<p><strong>Combined QC report using MultiQC</strong></p>
<p>We will use the tool <strong>MultiQC</strong> to crawl through the output, log files etc from FastQC, HISAT2, QualiMap and featureCounts to create a combined QC report.</p>
<p>Run MultiQC as shown below in the <code>6_multiqc</code> directory. You should be standing here to run this:</p>
<pre><code>/proj/g2020009/nobackup/&lt;username&gt;/rnaseq/6_multiqc</code></pre>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load MultiQC/1.8

multiqc --interactive ../</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="sh"><code>[INFO   ]         multiqc : This is MultiQC v1.8
[INFO   ]         multiqc : Template    : default
[INFO   ]         multiqc : Searching   : /crex/proj/gXXXXXXX/nobackup/&lt;user&gt;/rnaseq
Searching 297 files..  [####################################]  100%          
[INFO   ]        qualimap : Found 6 RNASeq reports
[INFO   ]  feature_counts : Found 6 reports
[INFO   ]         bowtie2 : Found 6 reports
[INFO   ]          fastqc : Found 12 reports
[INFO   ]         multiqc : Compressing plot data
[INFO   ]         multiqc : Report      : multiqc_report.html
[INFO   ]         multiqc : Data        : multiqc_data
[INFO   ]         multiqc : MultiQC complete</code></pre>
</div>
<p>The output should look like below:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>drwxrwsr-x 2 user gXXXXXXX 4.0K Sep  6 22:33 multiqc_data
-rw-rw-r-- 1 user gXXXXXXX 1.3M Sep  6 22:33 multiqc_report.html</code></pre>
</div>
<p><i class="fas fa-comments"></i> Open the MultiQC HTML report using <code>firefox</code> and/or transfer to your computer and inspect the report. You can also download the file locally to your computer.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>firefox multiqc_report.html &amp;</code></pre>
</div>
<p><br></p>
</div>
<div id="deseq2" class="section level2">
<h2><span class="header-section-number">3.7</span> DESeq2</h2>
<p><strong>Differential gene expression using DESeq2</strong></p>
<p>The easiest way to perform differential expression is to use one of the statistical packages, within R environment, that were specifically designed for analyses of read counts arising from RNA-seq, SAGE and similar technologies. Here, we will one of such packages called <strong>DESeq2</strong>. Learning R is beyond the scope of this course so we prepared basic ready to run R scripts to find DE genes between conditions <strong>KO</strong> and <strong>Wt</strong>.</p>
<p>Move to the <code>5_dge</code> directory and load R modules for use.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load R/4.0.0
module load R_packages/4.0.0</code></pre>
</div>
<p>Use <code>pwd</code> to check if you are standing in the correct directory. Copy the following file to the <code>5_dge</code> directory: <code>/sw/courses/ngsintro/rnaseq/main/5_dge/dge.R</code></p>
<p>Make sure you have the <code>counts_full.txt</code>. If not, you can copy this file too: <code>/sw/courses/ngsintro/rnaseq/main/5_dge/counts_full.txt</code></p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/main/5_dge/dge.R .
cp /sw/courses/ngsintro/rnaseq/main/5_dge/counts_full.txt .</code></pre>
</div>
<p>Now, run the R script from the schell in <code>5_dge</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript dge.R</code></pre>
</div>
<p>If you are curious what’s inside dge.R, you are welcome to explore it using a text editor.</p>
<p>This should have produced the following output files:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 282K Jan 22 23:16 counts_vst_full.Rds
-rw-rw-r-- 1 user gXXXXXXX 1.7M Jan 22 23:16 counts_vst_full.txt
-rw-rw-r-- 1 user gXXXXXXX 727K Jan 22 23:16 dge_results_full.Rds
-rw-rw-r-- 1 user gXXXXXXX 1.7M Jan 22 23:16 dge_results_full.txt</code></pre>
</div>
<p>Essentially, we have two outputs: <strong>dge_results_full</strong> and <strong>counts_vst_full</strong>. <strong>dge_results_full</strong> is the list of differentially expressed genes. This is available in human readable tab-delimited .txt file and R readable binary .Rds file. The <strong>counts_vst_full</strong> is variance-stabilised normalised counts, useful for exploratory analyses.</p>
<p><i class="fas fa-clipboard-list"></i> Copy the results text file (<code>dge_results_full.txt</code>) to your computer and inspect the results. What are the columns? How many differentially expressed genes are present after adjusted p-value of 0.05? How many genes are upregulated and how many are down-regulated? How does this change if we set a fold-change cut-off of 1?</p>
<p><i class="fas fa-lightbulb"></i> Open in a spreadsheet editor like Microsoft Excel or LibreOffice Calc.</p>
<p><i class="fas fa-lightbulb"></i> If you do not have the results or were unable to run the DGE step, you can copy these two here which will be required for functional annotation (optional).</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/main/5_dge/dge_results_full.txt .
cp /sw/courses/ngsintro/rnaseq/main/5_dge/dge_results_full.Rds .
cp /sw/courses/ngsintro/rnaseq/main/5_dge/counts_vst_full.txt .
cp /sw/courses/ngsintro/rnaseq/main/5_dge/counts_vst_full.Rds .</code></pre>
</div>
</div>
</div>
<div id="bonus-exercises" class="section level1">
<h1><span class="header-section-number">4</span> Bonus exercises</h1>
<div class="boxy boxy-primary boxy-exclamation">
<p>These exercises are optional and to be run only if you have time and if it interests you.</p>
</div>
<div id="functional-annotation" class="section level2">
<h2><span class="header-section-number">4.1</span> Functional annotation</h2>
<p>In this part of the exercise we will address the question which biological processes are affected in the experiment; in other words we will functionally annotate the results of the analysis of differential gene expression (performed in the main part of the exercise). We will use <strong>Gene Ontology (GO)</strong> and <strong>Reactome</strong> databases.</p>
<p>When performing this type of analysis, one has to keep in mind that the analysis is only as accurate as the annotation available for your organism. So, if working with non-model organisms which do have experimentally-validated annotations (computationally inferred), the results may not be fully reflecting the actual situation.</p>
<p>There are many methods to approach the question as to which biological processes and pathways are over-represented amongst the differentially expressed genes, compared to all the genes included in the DE analysis. They use several types of statistical tests (e.g. hypergeometric test, Fisher’s exact test etc.), and many have been developed with microarray data in mind. Not all of these methods are appropriate for RNA-seq data, which as you remember from the lecture, exhibit length bias in power of detection of differentially expressed genes (i.e. longer genes, which tend to gather more reads, are more likely to be detected as <strong>differentially expressed</strong> than shorter genes, solely because of the length).</p>
<p>We will use the R / Bioconductor package <strong>goseq</strong>, specifically designed to work with RNA-seq data. This package provides methods for performing Gene Ontology and pathway analysis of RNA-seq data, taking length bias into account.</p>
<p>In this part, we will use the same data as in the main workflow. The starting point of the exercise is the file with results of the differential expression produced in the main part of the exercise.</p>
<div id="preparation" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Preparation</h3>
<p>Copy this file <code>/sw/courses/ngsintro/rnaseq/bonus/funannot/annotate_de_results.R</code> to your <code>rnaseq/funannot</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/bonus/funannot/annotate_de_results.R /proj/g2020009/nobackup/[user]/rnaseq/funannot/</code></pre>
</div>
<p>We need some annotation information, so we will copy this file <code>/sw/courses/ngsintro/rnaseq/main/reference/mm-biomart99-genes.txt.gz</code> to your <code>rnaseq/reference</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/main/reference/mm-biomart99-genes.txt.gz /proj/g2020009/nobackup/[user]/rnaseq/reference/</code></pre>
</div>
<p>Then uncompress the file. <code>gunzip mm-biomart99-genes.txt.gz</code></p>
<p>Load R module and R packages</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load R/4.0.0
module load R_packages/4.0.0</code></pre>
</div>
<p>Change to the <code>funannot</code> directory in your <code>rnaseq</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cd funannot</code></pre>
</div>
<p>Run the functional annotation script from the linux terminal.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript annotate_de_results.R</code></pre>
</div>
<p>It will fetch <strong>dge_results_full.Rds</strong> from <code>5_dge/</code> and annotation files from <code>reference/</code>. When completed, you should find a directory named <strong>funannot_results</strong>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 44K Jan 22 23:37 go_down.txt
-rw-rw-r-- 1 user gXXXXXXX 15K Jan 22 23:37 go_up.txt
-rw-rw-r-- 1 user gXXXXXXX 800 Jan 22 23:37 reactome_down.txt
-rw-rw-r-- 1 user gXXXXXXX 794 Jan 22 23:37 reactome_up.txt</code></pre>
</div>
</div>
<div id="interpretation" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Interpretation</h3>
<p>The results are saved as tables in the directory <strong>funannot_results</strong>. There are four tables: GO terms for up-regulated genes (go_up.txt), GO terms for down-regulated genes (go_down.txt) and similarily, Reactome pathways for up-regulated genes (reactome_up.txt) and Reactome pathways for down-regulated genes (reactome_down.txt).</p>
<p><i class="fas fa-clipboard-list"></i> Take a quick look at some of these files.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>head go_down.txt</code></pre>
</div>
<p>The columns of the results tables are:</p>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code># go
category over_represented_pvalue under_represented_pvalue numDEInCat numInCat term ontology
# reactome
category over_represented_pvalue under_represented_pvalue numDEInCat numInCat path_name</code></pre>
</div>
<p>You can view the tables in a text editor (<code>nano</code>,<code>gedit</code> etc), and try to find GO terms and pathways relevant to the experiment using a word search functionality. You could download these files to your computer and import them into a spreadsheet program like MS Excel or LibreOffice Calc.</p>
<p><i class="fas fa-clipboard-list"></i> Try to use <code>grep</code> to find a match using a keyword, say <strong>phosphorylation</strong>.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cat go_down.txt | grep &quot;phosphorylation&quot;</code></pre>
</div>
<p><i class="fas fa-comments"></i> Have a look at the GO terms and see if you think the functional annotation reflects the biology of the experiments we have just analysed?</p>
</div>
</div>
<div id="rna-seq-plots" class="section level2">
<h2><span class="header-section-number">4.2</span> RNA-Seq plots</h2>
<p>Creating high quality plots of RNA-seq analysis are most easily done using R. Depending on your proficiency in reading R code and using R, you can in this section either just call scripts from the command lines with a set of arguments or you can open the R script in a text editor, and run the code step by step from an interactive R session.</p>
<p>Copy the R script files from the following directory: <code>/sw/courses/ngsintro/rnaseq/bonus/plots/</code> to your <code>plots</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>cp /sw/courses/ngsintro/rnaseq/bonus/plots/*.R /proj/g2020009/nobackup/&lt;username&gt;/rnaseq/plots/</code></pre>
</div>
<p>You should have the following files:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ls -l</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>-rw-rw-r-- 1 user gXXXXXXX 2.0K Sep 20  2016 gene.R
-rw-rw-r-- 1 user gXXXXXXX  842 Sep 22  2016 heatmap.R
-rw-rw-r-- 1 user gXXXXXXX  282 Sep 22  2016 ma.R
-rw-rw-r-- 1 user gXXXXXXX  340 Sep 22  2016 mds.R
-rw-rw-r-- 1 user gXXXXXXX  669 Sep 22  2016 volcano.R</code></pre>
</div>
<p>It is important that you load module R and R_packages.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load R/4.0.0
module load R_packages/4.0.0</code></pre>
</div>
<div id="pca-plot" class="section level3">
<h3><span class="header-section-number">4.2.1</span> PCA plot</h3>
<p>A popular way to visualise general patterns of gene expression in your data is to produce either PCA (Principal Component Analysis) or MDS (Multi Dimensional Scaling) plots. These methods aim at summarizing the main patterns of expression in the data and display them on a two-dimensional space and still retain as much information as possible. To properly evaluate these kind of results is non-trivial, but in the case of RNA-seq data we often use them to get an idea of the difference in expression between treatments and also to get an idea of the similarity among replicates. If the plots shows clear clusters of samples that corresponds to treatment it is an indication of treatment actually having an effect on gene expression. If the distance between replicates from a single treatment is very large it suggests large variance within the treatment, something that will influence the detection of differentially expressed genes between treatments.</p>
<p>Run the <code>pca.R</code> script like below in the <code>plots</code> folder.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript pca.R</code></pre>
</div>
<p>This generates a file named <strong>pca.png</strong> in the <code>plots</code> folder. To view it, use <code>eog pca.png &amp;</code> or copy it to your local disk.</p>
<p><img src="data/rnaseq/pca.png" width="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p><i class="fas fa-comments"></i> Do samples cluster as expected? Are there any odd or mislabelled samples? Based on these results, would you expect to find a large number of significant DE genes?</p>
</div>
<div id="ma-plot" class="section level3">
<h3><span class="header-section-number">4.2.2</span> MA plot</h3>
<p>An MA-plot plots the mean expression and estimated log-fold-change for all genes in an analysis.</p>
<p>Run the <code>ma.R</code> script in the <code>plots</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript ma.R</code></pre>
</div>
<p>This generates a file named <strong>ma.png</strong> in the <code>plots</code> folder. To view it, use <code>eog ma.png &amp;</code> or copy it to your local disk.</p>
<p><img src="data/rnaseq/ma.png" width="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p><i class="fas fa-comments"></i> What do you think the blue dots represent?</p>
</div>
<div id="volcano-plot" class="section level3">
<h3><span class="header-section-number">4.2.3</span> Volcano plot</h3>
<p>A related type of figure will instead plot fold change (on log2 scale) on the x-axis and -log10 p-value on the y-axis. Scaling like this means that genes with lowest p-value will be found at the top of the plot. In this example we will highlight (in red) the genes that are significant (p-val 0.05 after correction for multiple testing).</p>
<p>Run the script named <code>volcano.R</code> in the <code>plots</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript volcano.R</code></pre>
</div>
<p>This generates a file named <strong>volcano.png</strong> in the <code>plots</code> folder. To view it, use <code>eog volcano.png &amp;</code> or copy it to your local disk.</p>
<p><img src="data/rnaseq/volcano.png" width="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p><i class="fas fa-comments"></i> Anything noteworthy about the patterns in the plot? Is there a general trend in the direction of change in gene expression as a consequence of the experiment?</p>
</div>
<div id="heatmap" class="section level3">
<h3><span class="header-section-number">4.2.4</span> Heatmap</h3>
<p>Another popular plots for genome-wide expression patterns is heatmap for a set of genes. If you run the script called <code>heatmap.R</code>, it will extract the top 50 genes that have the lowest p-value in the experiment and create a heatmap from these. In addition to color-coding the expression levels over samples for the genes it also clusters the samples and genes based on inferred distance between them.</p>
<p>Run the script named <code>heatmap.R</code> in the <code>plots</code> directory.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>Rscript heatmap.R</code></pre>
</div>
<p>This generates a file named <strong>heatmap.png</strong> in the <code>plots</code> folder. To view it, use <code>eog heatmap.png &amp;</code> or copy it to your local disk.</p>
<p><img src="data/rnaseq/heatmap.png" width="500px" style="display: block; margin: auto auto auto 0;" /></p>
<p><i class="fas fa-clipboard-list"></i> Compare this plot to a similar plot in the paper behind the data.</p>
</div>
</div>
<div id="igv-browser" class="section level2">
<h2><span class="header-section-number">4.3</span> IGV browser</h2>
<p>Data visualisation is important to be able to clearly convey results, but can also be very helpful as tool for identifying issues and note-worthy patterns in the data. In this part you will use the BAM files you created earlier in the RNA-seq lab and use <a href="http://software.broadinstitute.org/software/igv/">IGV</a> (Integrated Genomic Viewer) to visualize the mapped reads and genome annotations. In addition we will produce high quality plots of both the mapped read data and the results from differential gene expression.</p>
<p>If you are already familiar with IGV you can load the mouse genome and at least one BAM file from each of the treatments that you created earlier. The functionality of IGV is the same as if you look at genomic data, but there are a few of the features that are more interesting to use for RNA-seq data.</p>
<p>Integrated genomics viewer from Broad Institute is a nice graphical interface to view bam files and genome annotations. It also has tools to export data and some functionality to look at splicing patterns in RNA-seq data sets. Even though it allows for some basic types of analysis it should be used more as a nice way to look at your mapped data. Looking at data in this way might seem like a daunting approach as you can not check more than a few regions, but in in many cases it can reveal mapping patterns that are hard to catch with just summary statistics.</p>
<p>For this tutorial you can chose to run IGV directly on your own computer <i class="fas fa-desktop"></i> or on Uppmax <i class="fas fa-cloud"></i> . If you chose to run it on your own computer you will have to download some of the BAM files (and the corresponding index files) from Uppmax. If you have not yet installed IGV you also have to <a href="http://software.broadinstitute.org/software/igv/download">download</a> the program.</p>
<div class="boxy boxy-primary boxy-desktop">
<p><strong>Local</strong></p>
<p>Copy two BAM files (one from each experimental group, for example; SRR3222409-19 and SRR3222412-19) and the associated index (<code>.bam.bai</code>) files to your computer by running the below command in a <strong>LOCAL</strong> terminal and <strong>NOT</strong> on Uppmax.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>scp user@rackham.uppmax.uu.se:/proj/g2020009/nobackup/rnaseq/[user]/3_mapping/SRR3222409-19.bam ./
scp user@rackham.uppmax.uu.se:/proj/g2020009/nobackup/rnaseq/[user]/3_mapping/SRR3222409-19.bam.bai ./
scp user@rackham.uppmax.uu.se:/proj/g2020009/nobackup/rnaseq/[user]/3_mapping/SRR3222412-19.bam ./
scp user@rackham.uppmax.uu.se:/proj/g2020009/nobackup/rnaseq/[user]/3_mapping/SRR3222412-19.bam.bai ./</code></pre>
</div>
<p>Alternatively, all users can use an SFTP browser like <a href="https://filezilla-project.org/">Filezilla</a> or <a href="https://cyberduck.io/">Cyberduck</a> for a GUI interface. Windows users can also use the MobaXterm SFTP file browser to drag and drop.</p>
</div>
<div class="boxy boxy-primary boxy-cloud">
<p><strong>Uppmax</strong></p>
<p>For Linux and Mac users, Log in to Uppmax in a way so that the generated graphics are exported via the network to your screen. This will allow any graphical interface that you start on your compute node to be exported to your computer. However, as the graphics are exported over the network, it can be fairly slow in redrawing windows and the experience can be fairly poor.</p>
<p>Login in to Uppmax with X-forwarding enabled:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>ssh -Y username@rackham.uppmax.uu.se
ssh -Y computenode</code></pre>
</div>
<p>An alternative method is to login through <a href="https://rackham-gui.uppmax.uu.se/main/">Rackham-GUI</a>. Once you log into this interface you will have a linux desktop interface in a browser window. This interface is running on the login node, so if you want to do any heavy lifting you need to login to your reserved compute node also here. This is done by opening a terminal in the running linux environment and log on to your compute node as before. NB! If you have no active reservation you have to do that first.</p>
<p>Load necessary modules and start IGV</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load IGV/2.4.2
igv-core</code></pre>
</div>
<p>This should start the IGV so that it is visible on your screen. If not please try to reconnect to Uppmax or consider running IGV locally as that is often the fastest and most convenient solution.</p>
</div>
<p>Once we have the program running, you select the genome that you would like to load. As seen in the image below. Choose <code>Mouse mm10</code>.</p>
<p><img src="data/rnaseq/igv.png" /></p>
<p>Note that if you are working with a genome that are not part of the available genomes in IGV, one can create genome files from within IGV. Please check the manual of IGV for more information on that.</p>
<p>To open your BAM files, go to <code>File &gt; Load from file...</code> and select your BAM file and make sure that you have a <code>.bai</code> index for that BAM file in the same folder. You can repeat this and open multiple BAM files in the same window, which makes it easy to compare samples. Then select <strong>Chr19</strong> since we only have data for that one chromosome.</p>
<p>For every file you open a number of panels are opened that visualize the data in different ways. The first panel named <strong>Coverage</strong> summarises the coverage of base-pairs in the window you have zoomed to. The second panel shows the reads as they are mapped to the genome. If one right click with the mouse on the read panel there many options to group and color reads. The bottom panel named Refseq genes shows the gene models from the annotation.</p>
<p>To see actual reads you have to zoom in until the reads are drawn on screen. If you have a gene of interest you can also use the search box to directly go to that gene.</p>
<p>Have a look at few of the interesting genes on Chr19. Look into gene <strong>Lbx1</strong>. Do you expect this gene to be differentially expressed? To see a gene with large number of reads, see <strong>Scd2</strong>. Here are a few more <strong>Cuedc2</strong>, <strong>Ahnak</strong> and <strong>Scd1</strong>.</p>
<p>For more detailed information on the splice reads you can instead of just looking at the splice panel right click on the read panel and select <strong>Sashimi plots</strong>. This will open a new window showing in an easy readable fashion how reads are spliced in mapping and you will also be able to see that there are differences in between what locations reads are spliced. This hence gives some indication on the isoform usage of the gene.</p>
<p><i class="fas fa-clipboard-list"></i> Are the reads you mapped from a stranded or unstranded library?</p>
<p><i class="fas fa-clipboard-list"></i> One can visualize all genes in a given pathway using the gene list option under <strong>Regions</strong> in the menu. If you need hints for how to proceed, see <a href="http://software.broadinstitute.org/software/igv/gene_list_view">Gene List tutorial</a> at Broad.</p>
</div>
</div>
<div id="sbatch" class="section level1">
<h1><span class="header-section-number">5</span> sbatch</h1>
<div class="boxy boxy-primary boxy-exclamation">
<p>You are not required to run anything practically in this section. This is just to read and understand.</p>
</div>
<p>We have throughout this tutorial written bash scripts and run them from the terminal directly. Remember that we are not running on the login node. We have pre-allocated resources, then logged in to a compute node to run tasks. This is called working <strong>interactively</strong> on Uppmax. This is fine for tutorials and testing purposes. But, if you were to actually work on Uppmax, you would follow a slightly different approach.</p>
<p>The standard workflow on Uppmax is to login to the login node and then submit tasks as jobs to something called a Slurm queue. We haven’t used this option, because it involves waiting for an unpredictable amount of time for your submitted job to execute. In this section, we will take a look at how to modify a standard bash script to work with Slurm job submission.</p>
<p>This is how our standard bash script for mapping looks like:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

# load modules
module load bioinfo-tools
module load HISAT2/2.1.0
module load samtools/1.8

# create output filename prefix
prefix=$( basename &quot;$1&quot; | sed -E &#39;s/_.+$//&#39; )

hisat2 \
-p 8 \
-x ../reference/mouse_chr19_hisat2/mouse_chr19_hisat2 \
--summary-file &quot;${prefix}.summary&quot; \
-1 $1 \
-2 $2 | samtools sort -O BAM &gt; &quot;${prefix}.bam&quot;</code></pre>
</div>
<p>We add <code>SBATCH</code> commands to the above script. The new script looks like this:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#!/bin/bash

#SBATCH -A g2020009
#SBATCH -p core
#SBATCH -n 8
#SBATCH -t 3:00:00
#SBATCH -J hisat2-align

# load modules
module load bioinfo-tools
module load HISAT2/2.1.0
module load samtools/1.8

# create output filename prefix
prefix=$( basename &quot;$1&quot; | sed -E &#39;s/_.+$//&#39; )

hisat2 \
-p 8 \
-x ../reference/mouse_chr19_hisat2/mouse_chr19_hisat2 \
--summary-file &quot;${prefix}-summary.txt&quot; \
-1 $1 \
-2 $2 | samtools sort -O BAM &gt; &quot;${prefix}.bam&quot;</code></pre>
</div>
<p>The <code>SBATCH</code> commands in the above script is specifying the account name to use resources from, the required number of cores, the time required for the job and a job name.</p>
<p><i class="fas fa-lightbulb"></i> If you run this as a normal bash script like this <code>./hisat2_align.sh ...</code> or <code>bash ./hisat2_align.sh ...</code>, the <code>SBATCH</code> comments have no effect (they are treated as comments) and the contents of the script will immediately start executing. But if you run this as script as <code>sbatch ./hisat2_align.sh ...</code>, the script is submitted as a job to the Uppmax Slurm queue. In this case, the <code>SBATCH</code> lines are interpreted and used by Slurm. At some point, your submitted job will reach the top of the queue and then the script will start to be executed.</p>
<p>You can check your jobs in the queue by running the following command.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>jobinfo -u user</code></pre>
</div>
<p>And this gives a list like this:</p>
<div class="block-title-parent">
<div class="block-title small">
output
</div>
<pre class="output"><code>CLUSTER: rackham
Running jobs:
   JOBID PARTITION   NAME USER        ACCOUNT ST          START_TIME  TIME_LEFT  NODES CPUS NODELIST(REASON)
 5006225      core (null) user       gXXXXXXX  R 2018-09-12T14:00:03      44:31      1    1 r169
 5006229      core (null) user       gXXXXXXX  R 2018-09-12T14:00:03      44:31      1    1 r169
 5006352      core (null) user       gXXXXXXX  R 2018-09-12T14:04:14      48:42      1    1 r178
 5006355      core (null) user       gXXXXXXX  R 2018-09-12T14:05:17      49:45      1    1 r169
 5006356      core (null) user       gXXXXXXX  R 2018-09-12T14:06:08      50:36      1    5 r179</code></pre>
</div>
<p>If the job is pending, then you will see <code>PD</code> in the <code>ST</code> column. If your job is running, you should see <code>R</code>. Once your job starts running, you will see a file named <code>slurm-XXXX.out</code> in the directory in which you submitted the job. This is the <strong>standard-out</strong> from that job. ie; everything that you would normally see printed to your screen when running locally, is printed to this file when running as a job. Once the job is over, one would inspect the slurm output file.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>head slurm-XXXX.out
tail slurm-XXXX.out
cat slurm-XXXX.out</code></pre>
</div>
</div>
<div id="simons-analysis" class="section level1">
<h1><span class="header-section-number">6</span> Simon’s analysis</h1>
<p>We had <a href="https://www.babraham.ac.uk/science-services/bioinformatics/simon-andrews">Simon Andrews</a> from the Babraham institute with us on one of the workshops and he did his own exploration into the dataset from this paper using his tool <a href="http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/">SeqMonk</a>. It’s an excellent read and highly recommended. His report is available <a href="data/rnaseq/simon-report/validating_knockout.html">here</a>.</p>
</div>
<div id="conclusion" class="section level1">
<h1><span class="header-section-number">7</span> Conclusion</h1>
<p>We hope that you enjoyed getting your hands wet working on some real-ish data. In this tutorial, we have covered the most important data processing steps that may be enough when the libraries are good. If not, there are plenty of troubleshooting procedures to try before discarding the data. And once the count table are in place, the biostatistics and data mining begins. There are no well-defined solutions here, all depends on the experiment and questions to be asked, but we strongly advise learning R. Not only to use the specifically designed statistical packages to analyze NGS count data, but also to be able to handle the data and results as well as to generate high-quality plots. There are many available tools and well-written tutorials with examples to learn from.</p>
<p>For those interested in RNA-Seq analysis, SciLifeLab offers a more advanced course in RNA-Seq analysis each semester. For more information, see <a href="https://www.scilifelab.se/events/">SciLifeLab events</a>.</p>
<hr />
<!-- --------------------- Do not edit this and below ---------------------- -->
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
