<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Malin Larsson/Diana Ekman">

<title>Variant Calling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-2b897086d7323cf94b4a601523d3bf70.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-039e2557270f3ab889f6a68f319ad311.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<style>

      .quarto-title-block .quarto-title-banner {
        background-image: url(../../assets/images/banner.webp);
background-size: cover;
      }
</style>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
<link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">
<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>


<meta property="og:title" content="Variant Calling">
<meta property="og:description" content="From reads to short variants">
<meta property="og:image" content="https://nbisweden.github.io/workshop-ngsintro/topics/vc/assets/1_onesamplevc.png">
<meta property="og:image:height" content="538">
<meta property="og:image:width" content="331">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbis-scilifelab.png" alt="logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_contents.html"> 
<span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_precourse.html"> 
<span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_syllabus.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_faq.html"> 
<span class="menu-text">FAQ</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../home_info.html"> 
<span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-ngsintro/"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Variant Calling</h1>
            <p class="subtitle lead">From reads to short variants</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Malin Larsson/Diana Ekman </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">21-Nov-2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
  <li><a href="#Genomicregion" id="toc-Genomicregion" class="nav-link" data-scroll-target="#Genomicregion">Genomic region</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  </ul></li>
  <li><a href="#preparations" id="toc-preparations" class="nav-link" data-scroll-target="#preparations">Preparations</a>
  <ul>
  <li><a href="#hpc-cluster" id="toc-hpc-cluster" class="nav-link" data-scroll-target="#hpc-cluster">HPC cluster</a>
  <ul class="collapse">
  <li><a href="#connect-to-pdc" id="toc-connect-to-pdc" class="nav-link" data-scroll-target="#connect-to-pdc">Connect to PDC</a></li>
  <li><a href="#workspace-on-dardel" id="toc-workspace-on-dardel" class="nav-link" data-scroll-target="#workspace-on-dardel">Workspace on Dardel</a></li>
  <li><a href="#symbolic-links-to-data" id="toc-symbolic-links-to-data" class="nav-link" data-scroll-target="#symbolic-links-to-data">Symbolic links to data</a></li>
  </ul></li>
  <li><a href="#login-to-a-node" id="toc-login-to-a-node" class="nav-link" data-scroll-target="#login-to-a-node">Login to a node</a>
  <ul class="collapse">
  <li><a href="#accessing-programs" id="toc-accessing-programs" class="nav-link" data-scroll-target="#accessing-programs">Accessing programs</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#part1-variant-calling-in-one-sample" id="toc-part1-variant-calling-in-one-sample" class="nav-link" data-scroll-target="#part1-variant-calling-in-one-sample"><span class="header-section-number">1</span> Part1: Variant calling in one sample</a>
  <ul>
  <li><a href="#bwamem" id="toc-bwamem" class="nav-link" data-scroll-target="#bwamem"><span class="header-section-number">1.1</span> Aligning to genome</a>
  <ul class="collapse">
  <li><a href="#index-the-genome" id="toc-index-the-genome" class="nav-link" data-scroll-target="#index-the-genome">Index the genome</a></li>
  <li><a href="#mapping-reads" id="toc-mapping-reads" class="nav-link" data-scroll-target="#mapping-reads">Mapping reads</a>
  <ul class="collapse">
  <li><a href="#check-sam-and-bam-files" id="toc-check-sam-and-bam-files" class="nav-link" data-scroll-target="#check-sam-and-bam-files">Check sam and bam files</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#question-1a" id="toc-question-1a" class="nav-link" data-scroll-target="#question-1a">Question 1A</a></li>
  <li><a href="#variant-calling" id="toc-variant-calling" class="nav-link" data-scroll-target="#variant-calling"><span class="header-section-number">1.2</span> Variant Calling</a>
  <ul class="collapse">
  <li><a href="#haplotypecaller" id="toc-haplotypecaller" class="nav-link" data-scroll-target="#haplotypecaller"><span class="header-section-number">1.2.1</span> HaplotypeCaller</a></li>
  <li><a href="#explore-the-vcf-file" id="toc-explore-the-vcf-file" class="nav-link" data-scroll-target="#explore-the-vcf-file"><span class="header-section-number">1.2.2</span> Explore the vcf file</a></li>
  </ul></li>
  <li><a href="#questions-1b" id="toc-questions-1b" class="nav-link" data-scroll-target="#questions-1b">Questions 1B</a></li>
  <li><a href="#igv-visualization" id="toc-igv-visualization" class="nav-link" data-scroll-target="#igv-visualization"><span class="header-section-number">1.3</span> IGV visualization</a>
  <ul class="collapse">
  <li><a href="#copy-files-to-local-workspace" id="toc-copy-files-to-local-workspace" class="nav-link" data-scroll-target="#copy-files-to-local-workspace"><span class="header-section-number">1.3.1</span> Copy files to local workspace</a>
  <ul class="collapse">
  <li><a href="#download-the-bam-file" id="toc-download-the-bam-file" class="nav-link" data-scroll-target="#download-the-bam-file"><span class="header-section-number">1.3.1.1</span> Download the bam file</a></li>
  </ul></li>
  <li><a href="#bam-files-in-igv" id="toc-bam-files-in-igv" class="nav-link" data-scroll-target="#bam-files-in-igv"><span class="header-section-number">1.3.2</span> Bam files in IGV</a></li>
  <li><a href="#vcfinigv" id="toc-vcfinigv" class="nav-link" data-scroll-target="#vcfinigv"><span class="header-section-number">1.3.3</span> Vcf files in IGV</a></li>
  </ul></li>
  <li><a href="#questions-1c" id="toc-questions-1c" class="nav-link" data-scroll-target="#questions-1c">Questions 1C</a></li>
  </ul></li>
  <li><a href="#jointvc" id="toc-jointvc" class="nav-link" data-scroll-target="#jointvc"><span class="header-section-number">2</span> Part2: Variant calling in cohort</a>
  <ul>
  <li><a href="#mapping---cohort" id="toc-mapping---cohort" class="nav-link" data-scroll-target="#mapping---cohort"><span class="header-section-number">2.1</span> Mapping - cohort</a>
  <ul class="collapse">
  <li><a href="#bwa_joint" id="toc-bwa_joint" class="nav-link" data-scroll-target="#bwa_joint"><span class="header-section-number">2.1.1</span> Alignment</a></li>
  <li><a href="#mark-duplicates" id="toc-mark-duplicates" class="nav-link" data-scroll-target="#mark-duplicates"><span class="header-section-number">2.1.2</span> Mark Duplicates</a></li>
  </ul></li>
  <li><a href="#generategvcf" id="toc-generategvcf" class="nav-link" data-scroll-target="#generategvcf"><span class="header-section-number">2.2</span> Variant calling - cohort</a>
  <ul class="collapse">
  <li><a href="#generate-g.vcf-files" id="toc-generate-g.vcf-files" class="nav-link" data-scroll-target="#generate-g.vcf-files"><span class="header-section-number">2.2.1</span> Generate g.vcf files</a></li>
  <li><a href="#jointgenotyping" id="toc-jointgenotyping" class="nav-link" data-scroll-target="#jointgenotyping"><span class="header-section-number">2.2.2</span> Joint genotyping</a></li>
  </ul></li>
  <li><a href="#questions-2a" id="toc-questions-2a" class="nav-link" data-scroll-target="#questions-2a">Questions 2A</a></li>
  <li><a href="#variant-filtration" id="toc-variant-filtration" class="nav-link" data-scroll-target="#variant-filtration"><span class="header-section-number">2.3</span> Variant filtration</a></li>
  <li><a href="#questions-2b" id="toc-questions-2b" class="nav-link" data-scroll-target="#questions-2b">Questions 2B</a></li>
  <li><a href="#igv-visualization-of-cohort-vcf" id="toc-igv-visualization-of-cohort-vcf" class="nav-link" data-scroll-target="#igv-visualization-of-cohort-vcf"><span class="header-section-number">2.4</span> IGV visualization of cohort vcf</a></li>
  <li><a href="#questions-2c" id="toc-questions-2c" class="nav-link" data-scroll-target="#questions-2c">Questions 2C</a></li>
  </ul></li>
  <li><a href="#part3-sbatch-scripts" id="toc-part3-sbatch-scripts" class="nav-link" data-scroll-target="#part3-sbatch-scripts"><span class="header-section-number">3</span> Part3: SBATCH scripts</a>
  <ul>
  <li><a href="#variant-calling-in-cohort" id="toc-variant-calling-in-cohort" class="nav-link" data-scroll-target="#variant-calling-in-cohort">Variant calling in cohort</a></li>
  </ul></li>
  <li><a href="#extra-material-gatks-best-practices" id="toc-extra-material-gatks-best-practices" class="nav-link" data-scroll-target="#extra-material-gatks-best-practices"><span class="header-section-number">4</span> Extra material: GATK’s best practices</a>
  <ul>
  <li><a href="#mapping---gatk-best-practise" id="toc-mapping---gatk-best-practise" class="nav-link" data-scroll-target="#mapping---gatk-best-practise"><span class="header-section-number">4.1</span> Mapping - GATK Best practise</a>
  <ul class="collapse">
  <li><a href="#bwa-mem" id="toc-bwa-mem" class="nav-link" data-scroll-target="#bwa-mem">BWA mem</a></li>
  <li><a href="#recalibrate-base-quality-scores" id="toc-recalibrate-base-quality-scores" class="nav-link" data-scroll-target="#recalibrate-base-quality-scores">Recalibrate Base Quality Scores</a></li>
  </ul></li>
  <li><a href="#variant-calling---gatk-best-practise" id="toc-variant-calling---gatk-best-practise" class="nav-link" data-scroll-target="#variant-calling---gatk-best-practise"><span class="header-section-number">4.2</span> Variant calling - GATK Best Practise</a>
  <ul class="collapse">
  <li><a href="#generate-g.vcf-files-1" id="toc-generate-g.vcf-files-1" class="nav-link" data-scroll-target="#generate-g.vcf-files-1">Generate g.vcf files</a></li>
  <li><a href="#joint-genotyping" id="toc-joint-genotyping" class="nav-link" data-scroll-target="#joint-genotyping">Joint genotyping</a></li>
  </ul></li>
  <li><a href="#variant-filtration-1" id="toc-variant-filtration-1" class="nav-link" data-scroll-target="#variant-filtration-1">Variant Filtration</a>
  <ul class="collapse">
  <li><a href="#hard-filters" id="toc-hard-filters" class="nav-link" data-scroll-target="#hard-filters"><span class="header-section-number">4.2.1</span> Hard filters</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#when-finished" id="toc-when-finished" class="nav-link" data-scroll-target="#when-finished"><span class="header-section-number">5</span> When finished</a>
  <ul>
  <li><a href="#clean-up" id="toc-clean-up" class="nav-link" data-scroll-target="#clean-up">Clean up</a></li>
  <li><a href="#answers" id="toc-answers" class="nav-link" data-scroll-target="#answers">Answers</a></li>
  </ul></li>
  <li><a href="#additional-information" id="toc-additional-information" class="nav-link" data-scroll-target="#additional-information">Additional information</a>
  <ul>
  <li><a href="#precomputed-files" id="toc-precomputed-files" class="nav-link" data-scroll-target="#precomputed-files">Precomputed files</a></li>
  </ul></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1 unnumbered">
<h1 class="unnumbered">Introduction</h1>
<p>Whole genome sequencing (WGS) is a comprehensive method for analyzing entire genomes. This workshop will take you through the process of calling germline short variants (SNVs and INDELs) in WGS data from three human samples.</p>
<ol type="1">
<li>The first part of the workshop will guide you through a basic variant calling workflow in one sample. The goals are that you should get familiar with the bam and vcf file formats, and be able to interpret vcf files in Integrative Genomics Viewer (IGV).</li>
<li>The second part of the workshop will show you how to refine the mapping and perform joint variant calling in three samples. The goals here are that you should be able to interpret multi-sample vcf files and explain the differences between the g.vcf and vcf file formats.</li>
<li>If you have time, the third part will let you practise running tools in a bash script. The goal is that you should be familiar with running commands using a script.</li>
<li>Extra material: will take you through the GATK best practices for germline short variant detection. The goal here is that you should learn how to use the GATK documentation so that you can analyze your own samples in the future.</li>
</ol>
<section id="Genomicregion" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="Genomicregion">Genomic region</h2>
<p>The LCT gene on chromosome 2 encodes the enzyme lactase, which is responsible for the metabolism of lactose in mammals. The variant rs4988235, located at position chr2:136608646 in the GRCh37 reference genome, has been shown to lead to lactose persistence.</p>
<p>In this workshop we will detect genetic variants in the region chr2:136545000-136617000 in three samples, and check if they carry the allele for lactase persistence.</p>
<details>
<summary>
Details about LCT and lactase persistence
</summary>
<p>The LCT gene on chromosome 2 encodes the enzyme lactase, which is responsible for the metabolism of lactose in mammals. Most mammals can not digest lactose as adults, but some humans can. Genetic variants upstream of the LCT gene cause lactase persistence, which means that lactase is expressed also in adulthood and the carrier can continue to digest lactose. The variant rs4988235, located at position chr2:136608646 in the GRCh37 reference genome, has been shown to lead to lactose persistence. The alternative allele (A on the forward strand and T on the reverse strand) creates a new transcription factor binding site that enables continued expression of the gene after weaning.</p>
<p>For those interested in the details of the genetic bases for lactose tolerance, please read the first three pages of <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3401057/pdf/ceg-5-113.pdf">Lactose intolerance: diagnosis, genetic, and clinical factors</a> by Mattar et al. The variant rs4988235 is here referred to as LCT-13910C&gt;T. You can also read about the variant in <a href="https://www.omim.org/entry/601806#0001">OMIM</a>.</p>
</details>
</section>
<section id="data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="data">Data</h2>
<p>We will use low coverage whole genome sequence data from three British individuals (HG00097, HG00100 and HG00101), generated in the first phase of the 1000 Genomes Project.</p>
<p>All input data for this exercise is located in this folder:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
General advice
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>If you change the node you are working on on the compute cluster you will need to reload the tool modules</li>
<li>Please type commands in the terminal instead of copying and pasting them which often result in formatting errors</li>
<li>Use tab completion</li>
<li>In paths, please replace <code>username</code> with your actual PDC username.</li>
<li>In commands, please replace <code>parameter</code> with the correct parameter, for example your input and output file names</li>
<li>A line starting with <code>#</code> is a comment and should not be run</li>
<li>Running a command without parameters will often return a help message</li>
<li>After a command is completed, please check that the desired output file was generated and that it has a reasonable size (use <code>ls -l</code>)</li>
<li>Search for errors, someone in the world has run into EXACTLY the same problem and asked about it in a forum somewhere</li>
</ul>
</div>
</div>
</section>
</section>
<section id="preparations" class="level1 unnumbered">
<h1 class="unnumbered">Preparations</h1>
<section id="hpc-cluster" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="hpc-cluster">HPC cluster</h2>
<section id="connect-to-pdc" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="connect-to-pdc">Connect to PDC</h3>
<p>Please follow the instructions for connecting to PDC at <a href="../other/lab_connect_pdc.html"><strong>Connecting to PDC</strong></a> under section 1.2 Connecting to PDC using SSH.</p>
</section>
<section id="workspace-on-dardel" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="workspace-on-dardel">Workspace on Dardel</h3>
<p>You should work in a subfolder of the course folder on Dardel, like you have done during the previous labs. Start by going to your course folder using this command (replace <code>username</code> with your actual username):</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /cfs/klemming/projects/supr/naiss2025-22-1240/username</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Create a folder for this exercise and move into it:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> vc</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> vc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Make sure you are located in this folder for the rest of this lab (use pwd to check).</p>
<pre><code>/cfs/klemming/projects/supr/naiss2025-22-1240/username/vc</code></pre>
</section>
<section id="symbolic-links-to-data" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="symbolic-links-to-data">Symbolic links to data</h3>
<p>The raw data files are located in the course data folder. Create a symbolic link in your workspace to the reference fasta file and the fastq files:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/ref/human_g1k_v37_chr2.fasta</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00097_1.fq</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00097_2.fq</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00100_1.fq</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00100_2.fq</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00101_1.fq</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/fastq/HG00101_2.fq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
<section id="login-to-a-node" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="login-to-a-node">Login to a node</h2>
<p>This lab should be done on a compute node (not the login node). First check if you already have an active allocation using this command, where <em>username</em> should be replaced with your username</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">squeue</span> <span class="at">-u</span> username</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If no jobs are listed you should allocate one for this lab. If you already have an active job allocation please proceed to <strong>Connect to the node</strong> below.<br></p>
<p>Use this code to allocate a job (note that the reservation changes each day):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Wednesday</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">salloc</span> <span class="at">-A</span> naiss2025-22-1240 <span class="at">--reservation</span><span class="op">=</span>edu25-11-19 <span class="at">-t</span> 04:00:00 <span class="at">-p</span> shared <span class="at">-c</span> 8 <span class="at">--no-shell</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Thursday</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="ex">salloc</span> <span class="at">-A</span> naiss2025-22-1240 <span class="at">--reservation</span><span class="op">=</span>edu25-11-20 <span class="at">-t</span> 04:00:00 <span class="at">-p</span> shared <span class="at">-c</span> 8 <span class="at">--no-shell</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once your job allocation has been granted (should not take long) please check the allocation again:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">squeue</span> <span class="at">-u</span> username</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You should now see that you have an active job allocation. The node name for your job is listed under the node list header. Replace <em>nodename</em> with that name.<br></p>
<p><strong>Connect to the node and move to your folder:</strong></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="at">-Y</span> nodename</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /cfs/klemming/projects/supr/naiss2025-22-1240/username/vc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="accessing-programs" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="accessing-programs">Accessing programs</h3>
<p>Load the tool modules that are needed during this workshop. Remember that these modules must be loaded <strong>every time you login to Dardel,</strong> or when you connect to a new compute node. Load the bioinfo-tools module first to make it possible to load the other modules:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bwa/0.7.17</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load samtools/1.20</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load gatk/4.5.0.0</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load vcftools/0.1.16</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#It is recommended to specify the tool version for reproducibility, although not required</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</section>
<section id="part1-variant-calling-in-one-sample" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Part1: Variant calling in one sample</h1>
<p>Now let’s start the first part of the exercise, which is variant calling in one sample. The workflow consists of aligning the reads with <a href="http://bio-bwa.sourceforge.net">BWA</a> and detecting variants with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller">HaplotypeCaller</a> as illustrated below.</p>
<p><a href="assets/1_onesamplevc.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="assets/1_onesamplevc.png" class="img-fluid"></a></p>
<section id="bwamem" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="bwamem"><span class="header-section-number">1.1</span> Aligning to genome</h2>
<section id="index-the-genome" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="index-the-genome">Index the genome</h3>
<p>Many tools need an index of the reference genome to work efficiently. You therefore need to create index files for each tool first.<br></p>
<p>Generate a BWA index, fasta index and a sequence dictionary:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> index <span class="at">-a</span> bwtsw human_g1k_v37_chr2.fasta</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx human_g1k_v37_chr2.fasta</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> CreateSequenceDictionary <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="at">-O</span> human_g1k_v37_chr2.dict</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Check what files were created using <code>ls -lrt</code>.</p>
</section>
<section id="mapping-reads" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="mapping-reads">Mapping reads</h3>
<p>You will use <code>BWA mem</code> to align the reads to the reference genome. You need to provide the reference fasta file and two fastq files (for paired reads).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem human_g1k_v37_chr2.fasta HG00097_1.fq HG00097_2.fq <span class="op">&gt;</span> HG00097.sam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Sort the aligned reads by chromosome position and convert the sam file to bam format.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> sort HG00097.sam <span class="op">&gt;</span> HG00097.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Next, you need to index the generated bam file so that programs can access the sorted data without reading the whole file.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index HG00097.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Please check what output files have been generated using <code>ls -lrt</code>. Is there a size difference between the sam and the bam file?</p>
<section id="check-sam-and-bam-files" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="check-sam-and-bam-files">Check sam and bam files</h4>
<p>You can look at the sam file using “less”. The bam file, on the other hand, is binary so we need to use <code>samtools view</code>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> HG00097.sam</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View the first 5 lines of the bam file</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view HG00097.bam <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">#The header section of the bam file can be viewed separately with the -H flag:</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-H</span> HG00097.bam </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</section>
<section id="question-1a" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="question-1a">Question 1A</h2>
<ol type="1">
<li>The <code>@SQ</code> tag of the bam header contains information about the reference sequence. What do you think SN:2 and LN:243199373 in this tag means?<br>
</li>
<li>What is the leftmost mapping position of the first read in the bamfile?</li>
</ol>
<p>For more information about the format, please have a look at the <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">Sequence Alignment/Map Format Specification</a>.</p>
</section>
<section id="variant-calling" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="variant-calling"><span class="header-section-number">1.2</span> Variant Calling</h2>
<section id="haplotypecaller" class="level3" data-number="1.2.1">
<h3 data-number="1.2.1" class="anchored" data-anchor-id="haplotypecaller"><span class="header-section-number">1.2.1</span> HaplotypeCaller</h3>
<p>We will detect short variants in the bam file using GATK’s HaplotypeCaller. First you need to add something called a <em>read group</em>, which contains information about the samples and sequencing. Read groups are required by HaplotypeCaller. We will use toy information here, but the RGSM tag should contain the sample name, and the RGID should be unique for each sample and sequencing run. To learn more about read groups please read <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups">this article</a>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> AddOrReplaceReadGroups <span class="at">-I</span> HG00097.bam <span class="at">-O</span> HG00097.rg.bam <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>--RGID rg_HG00097 <span class="at">--RGSM</span> HG00097 <span class="at">--RGPL</span> illumina <span class="at">--RGLB</span> libx <span class="at">--RGPU</span> XXX</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Index the resulting bam file</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index HG00097.rg.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now you can run the variant calling:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> HaplotypeCaller <span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>-R human_g1k_v37_chr2.fasta <span class="at">-I</span> HG00097.rg.bam <span class="at">-O</span> HG00097.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Check what new files were generated using <code>ls -lrt</code>.</p>
</section>
<section id="explore-the-vcf-file" class="level3" data-number="1.2.2">
<h3 data-number="1.2.2" class="anchored" data-anchor-id="explore-the-vcf-file"><span class="header-section-number">1.2.2</span> Explore the vcf file</h3>
<p>Now you have your first vcf file containing the variants on chr2:136545000-136617000. Please look at the vcf file with <code>less</code> and try to understand its structure.</p>
<p>Vcf files contain meta-information lines starting with <em>##</em>, a header line starting with <em>#CHROM</em>, and then data lines each containing information about one variant position in the genome. The header line defines the columns of the data lines. The meta-information lines starting with <em>##INFO</em> define how the data in the <em>INFO</em> column is encoded, and the meta-information lines starting with <em>##FORMAT</em> define how the data in the <em>FORMAT</em> column is encoded.</p>
<p>To view the different parts of the header you can try these commands:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'#CHROM'</span> HG00097.vcf</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'##INFO'</span> HG00097.vcf</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'##FORMAT'</span> HG00097.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To see only the variants you can use</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">"#"</span> HG00097.vcf <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="questions-1b" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="questions-1b">Questions 1B</h2>
<ol type="1">
<li>What column of the VCF file contains genotype information for the sample HG00097?<br>
</li>
<li>What does <em>GT</em> in the <em>FORMAT</em> column of the data lines mean?<br>
</li>
<li>What does <em>AD</em> in the <em>FORMAT</em> column of the data lines mean?<br>
</li>
<li>What genotype does the sample HG00097 have at position 2:136545844? (Use “grep” to find it)<br>
</li>
<li>What are the allelic depths for the reference and alternative alleles in sample HG00097 at position 2:136545844?</li>
</ol>
<p>For more detailed information about vcf files please have a look at <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">The Variant Call Format specification</a>.</p>
</section>
<section id="igv-visualization" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="igv-visualization"><span class="header-section-number">1.3</span> IGV visualization</h2>
<p>Integrated Genomics Viewer (IGV) provides an interactive visualization of the reads in a bam file as well as variants in a vcf file. Here we will show you how to use IGV on your laptop.</p>
<details>
<summary>
Install IGV or use ThinLinc
</summary>
<p>(Precourse preparations) <em>If you have not used IGV on your laptop before, then go to the IGV <a href="https://software.broadinstitute.org/software/igv/download">download page</a>, and follow the instructions to download it. It will prompt you to fill in some information and agree to license. Launch the viewer through web start.</em></p>
<p><em>If you have problems with IGV locally, you can connect to Dardel via ThinLinc, however the number of licences for ThinLinc on Dardel is limited, so during the workshop it might not work for everyone.</em></p>
</details>
<section id="copy-files-to-local-workspace" class="level3" data-number="1.3.1">
<h3 data-number="1.3.1" class="anchored" data-anchor-id="copy-files-to-local-workspace"><span class="header-section-number">1.3.1</span> Copy files to local workspace</h3>
<p>If you use IGV on your laptop, you need to copy some files from Dardel to your laptop. First, create a local workspace on your laptop, for example a folder called <em>vc_local</em> on your desktop. You need to have write permission in this folder. The folder you created will be referred to as <em>local workspace</em> in this lab.</p>
<section id="download-the-bam-file" class="level4" data-number="1.3.1.1">
<h4 data-number="1.3.1.1" class="anchored" data-anchor-id="download-the-bam-file"><span class="header-section-number">1.3.1.1</span> Download the bam file</h4>
<p>You need to download the bam file and the corresponding .bai file to your laptop. Navigate to your <a href="#preparelaptop">local workspace</a> (do not log in to Dardel). Copy the .bam and .bam.bai files you just generated with this command (replace <code>username</code> with your username):</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> username@dardel.pdc.kth.se:/cfs/klemming/projects/supr/naiss2025-22-1240/username/vc/HG00097.bam<span class="pp">*</span> .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
If scp does not work
</summary>
<p>Some users have problems with the wildcard “*”, try to write “HG00097.bam*”</p>
</details>
<p>Check that the files are now present in your <em>local workspace</em> using <code>ls -lrt</code>.</p>
</section>
</section>
<section id="bam-files-in-igv" class="level3" data-number="1.3.2">
<h3 data-number="1.3.2" class="anchored" data-anchor-id="bam-files-in-igv"><span class="header-section-number">1.3.2</span> Bam files in IGV</h3>
<p>Open IGV. In the upper left dropdown menu choose <code>Human hg19</code> (which is the same as GRCh37).<br>
In the <code>File</code> menu, select <code>Load from File</code>, navigate to your workspace, and select <code>HG00097.bam</code>. This data should now appear as a track in the tracks window.</p>
<p>You need to zoom in to see the reads. You can either select a region by click and drag, or by typing a region or a gene name in the text box at the top. We only have data for the region chr2:136545000-136617000, which covers the <em>LTC</em> gene, so you can type this region or the gene name in the search box .</p>
</section>
<section id="vcfinigv" class="level3" data-number="1.3.3">
<h3 data-number="1.3.3" class="anchored" data-anchor-id="vcfinigv"><span class="header-section-number">1.3.3</span> Vcf files in IGV</h3>
<p>Download the file HG00097.vcf to your local workspace as described for the bam file above.<br>
Load the file HG00097.vcf into IGV as you did with the HG00097.bam file. You will now see all the variants called in HG00097. You can view variants in the <em>LCT</em> gene by typing the gene name in the search box, and you can look specifically at the variant at position chr2:136545844 by typing that position in the search box.</p>
</section>
</section>
<section id="questions-1c" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="questions-1c">Questions 1C</h2>
<ol type="1">
<li>What is the read length?<br>
</li>
<li>Approximately how many reads cover an arbitrary position in the genomic region we are looking at?<br>
</li>
<li>Which genes are located within the region chr2:136545000-136617000?<br>
</li>
<li>Hover the mouse over (or click) the upper row of the vcf track. What is the reference and alternative alleles of the variant at position chr2:136545844?<br>
</li>
<li>Hover the mouse over the lower row of the vcf track and look under “Genotype Information”. What genotype does HG00097 have at position chr2:136545844? Is this the same as you found by looking directly in the vcf file in question 10?<br>
</li>
<li>Look at the bam track and count the number of reads that have “G” and “C”, respectively, at position chr2:136545844. How is this information captured under “Genotype Attributes”? (Again, hover the mouse over the lower row of the vcf track.)</li>
</ol>
</section>
</section>
<section id="jointvc" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Part2: Variant calling in cohort</h1>
<p>You will now try joint variant calling in three samples. Each sample will again be processed with <code>BWA mem</code> and <code>HaplotypeCaller</code>. We will however add a few steps and modify the variant calling to generate one g.vcf file per sample. The individual g.vcf files are combined with GATK’s <code>CombineGVCFs</code>, and translated into vcf format with GATK’s <code>GenotypeGVCFs</code>. The workflow below shows how the three samples are processed and combined.</p>
<p><a href="assets/2_jointvc.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="assets/2_jointvc.png" class="img-fluid"></a></p>
<section id="mapping---cohort" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="mapping---cohort"><span class="header-section-number">2.1</span> Mapping - cohort</h2>
<section id="bwa_joint" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="bwa_joint"><span class="header-section-number">2.1.1</span> Alignment</h3>
<p>We will run bwa a bit differently this time: mapping, sorting and conversion to bam will be performed in a single line. You will also add the <em>read group</em> information already in this step.</p>
<p>In addition, you can specify how many <em>threads</em> the program should use, which should be the same as the number of cores that you have access to. As we have asked for 8 cores we can specify <code>-t 8</code> below.</p>
<p>Remember to replace <em>sample</em> with the sample name, both in the read group and in file names.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem <span class="at">-t</span> 8 <span class="dt">\</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>-R <span class="st">"@RG\tID:rg_sample\tSM:sample\tPU:XXX\tLB:libx\tPL:illumina"</span> <span class="dt">\</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>human_g1k_v37_chr2.fasta sample_1.fq sample_2.fq <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="op">&gt;</span> sample.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
Exact command for the first sample
</summary>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem <span class="at">-t</span> 8 <span class="at">-R</span> <span class="st">"@RG\tID:rg_HG00097\tSM:HG00097\tPU:XXX\tLB:libx\tPL:illumina"</span> human_g1k_v37_chr2.fasta HG00097_1.fq HG00097_2.fq <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="op">&gt;</span> HG00097.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<p>Please check what output files were generated this time. What could the advantage be of running it in one line?</p>
</section>
<section id="mark-duplicates" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="mark-duplicates"><span class="header-section-number">2.1.2</span> Mark Duplicates</h3>
<p>Sometimes the same DNA fragment is sequenced multiple times, which leads to multiple reads from the same fragment in the fastq files. If a duplicated read contains a genetic variant, the ratio of the two alleles might be obscured, which can lead to incorrect genotyping. It is therefore recommended (in most cases) to mark duplicate reads so that they are counted as one during genotyping. Please read about Picard’s <code>MarkDuplicates</code> <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225972-MarkDuplicates-Picard-">here</a>.</p>
<p>Run MarkDuplicates on all three bam files generated in the previous step. <br></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> MarkDuplicates <span class="dt">\</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-I</span> sample.bam <span class="at">-O</span> sample.md.bam <span class="at">-M</span> sample_mdmetrics.txt</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index sample.md.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="generategvcf" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="generategvcf"><span class="header-section-number">2.2</span> Variant calling - cohort</h2>
<section id="generate-g.vcf-files" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="generate-g.vcf-files"><span class="header-section-number">2.2.1</span> Generate g.vcf files</h3>
<p><code>HaplotypeCaller</code> should also be run for all three samples, but this time the output for each sample needs to be in g.vcf format. This is accomplished by adding the <code>-ERC GVCF</code> flag.<br></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> HaplotypeCaller <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>-ERC GVCF <span class="at">-I</span> sample.md.bam <span class="at">-O</span> sample.g.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="jointgenotyping" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="jointgenotyping"><span class="header-section-number">2.2.2</span> Joint genotyping</h3>
<p>Once you have the g.vcf files for all samples you should perform joint genotype calling. To do this you first need to combine all individual .g.vcf files to one file using <code>CombineGVCFs</code>. Remember to replace <em>sample1</em>, <em>sample2</em>, <em>sample3</em> with the real sample names.<br></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> CombineGVCFs <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>-V sample1.g.vcf <span class="at">-V</span> sample2.g.vcf <span class="at">-V</span> sample3.g.vcf <span class="at">-O</span> cohort.g.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then run <code>GenoteypeGVC</code> to generate a vcf file:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> GenotypeGVCFs <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>-V cohort.g.vcf <span class="at">-O</span> cohort.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="questions-2a" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="questions-2a">Questions 2A</h2>
<ol type="1">
<li>How many data lines does the cohort.g.vcf file have? You can use the Linux command <code>grep -v "#" cohort.g.vcf | wc -l</code> to extract and count lines in “cohort.g.vcf” that don’t start with “#”.<br>
</li>
<li>How many data lines does the cohort.vcf file have? Explain the difference in number of lines.<br>
</li>
<li>What is encoded in the last three columns of the data lines?</li>
</ol>
</section>
<section id="variant-filtration" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="variant-filtration"><span class="header-section-number">2.3</span> Variant filtration</h2>
<p>HaplotypeCaller is designed to be very sensitive, which minimizes the risk of missing real variants. However, it means that the number of false positives can be quite large, so we need to filter the raw variant set. There are more sophisticated tools in the GATK suite for this, but we will use a simple filter, based on the quality value.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> VariantFiltration <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-V</span> cohort.vcf <span class="at">-O</span> cohort.filtered.vcf <span class="dt">\</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> QUALfilt <span class="at">--filter-expression</span> <span class="st">"QUAL &lt; 100"</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note that the FILTER column is now filled.</p>
<p><code>vcftools</code> is another tool that can be used to filter and manipulate the vcf file. For example, variants that did not pass the filters can be removed, and we can add other filters, for example mean sequencing depth.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--vcf</span> cohort.filtered.vcf <span class="at">--remove-filtered-all</span> <span class="at">--min-meanDP</span> 5 <span class="at">--recode</span> <span class="at">--recode-INFO-all</span> <span class="at">--out</span> cohort.PASS</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>See the vcftools <a href="https://vcftools.sourceforge.net/man_latest.html">manual</a> for more options.</p>
</section>
<section id="questions-2b" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="questions-2b">Questions 2B</h2>
<ol type="1">
<li>Check how many variants in total that are present in the cohort.filtered.vcf file and how many that have passed the filters. Is the difference big?<br>
</li>
<li>Look at the variants that did not pass the filters using <code>grep -v 'PASS' cohort.filtered.vcf</code>. (Do you understand why these variants didn’t pass the filter?)<br>
</li>
<li>How many variants were removed by vcftools (in cohort.PASS.recode.vcf)?<br>
</li>
<li>Try to remove all indels using vcftools. How many variants remain?</li>
</ol>
</section>
<section id="igv-visualization-of-cohort-vcf" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="igv-visualization-of-cohort-vcf"><span class="header-section-number">2.4</span> IGV visualization of cohort vcf</h2>
<p>Again, you have to download the files to your local workspace to look at them in IGV. This time you will view <strong>cohort.filtered.vcf, HG00097.md.bam, HG00100.md.bam and HG00101.md.bam and the corresponding index files</strong>.<br>
</p>
<p>Load the files into IGV as described earlier. This time let us look closer at the lactase persistence related variant, rs4988235, located at position chr2:136608646. Please use IGV to answer the questions below.</p>
</section>
<section id="questions-2c" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="questions-2c">Questions 2C</h2>
<ol type="1">
<li>What is the reference and alternative alleles at chr2:136608646?<br>
</li>
<li>What genotypes do the three samples have at chr2:136608646? Note how genotypes are color coded in IGV.<br>
</li>
<li>Should any of the individuals avoid drinking milk (i.e.&nbsp;homozygous reference)?<br>
</li>
<li>Now compare the data shown in IGV with the data in the VCF file. Extract the row for the chr2:136608646 variant in the cohort.vcf file, for example using <code>grep '136608646' cohort.vcf</code>. What columns of the vcf file contain the information shown in the upper part of the vcf track in IGV?<br>
</li>
<li>What columns of the vcf file contain the information shown in the lower part of the vcf track?<br>
</li>
<li>Zoom out so that you can see the <em>MCM6</em> and <em>LCT</em> genes. Is the variant at chr2:136608646 located within the LCT gene?</li>
</ol>
<p>When you have finished the exercises, you can have a look at this document with <a href="lab_vc_answers.pdf">answers to all questions</a>, and compare them with your answers.</p>
</section>
</section>
<section id="part3-sbatch-scripts" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Part3: SBATCH scripts</h1>
<p>This section is optional and intended for those who want to learn how to run all steps automatically in a bash script. Please make sure that you have understood all the individual steps before you start with this. To learn more about SLURM and SBATCH scripts please look at the <a href="https://www.pdc.kth.se/support/documents/run_jobs/job_scheduling.html">How to Run Jobs</a> page on the PDC website.</p>
<section id="variant-calling-in-cohort" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="variant-calling-in-cohort">Variant calling in cohort</h2>
<p>Below is a skeleton script that can be used as a template for running <a href="#jointvc">variant calling in a cohort</a>. Modify it to run all the steps in part two of this workshop.</p>
<div class="cell" data-accordion="true">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb29"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -A naiss2025-22-1240 </span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -p shared</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -c 8</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -t 1:00:00</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -J JointVariantCalling</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bwa/0.7.17</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load samtools/1.20</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load gatk/4.5.0.0</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a><span class="co">## loop through the samples:</span></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> HG00097 HG00100 HG00101<span class="kw">;</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"Now analyzing: "</span><span class="va">${sample}</span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Fill in the code for running bwa-mem for each sample here</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Fill in the code for samtools index for each sample here</span></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Fill in the code for MarkDuplicates here</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Fill in the code for HaplotypeCaller for each sample here</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Fill in the code for CombineGVCFs for all samples here</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a><span class="co">#Fill in the code for GenotypeGVCFs here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Save the template in your workspace and call it “joint_genotyping.sbatch” . Make the script executable by this command:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chmod</span> u+x joint_genotyping.sbatch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To run the sbatch script in the SLURM queue, use this command:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> joint_genotyping.sbatch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this lab you have an active node reservation so you can also run the script as a normal bash script:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">./joint_genotyping.sbatch</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If you would like more help with creating the sbatch script, please look at our example solution below (click on the link):</p>
<details>
<summary>
Example solution
</summary>
<div class="cell" data-accordion="true">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb33"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -A naiss2025-22-1240</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -p shared</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -c 8</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -t 2:00:00</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -J JointVariantCalling</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load bwa/0.7.17</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load samtools/1.20</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load gatk/4.5.0.0</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> HG00097 HG00100 HG00101<span class="kw">;</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"Now analyzing: "</span><span class="va">${sample}</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bwa</span> mem <span class="at">-R</span> <span class="dt">\</span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>  <span class="st">"@RG\tID:</span><span class="va">${sample}</span><span class="st">\tPU:flowcellx_lanex\tSM:</span><span class="va">${sample}</span><span class="st">\tLB:libraryx\tPL:illumina"</span> <span class="dt">\</span></span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">-t</span> 8 human_g1k_v37_chr2.fasta <span class="st">"</span><span class="va">${sample}</span><span class="st">_1.fq"</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">_2.fq"</span> <span class="kw">|</span> <span class="ex">samtools</span> sort <span class="op">&gt;</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">.bam"</span></span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">samtools</span> index <span class="st">"</span><span class="va">${sample}</span><span class="st">.bam"</span></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> MarkDuplicates <span class="dt">\</span></span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">-I</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">.bam"</span> <span class="dt">\</span></span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">.md.bam"</span> <span class="dt">\</span></span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">-M</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">_mdmetrics.txt"</span></span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a>  <span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> HaplotypeCaller <span class="dt">\</span></span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">-ERC</span> GVCF <span class="at">-I</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">.bam"</span> <span class="dt">\</span></span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">.g.vcf"</span></span>
<span id="cb33-33"><a href="#cb33-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-34"><a href="#cb33-34" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb33-35"><a href="#cb33-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-36"><a href="#cb33-36" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> CombineGVCFs <span class="dt">\</span></span>
<span id="cb33-37"><a href="#cb33-37" aria-hidden="true" tabindex="-1"></a>-R human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb33-38"><a href="#cb33-38" aria-hidden="true" tabindex="-1"></a>-V HG00097.g.vcf <span class="dt">\</span></span>
<span id="cb33-39"><a href="#cb33-39" aria-hidden="true" tabindex="-1"></a>-V HG00100.g.vcf <span class="dt">\</span></span>
<span id="cb33-40"><a href="#cb33-40" aria-hidden="true" tabindex="-1"></a>-V HG00101.g.vcf <span class="dt">\</span></span>
<span id="cb33-41"><a href="#cb33-41" aria-hidden="true" tabindex="-1"></a>-O cohort.g.vcf</span>
<span id="cb33-42"><a href="#cb33-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-43"><a href="#cb33-43" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> GenotypeGVCFs <span class="dt">\</span></span>
<span id="cb33-44"><a href="#cb33-44" aria-hidden="true" tabindex="-1"></a>-R human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb33-45"><a href="#cb33-45" aria-hidden="true" tabindex="-1"></a>-V cohort.g.vcf <span class="dt">\</span></span>
<span id="cb33-46"><a href="#cb33-46" aria-hidden="true" tabindex="-1"></a>-O cohort.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>
</section>
</section>
<section id="extra-material-gatks-best-practices" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Extra material: GATK’s best practices</h1>
<p>If you receive your sequencing data with variants called, it is likely that your data has been processed according to the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-">GATKs best practices for germline short variant discovery</a>. GATK is often used for human sequencing data, but also for many other organisms, especially model organisms. If you are interested you can try this optional section, where the variant calling is refined further.</p>
<details>
<summary>
GATK best practises germline variant calling
</summary>
<p>This will be included in the lecture on Thursday morning. If you reach this step earlier you can have a look at this <a href="https://youtu.be/b7VV6e5q6ss">prerecorded video of the same lecture</a>.</p>
<p><a href="assets/3_best_practise.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="assets/3_best_practise.png" class="img-fluid" style="width:80.0%"></a></p>
<section id="mapping---gatk-best-practise" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="mapping---gatk-best-practise"><span class="header-section-number">4.1</span> Mapping - GATK Best practise</h2>
<section id="bwa-mem" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="bwa-mem">BWA mem</h3>
<p>The first step in GATK’s best practice variant calling workflow is to run BWA mem and MarkDuplicates for each sample exactly as you did in <a href="#bwa_joint">Variant calling in cohort</a>. You have already done this step, so please use the bam files that you generated before (sample.md.bam).</p>
</section>
<section id="recalibrate-base-quality-scores" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="recalibrate-base-quality-scores">Recalibrate Base Quality Scores</h3>
<p>A potential source of error is systematic biases in the assignment of base quality scores by the sequencing instrument. This can be corrected by GATK’s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890531-Base-Quality-Score-Recalibration-BQSR-">Base Quality Score Recalibration</a>. In short, you first use <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593511-BaseRecalibrator">BaseRecalibrator</a> to build a recalibration model, and then <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225212-ApplyBQSR">ApplyBQSR</a> to recalibrate the base qualities in your bam file.<br>
<code>BaseRecalibrator</code> requires a file with known SNPs as input. This file is available in the data folder on Dardel. Make a soft link to this file:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /sw/courses/ngsintro/vc/data/ref/1000G_phase1.snps.high_confidence.b37.chr2.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>First, run BaseRecalibrator for the bam file generated in the previous step:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> BaseRecalibrator <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">--known-sites</span> 1000G_phase1.snps.high_confidence.b37.chr2.vcf <span class="dt">\</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">-I</span> HG00097.md.bam <span class="at">-O</span> HG00097.recal.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then run ApplyBQSR:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> ApplyBQSR <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-bqsr-recal-file</span> HG00097.recal.table <span class="dt">\</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-I</span> HG00097.md.bam <span class="at">-O</span> HG00097.recal.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Repeat for all three samples</p>
</section>
</section>
<section id="variant-calling---gatk-best-practise" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="variant-calling---gatk-best-practise"><span class="header-section-number">4.2</span> Variant calling - GATK Best Practise</h2>
<section id="generate-g.vcf-files-1" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="generate-g.vcf-files-1">Generate g.vcf files</h3>
<p>Next, <code>HaplotypeCaller</code> should be run for all three samples, and the output should be in g.vcf, as described <a href="#generategvcf">above</a>. This time use the recalibrated bam files as input.</p>
</section>
<section id="joint-genotyping" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="joint-genotyping">Joint genotyping</h3>
<p>Once you have the g.vcf files for all samples you should perform joint genotype calling. This is done with <code>CombineGVCFs</code> and <code>GenotypeGVCFs</code> as described <a href="#jointgenotyping">above</a>, but you should use the g.vcf files generated from the recalibrated bam files as input.</p>
</section>
</section>
<section id="variant-filtration-1" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="variant-filtration-1">Variant Filtration</h2>
<p>GATK offers two ways to filter variants:</p>
<ol type="1">
<li><p>The variant quality score recalibration (VQSR) method uses machine learning to identify variants that are likely to be real. This is the best method if you have a lot of data, for example one whole genome sequence sample or several whole exome samples, and work with human, mouse or any other well-studied organism. Read more here <a href="https://gatk.broadinstitute.org/hc/en-us/articles/4402736812443-Which-training-sets-arguments-should-I-use-for-running-VQSR">GATK VQSR</a></p></li>
<li><p>If you have less data, or a non-model organism, you can use hard filters as described <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531112?id=2806#2">here</a>.</p></li>
</ol>
<p>Since we have very little data we will use hard filters. The parameters are slightly different for SNVs and INDELs, so you need to first select all SNVs using <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225432-SelectVariants">SelectVariants</a> and filter them using <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037226192-VariantFiltration">VariantFiltration</a> with the parameters suggested for SNVs, and then the same for indels. Finally merge the SNVs and INDELs to get all variants in one file. The <strong>cutoffs are suggested starting values</strong>, but should be adjusted to the dataset. Read about hard filters <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants">here</a>.</p>
<section id="hard-filters" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="hard-filters"><span class="header-section-number">4.2.1</span> Hard filters</h3>
<p>Example solution for filtering SNVs:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> SelectVariants <span class="dt">\</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--select-type-to-include</span> SNP <span class="dt">\</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-V</span> cohort.vcf <span class="at">-O</span> cohort.snvs.vcf</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> VariantFiltration <span class="dt">\</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">-V</span> cohort.snvs.vcf <span class="at">-O</span> cohort.snvs.filtered.vcf <span class="dt">\</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> QDfilter <span class="at">--filter-expression</span> <span class="st">"QD &lt; 2.0"</span>  <span class="dt">\</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> MQfilter <span class="at">--filter-expression</span> <span class="st">"MQ &lt; 40.0"</span>  <span class="dt">\</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> FSfilter <span class="at">--filter-expression</span> <span class="st">"FS &gt; 60.0"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Example filters for INDELs:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> SelectVariants <span class="dt">\</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-V</span> cohort.vcf <span class="at">-O</span> cohort.indels.vcf <span class="dt">\</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>--select-type-to-include INDEL <span class="dt">\</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> VariantFiltration <span class="dt">\</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">-R</span> human_g1k_v37_chr2.fasta <span class="dt">\</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">-V</span> cohort.indels.vcf <span class="at">-O</span> cohort.indels.filtered.vcf <span class="dt">\</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> QDfilter <span class="at">--filter-expression</span> <span class="st">"QD &lt; 2.0"</span> <span class="dt">\</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filter-name</span> FSfilter <span class="at">--filter-expression</span> <span class="st">"FS &gt; 200.0"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Merge filtered SNPs and INDELs:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gatk</span> <span class="at">--java-options</span> <span class="at">-Xmx4g</span> MergeVcfs <span class="dt">\</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-I</span> cohort.snvs.filtered.vcf <span class="at">-I</span> cohort.indels.filtered.vcf <span class="at">-O</span> cohort.filtered.vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Look at your filtered vcf with e.g.&nbsp;<code>less</code>. It still has all the variant lines, but the FILTER column is now filled in, with PASS or the filters it failed. Note also that the filters that were run are described in the header section.</p>
</section></section></details>
</section>


<section id="when-finished" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> When finished</h1>
<section id="clean-up" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="clean-up">Clean up</h2>
<p>When the analysis is done and you are sure that you have the desired output, it is a good practice to remove intermediary files that are no longer needed. This will save disk space, and will be a crucial part of the routines when you work with your own data. Please think about which files you need to keep if you would like to go back and look at this lab later on. Remove the other files.</p>
</section>
<section id="answers" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="answers">Answers</h2>
<p>When you have finished the exercise, please have a look at this document with <a href="lab_vc_answers.pdf">answers to all questions</a>, and compare them with your answers.</p>
</section>
</section>
<section id="additional-information" class="level1 unnumbered">
<h1 class="unnumbered">Additional information</h1>
<section id="precomputed-files" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="precomputed-files">Precomputed files</h2>
<p>If you run out of time you can click below to get paths to precomputed bam files.</p>
<details>
<summary>
Precomputed files
</summary>
<div class="cell" data-accordion="true">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb40"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data/bam/HG00097.bam</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data/bam/HG00100.bam</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data/bam/HG00101.bam</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>If you run out of time you can click below to get paths to the precomputed cohort.g.vcf and cohort.vcf files.</p>
<div class="cell" data-accordion="true">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb41"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data/vcf/cohort.g.vcf</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="ex">/sw/courses/ngsintro/vc/data/vcf/cohort.vcf</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>If you run out of time, please click below to get the path to precomputed bam and vcf files for the GATK’s best practices section.</p>
<div class="cell" data-accordion="true">
<div class="cell-output cell-output-stdout">
<div class="sourceCode" id="cb42"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Path</span> to intermediary and final bam files: /sw/courses/ngsintro/vc/data/best_practise_bam</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Path</span> to intermediary and final vcf files: /sw/courses/ngsintro/vc/data/best_practise_vcf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>
</section>
</section>
<section id="section" class="level1 unnumbered">
<h1 class="unnumbered"></h1>
<ul>
<li>Here is a technical documentation of <a href="https://www.illumina.com/documents/products/technotes/technote_understanding_quality_scores.pdf">Illumina Quality Scores</a></li>
<li>Tools used or referenced
<ul>
<li><a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
<li><a href="http://multiqc.info/">MultiQC</a></li>
<li><a href="https://broadinstitute.github.io/picard/command-line-overview.html">Picard</a></li>
<li><a href="https://software.broadinstitute.org/gatk/">GATK</a></li>
<li><a href="http://www.htslib.org/">samtools</a></li>
</ul></li>
</ul>
<!--# Alternative cluster
The teachers may inform you that we will use the high performance computing center UPPMAX instead of PDC during this workshop. If so, please follow this link to [run the workshop on UPPMAX](https://nbisweden.github.io/workshop-ngsintro/2403/topics/vc/lab_vc.html).-->
<p><strong>Thanks for today!</strong></p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nbisweden\.github\.io\/workshop-ngsintro\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2025 <a href="https://nbis.se">NBIS</a> • <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Published with <a href="https://quarto.org/">Quarto</a> v1.6.42
</p>
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>