<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>National Bioinformatics infrastructure Sweden (NBIS)</title>
    <meta charset="utf-8" />
    <meta name="author" content="11-Mar-2022" />
    <meta name="keywords" content="bioinformatics, course, scilifelab, nbis" />
    <script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
    <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <link rel="stylesheet" href="assets/slide.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# National Bioinformatics infrastructure Sweden (NBIS)
## Introduction To Bioinformatics Using NGS Data
### 11-Mar-2022

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!-- ------------ Only edit title, subtitle &amp; author above this ------------ --&gt;

---
name: intro

**Enable world-class life science research and maximise scientific and societal impact of collected data by:**
- Providing Providing expert knowledge, innovative data integration, 
advanced training, efficient data publication for open science, 
and access to high-performance data analysis methods
- Coordinating bioinformatics support within Sweden and making 
bioinformatics easily accessible for life science researchers
- Swiftly responding to changes in support needs as new 
techniques are developed and utilised
- Forming the Swedish ELIXIR node and participating in relevant 
international projects

[Insert group photo here]
[Insert map here]
[Nbis units]

---
name: services

# Services

## Topics

- Genomics/NGS/Metagenomics
- Genome annotation and assembly
- MS-Proteomics and Protein bioinformatics
- Systems biology and Metabolomics
- Integrative bioinformatics
- Biostatistics 
- Data publication and FAIRification of data
- Data management and Data stewardship 
- Computational structural and chemical biology

## [Support](https://nbis.se/support/)

- Drop-Ins
- Consultation
- Short and medium term support (Fee based)
- Long term support
- Partner projects
- Bioimage informatics

## [Infrastructure](https://nbis.se/infrastructure/)

- Computational resources
- Data management
- Tools &amp; guidelines

## [Training](https://nbis.se/training/)

- ~13 workshops
- Introduction to bioinformatics using NGS data
- Introduction to data management practices
- Introduction to python - with applications in bioinformatics
- Tools for reproducible research
- R foundations for life scientists
- RNA-seq data analysis
- Single-cell RNA-seq data analysis
- Elixir omics integration and systems biology
- Introduction to biostatistics and machine learning
- Workshop in epigenomics data analysis
- Neural nets and deep learning
- Quick and clean advanced python
- Workshop on plotting in R

---
name: support-stats

- 267 PIs (104 female, 163 male)
- 212 active projects
- 104 consultations
- ~13 workshops

---
name: support

---
name: training

---
name: infrastructure

---
name: compute

---

nbis mission
funding sources
part of scilifelab
part of elixir
nbis locations, map
nbis units - support, infrastructure, training

support stats

support
- consultaion
- sms
- long-term
- partner
- bioimage bioinformatics
- focus areas

training
- workshops
- advisory program

infrastructure
- repositories
- hosting

compute and storage
snoc, uppmax, hpc

data management
- data centre
- data stewards, fair data

tools
- maintanence
- development
- pipelines

elixir
map
swedish contributions
- human protein atlas
- human data community
- federated ega and beacons
- data management
- biodiversity
- training

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
