<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.js" defer="true"></script>
<link href="../../site_libs/quarto-contrib/reveal-logo-1.0.0/reveal-logo.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.42">

  <meta name="author" content="Roy Francis / Dag Ahrén">
  <title>Bulk RNASeq Analysis</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-988246922734d86b16c6693a4ff32568.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
  <link href="https://fonts.googleapis.com/css2?family=Fira+Mono&amp;family=Nunito:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&amp;display=swap" rel="stylesheet">
  <link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
  <script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<meta property="og:title" content="Bulk RNASeq Analysis">
<meta property="og:description" content="Analysis of bulk RNA-Seq data to determine genome-wide gene expression">
<meta property="og:image" content="https://nbisweden.github.io/workshop-ngsintro/topics/rnaseq/assets/images/seo.webp">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/images/cover.webp" data-background-opacity="1" data-background-size="cover" class="center">
  <div class="title-container">

        <div class="title-parent title-two">
      <div class="title-child-text">
        <h1 class="title">Bulk RNASeq Analysis</h1>
                        <p class="author">Roy Francis / Dag Ahrén</p>
                                <p class="date">10-Nov-2025</p>
              </div>
      <div class="title-child-hero">
                <div class="title-hero">
          <img class="hero" src="../../assets/images/slide-hero.webp">
        </div>
              </div>
    </div>

    
  </div>
</section>

<section id="central-dogma-of-biology" class="slide level2">
<h2>Central Dogma of Biology</h2>
<p><img data-src="assets/rnaseq_transcription.svg" style="width:70.0%;height:100.0%"></p>
<ul>
<li>The transcriptome is spatially and temporally dynamic</li>
<li>Data comes from functional units (coding regions)</li>
<li>Only a tiny fraction of the genome</li>
</ul>
<aside class="notes">
<ul>
<li>Central dogma of molecular biology. DNA -&gt; RNA -&gt; Protein.</li>
<li>We have gene models to describe the organisation of a genome into functional units.</li>
<li>Very tiny portion of the genome is transcribed into RNA.</li>
<li>There are many types of RNA. Commonly protein coding RNA, also called mRNA. There are tRNA, sRNA, miRNA, siRNA, lincRNA, piRNA, snRNA etc.</li>
<li>What is the most abundant RNA? rRNA.</li>
<li>While DNA is mostly considered constant in all cells for an organism, expressed RNA varies from cell to cell and from time to time.</li>
<li>When we say RNA sequencing, we are not sequencing RNA molecules directly. RNA is first converted to DNA through reverse transcription followed by regular DNA sequencing.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="applications" class="slide level2">
<h2>Applications</h2>
<ul>
<li>Identify gene sequences in genomes (annotation)</li>
<li>Learn about gene function</li>
<li>Differential gene expression</li>
<li>Explore isoform and allelic expression</li>
<li>Understand co-expression, pathways and networks</li>
<li>Gene fusion</li>
<li>RNA editing</li>
</ul>
</section>
<section id="workflow" class="slide level2">
<h2>Workflow</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p><img data-src="assets/rnaseq_workflow.svg"></p>
</div><div class="column" style="width:35%;">
<p><img data-src="assets/sequence.jpg"></p>
</div></div>
<p><span class="citation" data-cites="conesa2016survey">Conesa et al. (<a href="#/references" role="doc-biblioref" onclick="">2016</a>)</span></p>
<aside class="notes">
<p><strong>De-Novo assembly</strong></p>
<ul>
<li>When no reference genome available</li>
<li>To identify novel genes/transcripts/isoforms</li>
<li>Identify fusion genes</li>
<li>Assemble transcriptome from short reads</li>
<li>Access quality of assembly and refine</li>
<li>Map reads back to assembled transcriptome</li>
</ul>
<div class="small">
<p><a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity</a>, <span class="citation" data-cites="hsieh2019effect">Hsieh et al. (<a href="#/references" role="doc-biblioref" onclick="">2019</a>)</span>, <span class="citation" data-cites="wang2017comprehensive">Wang &amp; Gribskov (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></p>
</div>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="experimental-design" class="slide level2">
<h2>Experimental design</h2>
<div class="columns">
<div class="column">
<ul>
<li>Biological replicates: 6 - 12 <span class="citation" data-cites="schurch2016many">Schurch et al. (<a href="#/references" role="doc-biblioref" onclick="">2016</a>)</span></li>
<li>Sample size estimation <span class="citation" data-cites="hart2013calculating">Hart et al. (<a href="#/references" role="doc-biblioref" onclick="">2013</a>)</span></li>
<li>Power analysis <span class="small"><a href="https://rnaseq-power.serve.scilifelab.se/">rnaseq-power web app</a></span>, <span class="citation" data-cites="zhao2018rnaseqsamplesize">Zhao et al. (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span></li>
<li>Balanced design to avoid batch effects</li>
<li>RIN values have strong effect <span class="citation" data-cites="gallego2014rna">Gallego Romero et al. (<a href="#/references" role="doc-biblioref" onclick="">2014</a>)</span></li>
</ul>
</div><div class="column">
<p><img data-src="assets/batch-effect.svg" style="width:75.0%"></p>
</div></div>
<aside class="notes">
<ul>
<li>Technical replicates are not necessary <span class="citation" data-cites="marioni2008rna">Marioni et al. (<a href="#/references" role="doc-biblioref" onclick="">2008</a>)</span></li>
<li>Previous publications can be useful in experimental planning to avoid repeating the same mistakes</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="library-sequencing" class="slide level2">
<h2>Library &amp; Sequencing</h2>
<div class="columns">
<div class="column">
<p><img data-src="assets/rnaseq_library_prep.svg" style="width:85.0%"></p>
</div><div class="column">
<p><img data-src="assets/rnaseq_read_type.svg"></p>
</div></div>
</section>
<section id="library-prep" class="slide level2 notes">
<h2>Library prep</h2>
<ul>
<li>80% of the RNA in a cell is ribosomal RNA (rRNA)</li>
<li>rRNA can be eliminated using polyA selection or rRNA depletion
<ul>
<li>PolyA selection mostly captures only protein-coding genes / mRNA but gives cleaner results</li>
<li>Depletion of rRNA is the solution if it’s important to retain non-mRNA species</li>
</ul></li>
<li>smallRNAs are purified through size selection</li>
<li>PCR amplification may be needed for low quantity input (See section PCR duplicates)</li>
<li>Use stranded (directional) libraries <span class="citation" data-cites="zhao2015comparison">Zhao et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span>, <span class="citation" data-cites="levin2010comprehensive">Levin et al. (<a href="#/references" role="doc-biblioref" onclick="">2010</a>)</span>
<ul>
<li>Accurately identify sense/antisense transcript</li>
<li>Resolve overlapping genes</li>
</ul></li>
<li>Libraries should have high complexity / low duplication. <span class="citation" data-cites="daley2013predicting">Daley &amp; Smith (<a href="#/references" role="doc-biblioref" onclick="">2013</a>)</span></li>
<li>Exome capture</li>
<li>Library normalisation to concentrate specific transcripts</li>
</ul>
</section>
<section id="sequencing" class="slide level2">
<h2>Sequencing</h2>
<ul>
<li>Short reads vs long reads (Illumina vs PacBio/Nanopore)
<ul>
<li>Short reads are cheap and good enough for DGE</li>
<li>Long reads are better for isoforms and structural variation</li>
</ul></li>
<li>Read length <span class="citation" data-cites="chhangawala2015impact">Chhangawala et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span>
<ul>
<li>Greater than 50bp does not improve DGE</li>
<li>Longer reads are better for isoforms</li>
</ul></li>
<li>Pooling samples</li>
<li>Sequencing depth (Coverage / Reads per sample)</li>
<li>Single-end reads (No longer cheaper!)</li>
<li>Use paired-end reads
<ul>
<li>Increased mappable reads</li>
<li>Increased power in assemblies</li>
<li>Better for structural variation and isoforms</li>
<li>Decreased false-positives for DGE</li>
</ul></li>
<li>More replicates are better than more depth <span class="citation" data-cites="liu2014rna">Liu et al. (<a href="#/references" role="doc-biblioref" onclick="">2014</a>)</span></li>
</ul>
<p><span class="citation" data-cites="corley2017differentially">Corley et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></p>
</section>
<section id="workflow-dge" class="slide level2">
<h2>Workflow • DGE</h2>
<p><img data-src="assets/rnaseq_workflow_dge.svg" style="width:70.0%;height:100.0%"></p>
</section>
<section id="read-qc" class="slide level2">
<h2>Read QC</h2>
<div class="columns">
<div class="column" style="width:65%;">
<ul>
<li>Number of reads</li>
<li>Per base sequence quality</li>
<li>Per sequence quality score</li>
<li>Per base sequence content</li>
<li>Per sequence GC content</li>
<li>Per base N content</li>
<li>Sequence length distribution</li>
<li>Sequence duplication levels</li>
<li>Overrepresented sequences</li>
<li>Adapter content</li>
<li>Kmer content</li>
</ul>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, <a href="http://multiqc.info/">MultiQC</a>, <a href="https://sequencing.qcfail.com/" class="uri">https://sequencing.qcfail.com/</a></p>
<p><img data-src="assets/qcfail.jpg" style="width:60.0%"></p>
</div><div class="column" style="width:35%;">
<p><img data-src="assets/qc.jpg"></p>
</div></div>
</section>
<section id="fastqc" class="slide level2">
<h2>FastQC</h2>
<div class="columns">
<div class="column">
<p>Good quality</p>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html" width="50%"><img data-src="assets/fastqc_good.png"></a></p>
</div><div class="column">
<p>Poor quality</p>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html" width="50%"><img data-src="assets/fastqc_bad.png"></a></p>
</div></div>
</section>
<section id="read-qc-pbsq-psqs" class="slide level2">
<h2>Read QC • PBSQ, PSQS</h2>
<p><strong>Per base sequence quality</strong><br>
<img data-src="assets/pbsq.jpg" style="width:50.0%"></p>
<p><strong>Per sequence quality scores</strong><br>
<img data-src="assets/psqs.jpg" style="width:50.0%"></p>
</section>
<aside class="notes">
<section id="read-qc-pbsc-psgc" class="slide level2">
<h2>Read QC • PBSC, PSGC</h2>
<p><strong>Per base sequence content</strong><br>
<img data-src="assets/pbsc.jpg" style="width:50.0%"></p>
<p><strong>Per sequence GC content</strong><br>
<img data-src="assets/psgc.jpg" style="width:50.0%"></p>
</section>
<section id="read-qc-sdl-ac" class="slide level2">
<h2>Read QC • SDL, AC</h2>
<p><strong>Sequence duplication level</strong><br>
<img data-src="assets/sdl.jpg" style="width:50.0%"></p>
<p><strong>Adapter content</strong><br>
<img data-src="assets/ac.jpg" style="width:50.0%"></p>
</section>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<section id="trimming" class="slide level2">
<h2>Trimming</h2>
<div class="columns">
<div class="column">
<ul>
<li>Trimming reads to remove adapter/readthrough or low quality bases</li>
<li>Related options are hard clipping, filtering reads</li>
<li>Sliding window trimming</li>
<li>Filter by min/max read length
<ul>
<li>Remove reads less than ~18nt</li>
</ul></li>
<li>Demultiplexing/Splitting</li>
</ul>
<p><strong>When to avoid trimming?</strong></p>
<ul>
<li>Read trimming may not always be necessary <span class="citation" data-cites="liao2020read">Liao &amp; Shi (<a href="#/references" role="doc-biblioref" onclick="">2020</a>)</span></li>
<li>Fixed read length may sometimes be more important</li>
<li>Expected insert size distribution may be more important for assemblers</li>
</ul>
<p><a href="https://github.com/marcelm/cutadapt/">Cutadapt</a>, <a href="https://github.com/OpenGene/fastp">fastp</a>, <a href="http://prinseq.sourceforge.net/">Prinseq</a>, <a href="https://github.com/relipmoc/skewer">skewer</a>, <a href="https://github.com/usadellab/Trimmomatic">Trimmomatic</a>, <a href="https://github.com/FelixKrueger/TrimGalore">TrimGalore</a></p>
</div><div class="column">
<p><img data-src="assets/rnaseq_read_through.svg" style="width:65.0%"></p>
</div></div>
</section>
<section id="mapping" class="slide level2">
<h2>Mapping</h2>
<p><img data-src="assets/rnaseq_mapping.svg" style="width:70.0%;height:100.0%"></p>
<ul>
<li>Aligning reads back to a reference sequence</li>
<li>Mapping to genome vs transcriptome</li>
<li>Splice-aware alignment (genome) (STAR, HISAT2 etc)</li>
</ul>
<p><a href="https://github.com/alexdobin/STAR">STAR</a>, <a href="https://ccb.jhu.edu/software/hisat2/index.shtml">HiSat2</a>, <span class="citation" data-cites="baruzzo2017simulation">Baruzzo et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></p>
</section>
<section id="aligners-metrics" class="slide level2">
<h2>Aligners • Metrics</h2>
<div class="columns">
<div class="column" style="width:40%;">
<p><img data-src="assets/compare-aligners-speed.png" style="width:90.0%"></p>
<p><span class="citation" data-cites="baruzzo2017simulation">Baruzzo et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></p>
</div><div class="column" style="width:60%;">
<p><img data-src="assets/compare-aligners-quality.jpg"></p>
<p><img data-src="assets/accuracy.svg" style="width:50.0%"></p>
</div></div>
</section>
<section id="aligners-time-and-ram" class="slide level2 notes">
<h2>Aligners time and RAM</h2>
<table class="caption-top">
<thead>
<tr class="header">
<th>Program</th>
<th>Time_Min</th>
<th>Memory_GB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>HISATx1</td>
<td>22.7</td>
<td>4.3</td>
</tr>
<tr class="even">
<td>HISATx2</td>
<td>47.7</td>
<td>4.3</td>
</tr>
<tr class="odd">
<td>HISAT</td>
<td>26.7</td>
<td>4.3</td>
</tr>
<tr class="even">
<td>STAR</td>
<td>25</td>
<td>28</td>
</tr>
<tr class="odd">
<td>STARx2</td>
<td>50.5</td>
<td>28</td>
</tr>
<tr class="even">
<td>GSNAP</td>
<td>291.9</td>
<td>20.2</td>
</tr>
<tr class="odd">
<td>TopHat2</td>
<td>1170</td>
<td>4.3</td>
</tr>
</tbody>
</table>
</section>
<section id="section" class="slide level2">
<h2></h2>
<ul>
<li>Reads (FASTQ)</li>
</ul>
<pre><code>@ST-E00274:179:HHYMLALXX:8:1101:1641:1309 1:N:0:NGATGT
NCATCGTGGTATTTGCACATCTTTTCTTATCAAATAAAAAGTTTAACCTACTCAGTTATGCGCATACGTTTTTTGATGGCATTTCCATAAACCGATTTTTTTTTTATGCACGTACCCAAAACGTGCAGAAAAATACGCTGCTAGAAATGTA
+
#AAAFAFA&lt;-AFFJJJAFA-FFJJJJFFFAJJJJ-&lt;FFJJJ-A-F-7--FA7F7-----FFFJFA&lt;FFFFJ&lt;AJ--FF-A&lt;A-&lt;JJ-7-7-&lt;FF-FFFJAFFAA--A--7FJ-7----77-A--7F7)---7F-A----7)7-----7&lt;&lt;-</code></pre>
<p><code>@instrument:runid:flowcellid:lane:tile:xpos:ypos read:isfiltered:controlnumber:sampleid</code></p>
<ul>
<li>Reference Genome/Transcriptome (FASTA)</li>
</ul>
<pre><code>&gt;1 dna:chromosome chromosome:GRCz10:1:1:58871917:1 REF
GATCTTAAACATTTATTCCCCCTGCAAACATTTTCAATCATTACATTGTCATTTCCCCTC</code></pre>
<ul>
<li>Annotation (GTF/GFF)</li>
</ul>
<pre><code>#!genome-build GRCz10
4       ensembl_havana  gene    6732    52059   .       -       .       gene_id "ENSDARG00000104632"; gene_version "2"; gene_name "rerg"; gene_source "ensembl_havana"; gene_biotype "protein_coding"; havana_gene "OTTDARG00000044080"; havana_gene_version "1";</code></pre>
<p><code>seq source feature start end score strand frame attribute</code></p>
<p><a href="https://emea.support.illumina.com/bulletins/2016/04/fastq-files-explained.html">Illumina FASTQ format</a>, <a href="https://www.ensembl.org/info/website/upload/gff.html">GTF format</a></p>
</section>
<section id="alignment" class="slide level2">
<h2>Alignment</h2>
<ul>
<li>SAM/BAM (Sequence Alignment Map format)</li>
</ul>
<pre><code>ST-E00274:188:H3JWNCCXY:4:1102:32431:49900      163     1       1       60      8S139M4S      =       385     535     TATTTAGAGATCTTAAACATCCATTCCCCCTGCAAACATTTTCAATCATTACATTGTCATTTTCCCTCCAAATTAAATTTAGCCAGAGGCGCACAACATACGACCTCTAAAAAAGGTGCTGGAACATGTACCTATATGCAGCACCACCATC     AAAFAFFAFFFFJ7FFFFJ&lt;JAFA7F-&lt;AJ7JJ&lt;FFFJ--&lt;FAJF&lt;7&lt;7FAFJ-&lt;AFA&lt;-JJJ-AF-AJ-FF&lt;F--A&lt;FF&lt;-7777-7JA-77A---F-7AAFF-FJA--77FJ&lt;--77)))7&lt;JJA&lt;J77&lt;-------&lt;7--))7)))7-    NM:i:4   MD:Z:12T0T40C58T25      AS:i:119        XS:i:102        XA:Z:17,-53287490,4S33M4D114M,11;     MQ:i:60 MC:Z:151M       RG:Z:ST-E00274_188_H3JWNCCXY_4</code></pre>
<p><code>query flag ref pos mapq cigar mrnm mpos tlen seq qual opt</code></p>
<p><strong>Never store alignment files in raw SAM format. Always compress it!</strong> <a href="http://www.htslib.org/doc/sam.html">SAM format</a></p>
<table class="caption-top">
<thead>
<tr class="header">
<th>Format</th>
<th>Size_GB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SAM</td>
<td>7.4</td>
</tr>
<tr class="even">
<td>BAM</td>
<td>1.9</td>
</tr>
<tr class="odd">
<td>CRAM lossless Q</td>
<td>1.4</td>
</tr>
<tr class="even">
<td>CRAM 8 bins Q</td>
<td>0.8</td>
</tr>
<tr class="odd">
<td>CRAM no Q</td>
<td>0.26</td>
</tr>
</tbody>
</table>
</section>
<section id="visualisation-igv" class="slide level2">
<h2>Visualisation • IGV</h2>

<img data-src="assets/igv.png" class="r-stretch"><p><a href="http://software.broadinstitute.org/software/igv/">IGV</a>, <a href="https://genome.ucsc.edu/">UCSC Genome Browser</a>, <a href="https://www.bioinformatics.babraham.ac.uk/projects/seqmonk/">SeqMonk</a>, <a href="https://cmdcolin.github.io/awesome-genome-visualization/?latest=true">More</a></p>
</section>
<section id="visualisation-tview" class="slide level2">
<h2>Visualisation • <code>tview</code></h2>
<p><code>samtools tview alignment.bam genome.fasta</code></p>

<img data-src="assets/tview.png" class="r-stretch"></section>
<section id="visualisation-seqmonk" class="slide level2">
<h2>Visualisation • SeqMonk</h2>

<img data-src="assets/seqmonk.jpg" class="r-stretch"><p><a href="https://www.bioinformatics.babraham.ac.uk/projects/seqmonk/">SeqMonk</a></p>
</section>
<section id="alignment-qc" class="slide level2">
<h2>Alignment QC</h2>
<ul>
<li>Number of reads mapped/unmapped/paired etc</li>
<li>Uniquely mapped</li>
<li>Insert size distribution</li>
<li>Coverage</li>
<li>Gene body coverage</li>
<li>Biotype counts / Chromosome counts</li>
<li>Counts by region: gene/intron/non-genic</li>
<li>Sequencing saturation</li>
<li>Strand specificity</li>
</ul>
<p>STAR (final log file), samtools stats, bamtools stats, <a href="https://hartleys.github.io/QoRTs/">QoRTs</a>, <a href="http://rseqc.sourceforge.net/">RSeQC</a>, <a href="http://qualimap.bioinfo.cipf.es/">Qualimap</a></p>
</section>
<section id="alignment-qc-star-log" class="slide level2">
<h2>Alignment QC • STAR Log</h2>
<p>MultiQC can be used to summarise and plot STAR log files.</p>

<img data-src="assets/star_alignment_plot.svg" style="width:95.0%" class="r-stretch"></section>
<section id="alignment-qc-features" class="slide level2">
<h2>Alignment QC • Features</h2>
<p>QoRTs was run on all samples and summarised using MultiQC.</p>

<img data-src="assets/qorts_alignments.svg" style="width:95.0%" class="r-stretch"></section>
<section id="alignment-qc-qorts" class="slide level2">
<h2>Alignment QC • QoRTs</h2>

<img data-src="assets/qorts.png" style="width:95.0%" class="r-stretch"></section>
<section id="alignment-qc-examples" class="slide level2">
<h2>Alignment QC • Examples</h2>
<div class="columns">
<div class="column">
<p><strong>Read mapping profile</strong> <img data-src="assets/clipping_good.png"></p>
</div><div class="column">
<p><strong>Gene body coverage</strong> <img data-src="assets/gene-body-coverage.png"><br>
<span class="citation" data-cites="sigurgeirsson2014sequencing">Sigurgeirsson et al. (<a href="#/references" role="doc-biblioref" onclick="">2014</a>)</span></p>
</div></div>
<aside class="notes">
<p>The read mapping profile shows how well bases along a read mapped to the reference for all reads. Mapability usually decreases towards the 5’ and 3’ ends due to soft-clipping. This is is more pronounced in untrimmed reads.</p>
<p>Gene body coverage shows read coverage over the gene for all genes. An even coverage is ideally expected. A bias in either direction could be indicative of library quality. A bias toward the 3’ end is usually seen in polyA selected libraries.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="alignment-qc-examples-1" class="slide level2">
<h2>Alignment QC • Examples</h2>
<div class="columns">
<div class="column">
<p><strong>Insert size</strong><br>
<img data-src="assets/inner_distance.png"></p>
</div><div class="column">
<p><strong>Saturation curve</strong><br>
<img data-src="assets/saturation.jpg" style="width:70.0%"></p>
<p><span class="citation" data-cites="francis2013comparison">Francis et al. (<a href="#/references" role="doc-biblioref" onclick="">2013</a>)</span></p>
</div></div>
<aside class="notes">
<p>Negative insert size implies overlapping mate pairs.</p>
<p>Conserved genes in the mouse transcriptome. Saturation curves of discovery of genes in the mouse heart from a set of a subset of 248 conserved orthologs; genes which have any blast hit are shown in circles; genes which the translated protein was within the expected size range of the conserved gene are in squares; proteins which are 100% identical to a canonical protein in Uniprot/Swissprot mouse database are shown in triangles.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="quantification-counts" class="slide level2">
<h2>Quantification • Counts</h2>
<div class="columns">
<div class="column">
<ul>
<li>Read counts = gene expression</li>
<li>Intersection on gene models</li>
<li>Reads can be quantified on any feature (gene, transcript, exon etc)</li>
</ul>
<p><img data-src="assets/rnaseq_counts.svg"></p>
<p><a href="https://subread.sourceforge.net/">featureCounts</a>, <a href="https://github.com/simon-anders/htseq">HTSeq</a></p>
</div><div class="column">
<p><img data-src="assets/rnaseq_union.svg" style="width:60.0%"></p>
</div></div>
</section>
<section id="quantification" class="slide level2">
<h2>Quantification</h2>
<div class="columns">
<div class="column">
<p><strong>PCR duplicates</strong></p>
<ul>
<li>Computational deduplication not recommended <span class="citation" data-cites="klepikova2017effect">Klepikova et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span>, <span class="citation" data-cites="parekh2016impact">Parekh et al. (<a href="#/references" role="doc-biblioref" onclick="">2016</a>)</span></li>
<li>Use PCR-free library-prep kits</li>
<li>Use UMIs during library-prep <span class="citation" data-cites="fu2018elimination">Fu et al. (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span></li>
</ul>
<p><strong>Multi-mapping</strong></p>
<ul>
<li>Added (BEDTools multicov)</li>
<li>Discard (featureCounts, HTSeq)</li>
<li>Distribute counts (Cufflinks, featureCounts)</li>
<li>Rescue
<ul>
<li>Probabilistic assignment (Rcount, Cufflinks)</li>
<li>Prioritise features (Rcount)</li>
<li>Probabilistic assignment with EM (RSEM)</li>
</ul></li>
</ul>
</div><div class="column">
<p><img data-src="assets/pcr-duplicates.png"></p>
</div></div>
<aside class="notes">
<p>Tools that detect duplicate reads (like Picard MarkDuplicates) is looking for reads that start and end at the same coordinates. Such reads are expected to be PCR derived and can be collapsed down to 1 read. This makes sense in whole genome sequencing, but in RNA-Seq, this is naturally expected since highly expressed genes create huge number of identical transcripts which may fragment at hotspots leading to fragmentation bias <span class="citation" data-cites="roberts2011improving">Roberts et al. (<a href="#/references" role="doc-biblioref" onclick="">2011</a>)</span>. Therefore, removing PCR duplicates may remove real duplicates as well.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="quantification-abundance" class="slide level2">
<h2>Quantification • Abundance</h2>
<ul>
<li>Count methods
<ul>
<li>Provide no inference on isoforms</li>
<li>Cannot accurately measure fold change</li>
</ul></li>
</ul>
<!--.size-60[![](assets/rnaseq_count_issues.svg)]-->
<ul>
<li>Probabilistic assignment
<ul>
<li>Deconvolute ambiguous mappings</li>
<li>Transcript-level</li>
<li>cDNA reference</li>
</ul></li>
</ul>
<p><strong>Kallisto, Salmon</strong></p>
<ul>
<li>Ultra-fast &amp; alignment-free</li>
<li>Bootstrapping &amp; quantification confidence</li>
<li>Transcript-level counts</li>
<li>Transcript-level estimates improves gene-level estimates <span class="citation" data-cites="soneson2015differential">Soneson et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span>, <a href="https://bioconductor.org/packages/release/bioc/html/tximport.html">tximport</a></li>
<li>Evaluation and comparison of isoform quantification tools <span class="citation" data-cites="zhang2017evaluation">Zhang et al. (<a href="#/references" role="doc-biblioref" onclick="">2017</a>)</span></li>
</ul>
<p><a href="https://deweylab.github.io/RSEM/">RSEM</a>, <a href="https://pachterlab.github.io/kallisto/">Kallisto</a>, <a href="https://combine-lab.github.io/salmon/">Salmon</a></p>
<aside class="notes">
<p>Gene expression counts may be literal counts (reads overlapping to gene models) or probabilistically estimated based on the data. Proponents of probabilistically estimation claims that it is a better estimate that corrects for sampling biases.</p>
<p>Tools such as RSEM, Kallisto and Salmon produce probabilistic abundance estimates. <span class="citation" data-cites="soneson2015differential">Soneson et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span> claims that estimating abundance using Kallisto at the transcript levels and then summarising down to gene-level counts results in improved estimates of gene expression.</p>
<p>Kallisto/Salmon -&gt; transcript-counts -&gt; <a href="https://bioconductor.org/packages/release/bioc/html/tximport.html"><code>tximport()</code></a> -&gt; gene-counts</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="quantification-qc" class="slide level2">
<h2>Quantification QC</h2>
<pre><code>ENSG00000000003    140   242   188   143   287   344   438   280   253
ENSG00000000005    0     0     0     0     0     0     0     0     0
ENSG00000000419    69    98    77    55    52    94    116   79    69
ENSG00000000457    56    75    104   79    157   205   183   178   153
ENSG00000000460    33    27    23    19    27    42    69    44    40</code></pre>

<img data-src="assets/correlation.png" style="width:50.0%" class="r-stretch"><p>Pairwise correlation between samples must be high (&gt;0.9)</p>
</section>
<section id="multiqc" class="slide level2">
<h2>MultiQC</h2>
<p><a href="https://multiqc.info/examples/rna-seq/multiqc_report.html"><img data-src="assets/multiqc.png" style="width:90.0%"></a></p>
</section>
<section id="normalization" class="slide level2">
<h2>Normalization</h2>
<div class="columns">
<div class="column" style="width:70%;">
<ul>
<li>Control for Sequencing depth, compositional bias and more</li>
<li>Median of Ratios (DESeq2) and TMM (edgeR) perform the best</li>
</ul>
<p><img data-src="assets/normalisation.png"></p>
<ul>
<li>For DGE using DGE packages, use raw counts</li>
<li>For clustering, heatmaps etc use VST, VOOM or RLOG</li>
<li>For own analysis, plots etc, use TPM</li>
<li>Other solutions: spike-ins/house-keeping genes</li>
</ul>
<p><span class="citation" data-cites="dillies2013comprehensive">Dillies et al. (<a href="#/references" role="doc-biblioref" onclick="">2013</a>)</span>, <span class="citation" data-cites="evans2018selecting">Evans et al. (<a href="#/references" role="doc-biblioref" onclick="">2018</a>)</span>, <span class="citation" data-cites="wagner2012measurement">Wagner et al. (<a href="#/references" role="doc-biblioref" onclick="">2012</a>)</span></p>
</div><div class="column" style="width:30%;">
<p><img data-src="assets/distribution.jpg"></p>
</div></div>
</section>
<section id="exploratory" class="slide level2">
<h2>Exploratory</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li>Remove lowly expressed genes</li>
<li>Heatmaps, MDS, PCA etc.</li>
</ul>
<p><img data-src="assets/dge-clustering.png"></p>
<p><a href="https://github.com/raivokolde/pheatmap">pheatmap</a></p>
</div><div class="column" style="width:40%;">
<ul>
<li>Transform raw counts to VST, VOOM, RLOG, TPM etc</li>
</ul>
<p><img data-src="assets/dge-mds.png"></p>
</div></div>
</section>
<section id="batch-correction" class="slide level2">
<h2>Batch correction</h2>
<ul>
<li>Estimate variation explained by variables (<a href="https://bioconductor.org/packages/release/bioc/html/pvca.html">PVCA</a>)</li>
</ul>

<img data-src="assets/pvca.png" style="width:60.0%" class="r-stretch"><ul>
<li>Find confounding effects as surrogate variables (<a href="http://bioconductor.org/packages/release/bioc/html/sva.html">SVA</a>)</li>
<li>Model known batches in the LM/GLM model</li>
<li>Correct known batches (ComBat from SVA)(Can overcorrect! <span class="citation" data-cites="zindler2020simulating">Zindler et al. (<a href="#/references" role="doc-biblioref" onclick="">2020</a>)</span>)</li>
<li>Interactively evaluate batch effects and correction (<a href="http://bioconductor.org/packages/release/bioc/html/BatchQC.html">BatchQC</a>) <span class="citation" data-cites="manimaran2016batchqc">Manimaran et al. (<a href="#/references" role="doc-biblioref" onclick="">2016</a>)</span></li>
</ul>
</section>
<section id="differential-expression" class="slide level2">
<h2>Differential expression</h2>
<div class="columns">
<div class="column" style="width:65%;">
<p><img data-src="assets/dge.png"></p>
<ul>
<li>Univariate testing gene-by-gene</li>
<li>More descriptive, less predictive</li>
</ul>
</div><div class="column" style="width:35%;">
<div class="fragment">
<p><img data-src="assets/dge-scatter.png" style="width:80.0%"></p>
</div>
</div></div>
</section>
<section id="differential-expression-1" class="slide level2 notes">
<h2>Differential expression</h2>
<ul>
<li>DESeq2, edgeR (Neg-binom &gt; GLM &gt; Test)</li>
<li>Limma-Voom (Neg-binom &gt; Voom-transform &gt; LM &gt; Test)</li>
<li>DESeq2 <code>~age+condition</code>
<ul>
<li>Estimate size factors <code>estimateSizeFactors()</code></li>
<li>Estimate gene-wise dispersion <code>estimateDispersions()</code></li>
<li>Fit curve to gene-wise dispersion estimates</li>
<li>Shrink gene-wise dispersion estimates</li>
<li>GLM fit for each gene</li>
<li>Wald test <code>nbinomWaldTest()</code></li>
</ul></li>
</ul>

<img data-src="assets/dge-dispersion.png" style="width:50.0%" class="r-stretch"><p><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>, <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>, <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a>, <span class="citation" data-cites="seyednasrollah2015comparison">Seyednasrollah et al. (<a href="#/references" role="doc-biblioref" onclick="">2015</a>)</span></p>
</section>
<section id="dge" class="slide level2">
<h2>DGE</h2>
<ul>
<li>Results <code>results()</code></li>
</ul>
<pre><code>log2 fold change (MLE): type type2 vs control
Wald test p-value: type type2 vs control
DataFrame with 1 row and 6 columns
                        baseMean     log2FoldChange             lfcSE
                       &lt;numeric&gt;          &lt;numeric&gt;         &lt;numeric&gt;
ENSG00000000003 242.307796723287 -0.932926089608546 0.114285150312285
                             stat               pvalue                 padj
                        &lt;numeric&gt;            &lt;numeric&gt;            &lt;numeric&gt;
ENSG00000000003 -8.16314356729037 3.26416150242775e-16 1.36240609998527e-14</code></pre>
<ul>
<li>Summary <code>summary()</code></li>
</ul>
<pre><code>out of 17889 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4526, 25%
LFC &lt; 0 (down)     : 5062, 28%
outliers [1]       : 25, 0.14%
low counts [2]     : 0, 0%
(mean count &lt; 3)</code></pre>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column">
<ul>
<li>MA plot <code>plotMA()</code></li>
</ul>
<p><img data-src="assets/dge-ma.png"></p>
<ul>
<li>Volcano plot</li>
</ul>
<p><img data-src="assets/dge-volcano.png"></p>
</div><div class="column">
<ul>
<li>Normalised counts <code>plotCounts()</code></li>
</ul>
<p><img data-src="assets/dge-counts.png"></p>
<p><img data-src="assets/scattered.gif" style="width:55.0%"></p>
</div></div>
</section>
<section id="functional-analysis-gene-ontology" class="slide level2">
<h2>Functional analysis • Gene Ontology</h2>
<div class="columns">
<div class="column" style="width:70%;">
<ul>
<li>Gene set analysis (GSA)</li>
<li>Gene set enrichment analysis (GSEA)</li>
<li>Gene ontology / Reactome databases</li>
</ul>
<p><img data-src="assets/go.jpg" style="width:70.0%"></p>
</div><div class="column" style="width:30%;">
<p><img data-src="assets/systembio.png"></p>
</div></div>
</section>
<section id="functional-analysis-kegg" class="slide level2">
<h2>Functional analysis • Kegg</h2>
<ul>
<li>Pathway analysis (Kegg)</li>
</ul>

<img data-src="assets/kegg.png" style="width:80.0%" class="r-stretch"><p><a href="https://www.webgestalt.org/">Webgestalt</a>, <a href="https://maayanlab.cloud/Enrichr/">EnrichR</a>, <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>, <a href="http://apps.cytoscape.org/apps/cluego">ClueGO</a>, <a href="https://bioconductor.org/packages/release/bioc/html/pathview.html">pathview</a></p>
</section>
<section id="summary" class="slide level2">
<h2>Summary</h2>
<ul>
<li>Sound experimental design to avoid confounding</li>
<li>Plan carefully about lib prep, sequencing etc based on experimental objective</li>
<li>For DGE, biological replicates may be more important than other considerations (paired-end, sequencing depth, long reads etc)</li>
<li>Discard low quality bases, reads, genes and samples</li>
<li>Verify that tools and methods align with data assumptions</li>
<li>Experiment with multiple pipelines and tools</li>
<li>QC! QC everything at every step</li>
</ul>
<blockquote>
<p><span class="large">Conesa, A., Madrigal, P., Tarazona, S., Gomez-Cabrero, D., Cervera, A., McPherson, A., … &amp; Mortazavi, A. (2016). A survey of best practices for RNA-seq data analysis. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8">Genome biology, 17(1), 1-19.</a></span></p>
</blockquote>
</section>
<section id="further-learning" class="slide level2">
<h2>Further learning</h2>
<div class="columns">
<div class="column" style="width:75%;">
<ul>
<li>Griffith lab <a href="https://github.com/griffithlab/rnaseq_tutorial/wiki">RNA-Seq using HiSat &amp; StringTie tutorial</a></li>
<li>HBC Training <a href="https://hbctraining.github.io/main/">DGE using DeSeq2 tutorial</a></li>
<li>NBIS Workshop | <a href="https://nbisweden.github.io/workshop-RNAseq/">Analysis of RNA-Seq data</a></li>
<li>SciLifeLab <a href="https://training.scilifelab.se/">training</a></li>
<li><a href="https://www.rna-seqblog.com/">RNA-Seq Blog</a></li>
</ul>
</div><div class="column" style="width:25%;">
<p><img data-src="assets/help.png"></p>
</div></div>
</section>
<section id="section-2" class="slide level2" data-background-image="../../assets/images/cover.jpg">
<h2></h2>
<section id="thank-you.-questions" class="bottom-align">
<h3>Thank you. Questions?</h3>
</section>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-baruzzo2017simulation" class="csl-entry" role="listitem">
Baruzzo, G., Hayer, K. E., Kim, E. J., Di Camillo, B., FitzGerald, G. A., &amp; Grant, G. R. (2017). Simulation-based comprehensive benchmarking of RNA-seq aligners. <em>Nature Methods</em>, <em>14</em>(2), 135–139. <a href="https://www.nature.com/articles/nmeth.4106">https://www.nature.com/articles/nmeth.4106</a>
</div>
<div id="ref-chhangawala2015impact" class="csl-entry" role="listitem">
Chhangawala, S., Rudy, G., Mason, C. E., &amp; Rosenfeld, J. A. (2015). The impact of read length on quantification of differentially expressed genes and splice junction detection. <em>Genome Biology</em>, <em>16</em>(1), 1–10. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531809/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531809/</a>
</div>
<div id="ref-conesa2016survey" class="csl-entry" role="listitem">
Conesa, A., Madrigal, P., Tarazona, S., Gomez-Cabrero, D., Cervera, A., McPherson, A., Szcześniak, M. W., Gaffney, D. J., Elo, L. L., Zhang, X., et al. (2016). A survey of best practices for RNA-seq data analysis. <em>Genome Biology</em>, <em>17</em>(1), 1–19.
</div>
<div id="ref-corley2017differentially" class="csl-entry" role="listitem">
Corley, S. M., MacKenzie, K. L., Beverdam, A., Roddam, L. F., &amp; Wilkins, M. R. (2017). Differentially expressed genes from RNA-seq and functional enrichment results are affected by the choice of single-end versus paired-end reads and stranded versus non-stranded protocols. <em>BMC Genomics</em>, <em>18</em>(1), 1–13. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5442695/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5442695/</a>
</div>
<div id="ref-daley2013predicting" class="csl-entry" role="listitem">
Daley, T., &amp; Smith, A. D. (2013). Predicting the molecular complexity of sequencing libraries. <em>Nature Methods</em>, <em>10</em>(4), 325–327. <a href="https://www.nature.com/articles/nmeth.2375">https://www.nature.com/articles/nmeth.2375</a>
</div>
<div id="ref-dillies2013comprehensive" class="csl-entry" role="listitem">
Dillies, M.-A., Rau, A., Aubert, J., Hennequet-Antier, C., Jeanmougin, M., Servant, N., Keime, C., Marot, G., Castel, D., Estelle, J., et al. (2013). A comprehensive evaluation of normalization methods for illumina high-throughput RNA sequencing data analysis. <em>Briefings in Bioinformatics</em>, <em>14</em>(6), 671–683.
</div>
<div id="ref-evans2018selecting" class="csl-entry" role="listitem">
Evans, C., Hardin, J., &amp; Stoebel, D. M. (2018). Selecting between-sample RNA-seq normalization methods from the perspective of their assumptions. <em>Briefings in Bioinformatics</em>, <em>19</em>(5), 776–792.
</div>
<div id="ref-francis2013comparison" class="csl-entry" role="listitem">
Francis, W. R., Christianson, L. M., Kiko, R., Powers, M. L., Shaner, N. C., &amp; D Haddock, S. H. (2013). A comparison across non-model animals suggests an optimal sequencing depth for de novotranscriptome assembly. <em>BMC Genomics</em>, <em>14</em>(1), 1–12. <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-167">https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-167</a>
</div>
<div id="ref-fu2018elimination" class="csl-entry" role="listitem">
Fu, Y., Wu, P.-H., Beane, T., Zamore, P. D., &amp; Weng, Z. (2018). Elimination of PCR duplicates in RNA-seq and small RNA-seq using unique molecular identifiers. <em>Bmc Genomics</em>, <em>19</em>, 1–14.
</div>
<div id="ref-gallego2014rna" class="csl-entry" role="listitem">
Gallego Romero, I., Pai, A. A., Tung, J., &amp; Gilad, Y. (2014). RNA-seq: Impact of RNA degradation on transcript quantification. <em>BMC Biology</em>, <em>12</em>(1), 1–13. <a href="https://bmcbiol.biomedcentral.com/articles/10.1186/1741-7007-12-42">https://bmcbiol.biomedcentral.com/articles/10.1186/1741-7007-12-42</a>
</div>
<div id="ref-hart2013calculating" class="csl-entry" role="listitem">
Hart, S. N., Therneau, T. M., Zhang, Y., Poland, G. A., &amp; Kocher, J.-P. (2013). Calculating sample size estimates for RNA sequencing data. <em>Journal of Computational Biology</em>, <em>20</em>(12), 970–978. <a href="https://www.liebertpub.com/doi/10.1089/cmb.2012.0283">https://www.liebertpub.com/doi/10.1089/cmb.2012.0283</a>
</div>
<div id="ref-hsieh2019effect" class="csl-entry" role="listitem">
Hsieh, P.-H., Oyang, Y.-J., &amp; Chen, C.-Y. (2019). Effect of de novo transcriptome assembly on transcript quantification. <em>Scientific Reports</em>, <em>9</em>(1), 8304. <a href="https://www.nature.com/articles/s41598-019-44499-3">https://www.nature.com/articles/s41598-019-44499-3</a>
</div>
<div id="ref-klepikova2017effect" class="csl-entry" role="listitem">
Klepikova, A. V., Kasianov, A. S., Chesnokov, M. S., Lazarevich, N. L., Penin, A. A., &amp; Logacheva, M. (2017). Effect of method of deduplication on estimation of differential gene expression using RNA-seq. <em>PeerJ</em>, <em>5</em>, e3091. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5357343/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5357343/</a>
</div>
<div id="ref-levin2010comprehensive" class="csl-entry" role="listitem">
Levin, J. Z., Yassour, M., Adiconis, X., Nusbaum, C., Thompson, D. A., Friedman, N., Gnirke, A., &amp; Regev, A. (2010). Comprehensive comparative analysis of strand-specific RNA sequencing methods. <em>Nature Methods</em>, <em>7</em>(9), 709–715. <a href="https://www.nature.com/articles/nmeth.1491">https://www.nature.com/articles/nmeth.1491</a>
</div>
<div id="ref-liao2020read" class="csl-entry" role="listitem">
Liao, Y., &amp; Shi, W. (2020). Read trimming is not required for mapping and quantification of RNA-seq reads at the gene level. <em>NAR Genomics and Bioinformatics</em>, <em>2</em>(3), lqaa068. <a href="https://pubmed.ncbi.nlm.nih.gov/33575617/">https://pubmed.ncbi.nlm.nih.gov/33575617/</a>
</div>
<div id="ref-liu2014rna" class="csl-entry" role="listitem">
Liu, Y., Zhou, J., &amp; White, K. P. (2014). RNA-seq differential expression studies: More sequence or more replication? <em>Bioinformatics</em>, <em>30</em>(3), 301–304. <a href="https://academic.oup.com/bioinformatics/article/30/3/301/228651">https://academic.oup.com/bioinformatics/article/30/3/301/228651</a>
</div>
<div id="ref-manimaran2016batchqc" class="csl-entry" role="listitem">
Manimaran, S., Selby, H. M., Okrah, K., Ruberman, C., Leek, J. T., Quackenbush, J., Haibe-Kains, B., Bravo, H. C., &amp; Johnson, W. E. (2016). BatchQC: Interactive software for evaluating sample and batch effects in genomic data. <em>Bioinformatics</em>, <em>32</em>(24), 3836–3838.
</div>
<div id="ref-marioni2008rna" class="csl-entry" role="listitem">
Marioni, J. C., Mason, C. E., Mane, S. M., Stephens, M., &amp; Gilad, Y. (2008). RNA-seq: An assessment of technical reproducibility and comparison with gene expression arrays. <em>Genome Research</em>, <em>18</em>(9), 1509–1517. <a href="https://genome.cshlp.org/content/18/9/1509.long">https://genome.cshlp.org/content/18/9/1509.long</a>
</div>
<div id="ref-parekh2016impact" class="csl-entry" role="listitem">
Parekh, S., Ziegenhain, C., Vieth, B., Enard, W., &amp; Hellmann, I. (2016). The impact of amplification on differential expression analyses by RNA-seq. <em>Scientific Reports</em>, <em>6</em>(1), 25533. <a href="https://www.nature.com/articles/srep25533">https://www.nature.com/articles/srep25533</a>
</div>
<div id="ref-roberts2011improving" class="csl-entry" role="listitem">
Roberts, A., Trapnell, C., Donaghey, J., Rinn, J. L., &amp; Pachter, L. (2011). Improving RNA-seq expression estimates by correcting for fragment bias. <em>Genome Biology</em>, <em>12</em>(3), 1–14.
</div>
<div id="ref-schurch2016many" class="csl-entry" role="listitem">
Schurch, N. J., Schofield, P., Gierliński, M., Cole, C., Sherstnev, A., Singh, V., Wrobel, N., Gharbi, K., Simpson, G. G., Owen-Hughes, T., et al. (2016). How many biological replicates are needed in an RNA-seq experiment and which differential expression tool should you use? <em>Rna</em>, <em>22</em>(6), 839–851. <a href="https://rnajournal.cshlp.org/content/early/2016/03/30/rna.053959.115.abstract">https://rnajournal.cshlp.org/content/early/2016/03/30/rna.053959.115.abstract</a>
</div>
<div id="ref-seyednasrollah2015comparison" class="csl-entry" role="listitem">
Seyednasrollah, F., Laiho, A., &amp; Elo, L. L. (2015). Comparison of software packages for detecting differential expression in RNA-seq studies. <em>Briefings in Bioinformatics</em>, <em>16</em>(1), 59–70.
</div>
<div id="ref-sigurgeirsson2014sequencing" class="csl-entry" role="listitem">
Sigurgeirsson, B., Emanuelsson, O., &amp; Lundeberg, J. (2014). Sequencing degraded RNA addressed by 3’tag counting. <em>PloS One</em>, <em>9</em>(3), e91851. <a href="https://pubmed.ncbi.nlm.nih.gov/24632678/">https://pubmed.ncbi.nlm.nih.gov/24632678/</a>
</div>
<div id="ref-soneson2015differential" class="csl-entry" role="listitem">
Soneson, C., Love, M. I., &amp; Robinson, M. D. (2015). Differential analyses for RNA-seq: Transcript-level estimates improve gene-level inferences. <em>F1000Research</em>, <em>4</em>.
</div>
<div id="ref-wagner2012measurement" class="csl-entry" role="listitem">
Wagner, G. P., Kin, K., &amp; Lynch, V. J. (2012). Measurement of mRNA abundance using RNA-seq data: RPKM measure is inconsistent among samples. <em>Theory in Biosciences</em>, <em>131</em>, 281–285.
</div>
<div id="ref-wang2017comprehensive" class="csl-entry" role="listitem">
Wang, S., &amp; Gribskov, M. (2017). Comprehensive evaluation of de novo transcriptome assembly programs and their effects on differential gene expression analysis. <em>Bioinformatics</em>, <em>33</em>(3), 327–333. <a href="https://academic.oup.com/bioinformatics/article/33/3/327/2580374">https://academic.oup.com/bioinformatics/article/33/3/327/2580374</a>
</div>
<div id="ref-zhang2017evaluation" class="csl-entry" role="listitem">
Zhang, C., Zhang, B., Lin, L.-L., &amp; Zhao, S. (2017). Evaluation and comparison of computational tools for RNA-seq isoform quantification. <em>BMC Genomics</em>, <em>18</em>(1), 1–11.
</div>
<div id="ref-zhao2018rnaseqsamplesize" class="csl-entry" role="listitem">
Zhao, S., Li, C.-I., Guo, Y., Sheng, Q., &amp; Shyr, Y. (2018). RnaSeqSampleSize: Real data based sample size estimation for RNA sequencing. <em>BMC Bioinformatics</em>, <em>19</em>(1), 1–8. <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2191-5">https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2191-5</a>
</div>
<div id="ref-zhao2015comparison" class="csl-entry" role="listitem">
Zhao, S., Zhang, Y., Gordon, W., Quan, J., Xi, H., Du, S., Schack, D. von, &amp; Zhang, B. (2015). Comparison of stranded and non-stranded RNA-seq transcriptome profiling and investigation of gene overlap. <em>BMC Genomics</em>, <em>16</em>(1), 1–14. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4559181/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4559181/</a>
</div>
<div id="ref-zindler2020simulating" class="csl-entry" role="listitem">
Zindler, T., Frieling, H., Neyazi, A., Bleich, S., &amp; Friedel, E. (2020). Simulating ComBat: How batch correction can lead to the systematic introduction of false positive results in DNA methylation microarray studies. <em>BMC Bioinformatics</em>, <em>21</em>, 1–15.
</div>
</div>
</section>
<section id="hands-on-tutorial" class="slide level2">
<h2>Hands-On tutorial</h2>
<p><strong>Main exercise</strong></p>
<ul>
<li>01 Check the quality of the raw reads with <strong>FastQC</strong></li>
<li>02 Map the reads to the reference genome using <strong>HISAT2</strong></li>
<li>03 Assess the post-alignment quality using <strong>QualiMap</strong></li>
<li>04 Count the reads overlapping with genes using <strong>featureCounts</strong></li>
<li>05 Find differentially expressed genes using <strong>DESeq2</strong> in R</li>
<li>06 RNA-Seq figures and plots using <strong>R</strong></li>
</ul>
<p><strong>Bonus exercises</strong></p>
<ul>
<li>01 Functional annotation of DE genes using <strong>GO</strong> database</li>
<li>02 Visualisation of RNA-seq BAM files using <strong>IGV</strong> genome browser</li>
</ul>
<p><span class="large"><strong>Data:</strong> <code>/sw/courses/ngsintro/rnaseq/dardel</code></span><br>
<span class="large"><strong>Work:</strong> <code>/cfs/klemming/projects/supr/naiss2025-22-1240/username/rnaseq/</code></span></p>
</section>
<section id="hands-on-tutorial-1" class="slide level2">
<h2>Hands-On tutorial</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>Course data directory</li>
</ul>
<p><code>/sw/courses/ngsintro/rnaseq/dardel</code></p>
<pre><code>dardel/
├── extras
├── main
│&nbsp;&nbsp; ├── 1_raw
│&nbsp;&nbsp; ├── 2_fastqc
│&nbsp;&nbsp; ├── 3_mapping
│&nbsp;&nbsp; ├── 4_qorts
│&nbsp;&nbsp; ├── 4_qualimap
│&nbsp;&nbsp; ├── 5_dge
│&nbsp;&nbsp; ├── 6_multiqc
│&nbsp;&nbsp; ├── reference
│   │   └── mouse_chr19_hisat2
│&nbsp;&nbsp; └── scripts
├── main_full
│&nbsp;&nbsp; └── nextflow
├── r
└── README.md</code></pre>
</div><div class="column" style="width:60%;">
<ul>
<li>Your work directory</li>
</ul>
<p><code>/cfs/klemming/projects/supr/naiss2025-22-1240/username/rnaseq/</code></p>
<pre><code>rnaseq/
├── 1_raw
├── 2_fastqc
├── 3_mapping
├── 4_qualimap
├── 5_dge
├── 6_multiqc
├── reference
└── scripts</code></pre>
</div></div>


<div class="reveal-header">
<div class="header-logo-left">
<a href="https://nbisweden.github.io/workshop-ngsintro/2511/home_contents.html"><img data-src="../../assets/logos/nbis.png" style="height:25px;max-width:none;"></a>
</div>
<div class="header-logo-right">
<a href="https://nbisweden.github.io/workshop-ngsintro/2511/home_schedule.html"><img data-src="../../assets/logos/scilifelab.png" style="height:25px;max-width:none;"></a>
</div>
</div>
</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'fade',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'fast',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/nbisweden\.github\.io\/workshop-ngsintro\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>